// @initia/initia.js/dist/core/evm/msgs/MsgUpdateEvmParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateEvmParams = void 0;
const json_1 = require("../../../util/json");
const EvmParams_1 = require("../EvmParams");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/minievm/evm/v1/tx");
class MsgUpdateEvmParams extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param params params defines the x/evm parameters to update
     */
    constructor(authority, params) {
        super();
        this.authority = authority;
        this.params = params;
    }
    static fromAmino(data) {
        const { value: { authority, params }, } = data;
        return new MsgUpdateEvmParams(authority, EvmParams_1.EvmParams.fromAmino(params));
    }
    toAmino() {
        const { authority, params } = this;
        return {
            type: 'evm/MsgUpdateParams',
            value: {
                authority,
                params: params.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, params } = data;
        return new MsgUpdateEvmParams(authority, EvmParams_1.EvmParams.fromData(params));
    }
    toData() {
        const { authority, params } = this;
        return {
            '@type': '/minievm.evm.v1.MsgUpdateParams',
            authority,
            params: params.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateEvmParams(data.authority, EvmParams_1.EvmParams.fromProto(data.params));
    }
    toProto() {
        const { authority, params } = this;
        return tx_1.MsgUpdateParams.fromPartial({
            authority,
            params: params.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/minievm.evm.v1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateEvmParams.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    }
}
exports.MsgUpdateEvmParams = MsgUpdateEvmParams;
