// @initia/initia.js/dist/core/opchild/msgs/MsgAddValidator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgAddValidator = void 0;
const json_1 = require("../../../util/json");
const PublicKey_1 = require("../../PublicKey");
const tx_1 = require("@initia/opinit.proto/opinit/opchild/v1/tx");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
class MsgAddValidator extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param moniker
     * @param validator_address
     * @param pubkey
     */
    constructor(authority, moniker, validator_address, pubkey) {
        super();
        this.authority = authority;
        this.moniker = moniker;
        this.validator_address = validator_address;
        this.pubkey = pubkey;
    }
    static fromAmino(data) {
        const { value: { authority, moniker, validator_address, pubkey }, } = data;
        return new MsgAddValidator(authority, moniker, validator_address, PublicKey_1.ValConsPublicKey.fromAmino(pubkey));
    }
    toAmino() {
        const { authority, moniker, validator_address, pubkey } = this;
        return {
            type: 'opchild/MsgAddValidator',
            value: {
                authority,
                moniker,
                validator_address,
                pubkey: pubkey.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, moniker, validator_address, pubkey } = data;
        return new MsgAddValidator(authority, moniker, validator_address, PublicKey_1.ValConsPublicKey.fromData(pubkey));
    }
    toData() {
        const { authority, moniker, validator_address, pubkey } = this;
        return {
            '@type': '/opinit.opchild.v1.MsgAddValidator',
            authority,
            moniker,
            validator_address,
            pubkey: pubkey.toData(),
        };
    }
    static fromProto(data) {
        return new MsgAddValidator(data.authority, data.moniker, data.validatorAddress, PublicKey_1.ValConsPublicKey.unpackAny(data.pubkey));
    }
    toProto() {
        const { authority, moniker, validator_address, pubkey } = this;
        return tx_1.MsgAddValidator.fromPartial({
            authority,
            moniker,
            validatorAddress: validator_address,
            pubkey: pubkey.packAny(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/opinit.opchild.v1.MsgAddValidator',
            value: tx_1.MsgAddValidator.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgAddValidator.fromProto(tx_1.MsgAddValidator.decode(msgAny.value));
    }
}
exports.MsgAddValidator = MsgAddValidator;
