// @initia/initia.js/dist/core/opchild/msgs/MsgRemoveValidator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgRemoveValidator = void 0;
const json_1 = require("../../../util/json");
const tx_1 = require("@initia/opinit.proto/opinit/opchild/v1/tx");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
class MsgRemoveValidator extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param validator_address
     */
    constructor(authority, validator_address) {
        super();
        this.authority = authority;
        this.validator_address = validator_address;
    }
    static fromAmino(data) {
        const { value: { authority, validator_address }, } = data;
        return new MsgRemoveValidator(authority, validator_address);
    }
    toAmino() {
        const { authority, validator_address } = this;
        return {
            type: 'opchild/MsgRemoveValidator',
            value: {
                authority,
                validator_address,
            },
        };
    }
    static fromData(data) {
        const { authority, validator_address } = data;
        return new MsgRemoveValidator(authority, validator_address);
    }
    toData() {
        const { authority, validator_address } = this;
        return {
            '@type': '/opinit.opchild.v1.MsgRemoveValidator',
            authority,
            validator_address,
        };
    }
    static fromProto(data) {
        return new MsgRemoveValidator(data.authority, data.validatorAddress);
    }
    toProto() {
        const { authority, validator_address } = this;
        return tx_1.MsgRemoveValidator.fromPartial({
            authority,
            validatorAddress: validator_address,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/opinit.opchild.v1.MsgRemoveValidator',
            value: tx_1.MsgRemoveValidator.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgRemoveValidator.fromProto(tx_1.MsgRemoveValidator.decode(msgAny.value));
    }
}
exports.MsgRemoveValidator = MsgRemoveValidator;
