// @initia/initia.js/dist/core/opchild/msgs/MsgUpdateOpchildParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateOpchildParams = void 0;
const json_1 = require("../../../util/json");
const OpchildParams_1 = require("../OpchildParams");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/opinit.proto/opinit/opchild/v1/tx");
class MsgUpdateOpchildParams extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param params params defines the x/move parameters to update
     */
    constructor(authority, params) {
        super();
        this.authority = authority;
        this.params = params;
    }
    static fromAmino(data) {
        const { value: { authority, params }, } = data;
        return new MsgUpdateOpchildParams(authority, OpchildParams_1.OpchildParams.fromAmino(params));
    }
    toAmino() {
        const { authority, params } = this;
        return {
            type: 'opchild/MsgUpdateParams',
            value: {
                authority,
                params: params.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, params } = data;
        return new MsgUpdateOpchildParams(authority, OpchildParams_1.OpchildParams.fromData(params));
    }
    toData() {
        const { authority, params } = this;
        return {
            '@type': '/opinit.opchild.v1.MsgUpdateParams',
            authority,
            params: params.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateOpchildParams(data.authority, OpchildParams_1.OpchildParams.fromProto(data.params));
    }
    toProto() {
        const { authority, params } = this;
        return tx_1.MsgUpdateParams.fromPartial({
            authority,
            params: params.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/opinit.opchild.v1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateOpchildParams.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    }
}
exports.MsgUpdateOpchildParams = MsgUpdateOpchildParams;
