// @initia/initia.js/dist/core/ibc/core/connection/msgs/MsgConnectionOpenAck.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgConnectionOpenAck = void 0;
const json_1 = require("../../../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/core/connection/v1/tx");
const core_1 = require("../../../core");
/**
 * MsgConnectionOpenAck defines a msg sent by a Relayer to Chain A to
 * acknowledge the change of connection state to TRYOPEN on Chain B.
 */
class MsgConnectionOpenAck extends json_1.JSONSerializable {
    /**
     * @param connection_id
     * @param counterparty_connection_id
     * @param version
     * @param client_state
     * @param proof_height proof of the initialization the connection on Chain B: `UNITIALIZED -> TRYOPEN`
     * @param proof_try proof of client state included in message
     * @param proof_client proof of client consensus state
     * @param proof_consensus
     * @param consenesus_height
     * @param signer signer address
     */
    constructor(connection_id, counterparty_connection_id, version, client_state, proof_height, proof_try, proof_client, proof_consensus, consensus_height, signer) {
        super();
        this.connection_id = connection_id;
        this.counterparty_connection_id = counterparty_connection_id;
        this.version = version;
        this.client_state = client_state;
        this.proof_height = proof_height;
        this.proof_try = proof_try;
        this.proof_client = proof_client;
        this.proof_consensus = proof_consensus;
        this.consensus_height = consensus_height;
        this.signer = signer;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { connection_id, counterparty_connection_id, version, client_state, proof_height, proof_try, proof_client, proof_consensus, consensus_height, signer, } = data;
        return new MsgConnectionOpenAck(connection_id, counterparty_connection_id, version ? core_1.IbcVersion.fromData(version) : undefined, client_state, proof_height ? core_1.Height.fromData(proof_height) : undefined, proof_try, proof_client, proof_consensus, consensus_height ? core_1.Height.fromData(consensus_height) : undefined, signer);
    }
    toData() {
        const { connection_id, counterparty_connection_id, version, client_state, proof_height, proof_try, proof_client, proof_consensus, consensus_height, signer, } = this;
        return {
            '@type': '/ibc.core.connection.v1.MsgConnectionOpenAck',
            connection_id,
            counterparty_connection_id,
            version: version === null || version === void 0 ? void 0 : version.toData(),
            client_state,
            proof_height: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toData(),
            proof_try,
            proof_client,
            proof_consensus,
            consensus_height: consensus_height === null || consensus_height === void 0 ? void 0 : consensus_height.toData(),
            signer,
        };
    }
    static fromProto(proto) {
        return new MsgConnectionOpenAck(proto.connectionId, proto.counterpartyConnectionId, proto.version ? core_1.IbcVersion.fromProto(proto.version) : undefined, proto.clientState, proto.proofHeight ? core_1.Height.fromProto(proto.proofHeight) : undefined, Buffer.from(proto.proofTry).toString('base64'), Buffer.from(proto.proofClient).toString('base64'), Buffer.from(proto.proofConsensus).toString('base64'), proto.consensusHeight
            ? core_1.Height.fromProto(proto.consensusHeight)
            : undefined, proto.signer);
    }
    toProto() {
        const { connection_id, counterparty_connection_id, version, client_state, proof_height, proof_try, proof_client, proof_consensus, consensus_height, signer, } = this;
        return tx_1.MsgConnectionOpenAck.fromPartial({
            connectionId: connection_id,
            counterpartyConnectionId: counterparty_connection_id,
            version: version === null || version === void 0 ? void 0 : version.toProto(),
            clientState: client_state,
            proofHeight: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toProto(),
            proofTry: Buffer.from(proof_try, 'base64'),
            proofClient: Buffer.from(proof_client, 'base64'),
            proofConsensus: Buffer.from(proof_consensus, 'base64'),
            consensusHeight: consensus_height === null || consensus_height === void 0 ? void 0 : consensus_height.toProto(),
            signer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.core.connection.v1.MsgConnectionOpenAck',
            value: tx_1.MsgConnectionOpenAck.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgConnectionOpenAck.fromProto(tx_1.MsgConnectionOpenAck.decode(msgAny.value));
    }
}
exports.MsgConnectionOpenAck = MsgConnectionOpenAck;
