// @initia/initia.js/dist/core/ibc/core/connection/msgs/MsgConnectionOpenConfirm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgConnectionOpenConfirm = void 0;
const json_1 = require("../../../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/core/connection/v1/tx");
const Height_1 = require("../../client/Height");
/**
 * MsgConnectionOpenConfirm defines a msg sent by a Relayer to Chain B to
 * acknowledge the change of connection state to OPEN on Chain A.
 */
class MsgConnectionOpenConfirm extends json_1.JSONSerializable {
    /**
     * @param connection_id
     * @param proof_ack proof for the change of the connection state on Chain A: `INIT -> OPEN`
     * @param proof_height
     * @param signer signer address
     */
    constructor(connection_id, proof_ack, proof_height, signer) {
        super();
        this.connection_id = connection_id;
        this.proof_ack = proof_ack;
        this.proof_height = proof_height;
        this.signer = signer;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { connection_id, proof_ack, proof_height, signer } = data;
        return new MsgConnectionOpenConfirm(connection_id, proof_ack, proof_height ? Height_1.Height.fromData(proof_height) : undefined, signer);
    }
    toData() {
        const { connection_id, proof_ack, proof_height, signer } = this;
        return {
            '@type': '/ibc.core.connection.v1.MsgConnectionOpenConfirm',
            connection_id,
            proof_ack,
            proof_height: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toData(),
            signer,
        };
    }
    static fromProto(proto) {
        return new MsgConnectionOpenConfirm(proto.connectionId, Buffer.from(proto.proofAck).toString('base64'), proto.proofHeight ? Height_1.Height.fromProto(proto.proofHeight) : undefined, proto.signer);
    }
    toProto() {
        const { connection_id, proof_ack, proof_height, signer } = this;
        return tx_1.MsgConnectionOpenConfirm.fromPartial({
            connectionId: connection_id,
            proofAck: Buffer.from(proof_ack, 'base64'),
            proofHeight: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toProto(),
            signer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.core.connection.v1.MsgConnectionOpenConfirm',
            value: tx_1.MsgConnectionOpenConfirm.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgConnectionOpenConfirm.fromProto(tx_1.MsgConnectionOpenConfirm.decode(msgAny.value));
    }
}
exports.MsgConnectionOpenConfirm = MsgConnectionOpenConfirm;
