// @initia/initia.js/dist/core/ibc/core/connection/msgs/MsgConnectionOpenTry.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgConnectionOpenTry = void 0;
const json_1 = require("../../../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/core/connection/v1/tx");
const ConnectionCounterparty_1 = require("../ConnectionCounterparty");
const IbcVersion_1 = require("../IbcVersion");
const client_1 = require("../../client");
const long_1 = __importDefault(require("long"));
/**
 *  MsgConnectionOpenTry defines a msg sent by a Relayer to try to open a connection on Chain B.
 */
class MsgConnectionOpenTry extends json_1.JSONSerializable {
    /**
     * @param client_id in the case of crossing hello's, when both chains call OpenInit, we need the connection identifier of the previous connection in state INIT
     * @param client_state
     * @param counterparty
     * @param delay_period
     * @param counterparty_versions
     * @param proof_height proof of the initialization the connection on Chain A: `UNITIALIZED -> INIT`
     * @param proof_init proof of client state included in message
     * @param proof_client proof of client consensus state
     * @param proof_consensus
     * @param consensus_height
     * @param signer signer address
     */
    constructor(client_id, client_state, counterparty, delay_period, counterparty_versions, proof_height, proof_init, proof_client, proof_consensus, consensus_height, signer) {
        super();
        this.client_id = client_id;
        this.client_state = client_state;
        this.counterparty = counterparty;
        this.delay_period = delay_period;
        this.counterparty_versions = counterparty_versions;
        this.proof_height = proof_height;
        this.proof_init = proof_init;
        this.proof_client = proof_client;
        this.proof_consensus = proof_consensus;
        this.consensus_height = consensus_height;
        this.signer = signer;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { client_id, client_state, counterparty, delay_period, counterparty_versions, proof_height, proof_init, proof_client, proof_consensus, consensus_height, signer, } = data;
        return new MsgConnectionOpenTry(client_id, client_state, counterparty ? ConnectionCounterparty_1.ConnectionCounterparty.fromData(counterparty) : undefined, Number.parseInt(delay_period), counterparty_versions.map(cv => IbcVersion_1.IbcVersion.fromData(cv)), proof_height ? client_1.Height.fromData(proof_height) : undefined, Buffer.from(proof_init).toString('base64'), Buffer.from(proof_client).toString('base64'), Buffer.from(proof_consensus).toString('base64'), consensus_height ? client_1.Height.fromData(consensus_height) : undefined, signer);
    }
    toData() {
        const { client_id, client_state, counterparty, delay_period, counterparty_versions, proof_height, proof_init, proof_client, proof_consensus, consensus_height, signer, } = this;
        return {
            '@type': '/ibc.core.connection.v1.MsgConnectionOpenTry',
            client_id,
            client_state,
            counterparty: counterparty === null || counterparty === void 0 ? void 0 : counterparty.toData(),
            delay_period: delay_period.toFixed(),
            counterparty_versions: counterparty_versions.map(cv => cv.toData()),
            proof_height: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toData(),
            proof_init,
            proof_client,
            proof_consensus,
            consensus_height: consensus_height === null || consensus_height === void 0 ? void 0 : consensus_height.toData(),
            signer,
        };
    }
    static fromProto(proto) {
        return new MsgConnectionOpenTry(proto.clientId, proto.clientState, proto.counterparty
            ? ConnectionCounterparty_1.ConnectionCounterparty.fromProto(proto.counterparty)
            : undefined, proto.delayPeriod.toNumber(), proto.counterpartyVersions.map(cv => IbcVersion_1.IbcVersion.fromProto(cv)), proto.proofHeight ? client_1.Height.fromProto(proto.proofHeight) : undefined, Buffer.from(proto.proofInit).toString('base64'), Buffer.from(proto.proofClient).toString('base64'), Buffer.from(proto.proofConsensus).toString('base64'), proto.consensusHeight
            ? client_1.Height.fromProto(proto.consensusHeight)
            : undefined, proto.signer);
    }
    toProto() {
        const { client_id, client_state, counterparty, delay_period, counterparty_versions, proof_height, proof_init, proof_client, proof_consensus, consensus_height, signer, } = this;
        return tx_1.MsgConnectionOpenTry.fromPartial({
            clientId: client_id,
            clientState: client_state.toProto(),
            counterparty: counterparty === null || counterparty === void 0 ? void 0 : counterparty.toProto(),
            delayPeriod: long_1.default.fromNumber(delay_period),
            counterpartyVersions: counterparty_versions.map(cv => cv.toProto()),
            proofHeight: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toProto(),
            proofInit: Buffer.from(proof_init, 'base64'),
            proofClient: Buffer.from(proof_client, 'base64'),
            proofConsensus: Buffer.from(proof_consensus, 'base64'),
            consensusHeight: consensus_height === null || consensus_height === void 0 ? void 0 : consensus_height.toProto(),
            signer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.core.connection.v1.MsgConnectionOpenTry',
            value: tx_1.MsgConnectionOpenTry.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgConnectionOpenTry.fromProto(tx_1.MsgConnectionOpenTry.decode(msgAny.value));
    }
}
exports.MsgConnectionOpenTry = MsgConnectionOpenTry;
