// @initia/initia.js/dist/core/ibc/core/connection/msgs/MsgConnectionOpenInit.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgConnectionOpenInit = void 0;
const json_1 = require("../../../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/core/connection/v1/tx");
const ConnectionCounterparty_1 = require("../ConnectionCounterparty");
const IbcVersion_1 = require("../IbcVersion");
const long_1 = __importDefault(require("long"));
/**
 * MsgConnectionOpenInit defines the msg sent by an account on Chain A to initialize a connection with Chain B.
 */
class MsgConnectionOpenInit extends json_1.JSONSerializable {
    /**
     * @param client_id identifier of the port to use
     * @param counterparty
     * @param version
     * @param delay_period
     * @param signer signer address
     */
    constructor(client_id, counterparty, version, delay_period, signer) {
        super();
        this.client_id = client_id;
        this.counterparty = counterparty;
        this.version = version;
        this.delay_period = delay_period;
        this.signer = signer;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { client_id, counterparty, version, delay_period, signer } = data;
        return new MsgConnectionOpenInit(client_id, ConnectionCounterparty_1.ConnectionCounterparty.fromData(counterparty), version ? IbcVersion_1.IbcVersion.fromData(version) : undefined, Number.parseInt(delay_period), signer);
    }
    toData() {
        const { client_id, counterparty, version, delay_period, signer } = this;
        return {
            '@type': '/ibc.core.connection.v1.MsgConnectionOpenInit',
            client_id,
            counterparty: counterparty.toData(),
            version: version === null || version === void 0 ? void 0 : version.toData(),
            delay_period: delay_period.toFixed(),
            signer,
        };
    }
    static fromProto(proto) {
        return new MsgConnectionOpenInit(proto.clientId, ConnectionCounterparty_1.ConnectionCounterparty.fromProto(proto.counterparty), proto.version ? IbcVersion_1.IbcVersion.fromProto(proto.version) : undefined, proto.delayPeriod.toNumber(), proto.signer);
    }
    toProto() {
        const { client_id, counterparty, version, delay_period, signer } = this;
        return tx_1.MsgConnectionOpenInit.fromPartial({
            clientId: client_id,
            counterparty: counterparty.toProto(),
            version: version === null || version === void 0 ? void 0 : version.toProto(),
            delayPeriod: long_1.default.fromNumber(delay_period),
            signer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.core.connection.v1.MsgConnectionOpenInit',
            value: tx_1.MsgConnectionOpenInit.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgConnectionOpenInit.fromProto(tx_1.MsgConnectionOpenInit.decode(msgAny.value));
    }
}
exports.MsgConnectionOpenInit = MsgConnectionOpenInit;
