// @initia/initia.js/dist/core/ibc/core/channel/msgs/MsgChannelOpenConfirm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgChannelOpenConfirm = void 0;
const json_1 = require("../../../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/core/channel/v1/tx");
const client_1 = require("../../client");
/**
 *  MsgChannelOpenConfirm defines a msg sent by a Relayer to Chain B to acknowledge the change of channel state to OPEN on Chain A.
 */
class MsgChannelOpenConfirm extends json_1.JSONSerializable {
    /**
     * @param port_id identifier of the port to use
     * @param channel_id
     * @param proof_ack
     * @param proof_height
     * @param signer signer address
     */
    constructor(port_id, channel_id, proof_ack, proof_height, signer) {
        super();
        this.port_id = port_id;
        this.channel_id = channel_id;
        this.proof_ack = proof_ack;
        this.proof_height = proof_height;
        this.signer = signer;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { port_id, channel_id, proof_ack, proof_height, signer } = data;
        return new MsgChannelOpenConfirm(port_id, channel_id, proof_ack, proof_height ? client_1.Height.fromData(proof_height) : undefined, signer);
    }
    toData() {
        const { port_id, channel_id, proof_ack, proof_height, signer } = this;
        return {
            '@type': '/ibc.core.channel.v1.MsgChannelOpenConfirm',
            port_id,
            channel_id,
            proof_ack,
            proof_height: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toData(),
            signer,
        };
    }
    static fromProto(proto) {
        return new MsgChannelOpenConfirm(proto.portId, proto.channelId, Buffer.from(proto.proofAck).toString('base64'), proto.proofHeight ? client_1.Height.fromProto(proto.proofHeight) : undefined, proto.signer);
    }
    toProto() {
        const { port_id, channel_id, proof_ack, proof_height, signer } = this;
        return tx_1.MsgChannelOpenConfirm.fromPartial({
            portId: port_id,
            channelId: channel_id,
            proofAck: Buffer.from(proof_ack, 'base64'),
            proofHeight: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toProto(),
            signer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.core.channel.v1.MsgChannelOpenConfirm',
            value: tx_1.MsgChannelOpenConfirm.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgChannelOpenConfirm.fromProto(tx_1.MsgChannelOpenConfirm.decode(msgAny.value));
    }
}
exports.MsgChannelOpenConfirm = MsgChannelOpenConfirm;
