// @initia/initia.js/dist/core/ibc/core/channel/msgs/MsgChannelOpenTry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgChannelOpenTry = void 0;
const json_1 = require("../../../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/core/channel/v1/tx");
const core_1 = require("../../../core");
/**
 * MsgChannelOpenTry defines a msg sent by a Relayer to try to open a channel on Chain B
 */
class MsgChannelOpenTry extends json_1.JSONSerializable {
    /**
     * @param port_id identifier of the port to use
     * @param previous_channel_id
     * @param channel channel info
     * @param counterparty_version
     * @param proof_init
     * @param proof_height
     * @param signer signer address
     */
    constructor(port_id, previous_channel_id, channel, counterparty_version, proof_init, proof_height, signer) {
        super();
        this.port_id = port_id;
        this.previous_channel_id = previous_channel_id;
        this.channel = channel;
        this.counterparty_version = counterparty_version;
        this.proof_init = proof_init;
        this.proof_height = proof_height;
        this.signer = signer;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { port_id, previous_channel_id, channel, counterparty_version, proof_init, proof_height, signer, } = data;
        return new MsgChannelOpenTry(port_id, previous_channel_id, channel ? core_1.Channel.fromData(channel) : undefined, counterparty_version, proof_init, proof_height ? core_1.Height.fromData(proof_height) : undefined, signer);
    }
    toData() {
        const { port_id, previous_channel_id, channel, counterparty_version, proof_init, proof_height, signer, } = this;
        return {
            '@type': '/ibc.core.channel.v1.MsgChannelOpenTry',
            port_id,
            previous_channel_id,
            channel: channel === null || channel === void 0 ? void 0 : channel.toData(),
            counterparty_version,
            proof_init,
            proof_height: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toData(),
            signer,
        };
    }
    static fromProto(proto) {
        return new MsgChannelOpenTry(proto.portId, proto.previousChannelId, proto.channel ? core_1.Channel.fromProto(proto.channel) : undefined, proto.counterpartyVersion, Buffer.from(proto.proofInit).toString('base64'), proto.proofHeight ? core_1.Height.fromProto(proto.proofHeight) : undefined, proto.signer);
    }
    toProto() {
        const { port_id, previous_channel_id, channel, counterparty_version, proof_init, proof_height, signer, } = this;
        return tx_1.MsgChannelOpenTry.fromPartial({
            portId: port_id,
            previousChannelId: previous_channel_id,
            channel: channel === null || channel === void 0 ? void 0 : channel.toProto(),
            counterpartyVersion: counterparty_version,
            proofInit: Buffer.from(proof_init, 'base64'),
            proofHeight: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toProto(),
            signer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.core.channel.v1.MsgChannelOpenTry',
            value: tx_1.MsgChannelOpenTry.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgChannelOpenTry.fromProto(tx_1.MsgChannelOpenTry.decode(msgAny.value));
    }
}
exports.MsgChannelOpenTry = MsgChannelOpenTry;
