// @initia/initia.js/dist/core/ibc/core/channel/msgs/MsgChannelOpenInit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgChannelOpenInit = void 0;
const json_1 = require("../../../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/core/channel/v1/tx");
const Channel_1 = require("../Channel");
/**
 * MsgChannelOpenInit defines an sdk.Msg to initialize a channel handshake. It is called by a relayer on Chain A.
 */
class MsgChannelOpenInit extends json_1.JSONSerializable {
    /**
     * @param port_id identifier of the port to use
     * @param channel channel info
     * @param signer signer address
     */
    constructor(port_id, channel, signer) {
        super();
        this.port_id = port_id;
        this.channel = channel;
        this.signer = signer;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { port_id, channel, signer } = data;
        return new MsgChannelOpenInit(port_id, channel ? Channel_1.Channel.fromData(channel) : undefined, signer);
    }
    toData() {
        const { port_id, channel, signer } = this;
        return {
            '@type': '/ibc.core.channel.v1.MsgChannelOpenInit',
            port_id,
            channel: channel === null || channel === void 0 ? void 0 : channel.toData(),
            signer,
        };
    }
    static fromProto(proto) {
        return new MsgChannelOpenInit(proto.portId, proto.channel ? Channel_1.Channel.fromProto(proto.channel) : undefined, proto.signer);
    }
    toProto() {
        const { port_id, channel, signer } = this;
        return tx_1.MsgChannelOpenInit.fromPartial({
            portId: port_id,
            channel: channel === null || channel === void 0 ? void 0 : channel.toProto(),
            signer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.core.channel.v1.MsgChannelOpenInit',
            value: tx_1.MsgChannelOpenInit.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgChannelOpenInit.fromProto(tx_1.MsgChannelOpenInit.decode(msgAny.value));
    }
}
exports.MsgChannelOpenInit = MsgChannelOpenInit;
