// @initia/initia.js/dist/core/ibc/core/channel/msgs/MsgChannelOpenAck.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgChannelOpenAck = void 0;
const json_1 = require("../../../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/core/channel/v1/tx");
const client_1 = require("../../client");
/**
 * MsgChannelOpenAck defines a msg sent by a Relayer to Chain A to acknowledge the change of channel state to TRYOPEN on Chain B.
 */
class MsgChannelOpenAck extends json_1.JSONSerializable {
    /**
     * @param port_id identifier of the port to use
     * @param channel_id
     * @param counterparty_channel_id
     * @param counterparty_version
     * @param proof_try
     * @param proof_height
     * @param signer signer address
     */
    constructor(port_id, channel_id, counterparty_channel_id, counterparty_version, proof_try, proof_height, signer) {
        super();
        this.port_id = port_id;
        this.channel_id = channel_id;
        this.counterparty_channel_id = counterparty_channel_id;
        this.counterparty_version = counterparty_version;
        this.proof_try = proof_try;
        this.proof_height = proof_height;
        this.signer = signer;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { port_id, channel_id, counterparty_channel_id, counterparty_version, proof_try, proof_height, signer, } = data;
        return new MsgChannelOpenAck(port_id, channel_id, counterparty_channel_id, counterparty_version, proof_try, proof_height ? client_1.Height.fromData(proof_height) : undefined, signer);
    }
    toData() {
        const { port_id, channel_id, counterparty_channel_id, counterparty_version, proof_try, proof_height, signer, } = this;
        return {
            '@type': '/ibc.core.channel.v1.MsgChannelOpenAck',
            port_id,
            channel_id,
            counterparty_channel_id,
            counterparty_version,
            proof_try,
            proof_height: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toData(),
            signer,
        };
    }
    static fromProto(proto) {
        return new MsgChannelOpenAck(proto.portId, proto.channelId, proto.counterpartyChannelId, proto.counterpartyVersion, Buffer.from(proto.proofTry).toString('base64'), proto.proofHeight ? client_1.Height.fromProto(proto.proofHeight) : undefined, proto.signer);
    }
    toProto() {
        const { port_id, channel_id, counterparty_channel_id, counterparty_version, proof_try, proof_height, signer, } = this;
        return tx_1.MsgChannelOpenAck.fromPartial({
            portId: port_id,
            channelId: channel_id,
            counterpartyChannelId: counterparty_channel_id,
            counterpartyVersion: counterparty_version,
            proofTry: Buffer.from(proof_try, 'base64'),
            proofHeight: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toProto(),
            signer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.core.channel.v1.MsgChannelOpenAck',
            value: tx_1.MsgChannelOpenAck.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgChannelOpenAck.fromProto(tx_1.MsgChannelOpenAck.decode(msgAny.value));
    }
}
exports.MsgChannelOpenAck = MsgChannelOpenAck;
