// @initia/initia.js/dist/core/ibc/core/channel/msgs/MsgRecvPacket.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgRecvPacket = void 0;
const json_1 = require("../../../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/core/channel/v1/tx");
const core_1 = require("../../../core");
/**
 * MsgRecvPacket receives incoming IBC packet
 */
class MsgRecvPacket extends json_1.JSONSerializable {
    /**
     * @param packet
     * @param proof_commitment
     * @param proof_height
     * @param signer signer address
     */
    constructor(packet, proof_commitment, proof_height, signer) {
        super();
        this.packet = packet;
        this.proof_commitment = proof_commitment;
        this.proof_height = proof_height;
        this.signer = signer;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { packet, proof_commitment, proof_height, signer } = data;
        return new MsgRecvPacket(packet ? core_1.Packet.fromData(packet) : undefined, proof_commitment, proof_height ? core_1.Height.fromData(proof_height) : undefined, signer);
    }
    toData() {
        const { packet, proof_commitment, proof_height, signer } = this;
        return {
            '@type': '/ibc.core.channel.v1.MsgRecvPacket',
            packet: packet === null || packet === void 0 ? void 0 : packet.toData(),
            proof_commitment,
            proof_height: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toData(),
            signer,
        };
    }
    static fromProto(proto) {
        return new MsgRecvPacket(proto.packet ? core_1.Packet.fromProto(proto.packet) : undefined, Buffer.from(proto.proofCommitment).toString('base64'), proto.proofHeight ? core_1.Height.fromProto(proto.proofHeight) : undefined, proto.signer);
    }
    toProto() {
        const { packet, proof_commitment, proof_height, signer } = this;
        return tx_1.MsgRecvPacket.fromPartial({
            packet: packet === null || packet === void 0 ? void 0 : packet.toProto(),
            proofCommitment: Buffer.from(proof_commitment, 'base64'),
            proofHeight: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toProto(),
            signer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.core.channel.v1.MsgRecvPacket',
            value: tx_1.MsgRecvPacket.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgRecvPacket.fromProto(tx_1.MsgRecvPacket.decode(msgAny.value));
    }
}
exports.MsgRecvPacket = MsgRecvPacket;
