// @initia/initia.js/dist/core/ibc/core/channel/msgs/MsgRecvAcknowledgement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgAcknowledgement = void 0;
const json_1 = require("../../../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/core/channel/v1/tx");
const core_1 = require("../../../core");
/**
 * MsgAcknowledgement receives incoming IBC acknowledgement
 */
class MsgAcknowledgement extends json_1.JSONSerializable {
    /**
     * @param packet
     * @param acknowledgement
     * @param proof_acked
     * @param proof_height
     * @param signer signer address
     */
    constructor(packet, acknowledgement, proof_acked, proof_height, signer) {
        super();
        this.packet = packet;
        this.acknowledgement = acknowledgement;
        this.proof_acked = proof_acked;
        this.proof_height = proof_height;
        this.signer = signer;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { packet, acknowledgement, proof_acked, proof_height, signer } = data;
        return new MsgAcknowledgement(packet ? core_1.Packet.fromData(packet) : undefined, proof_acked, acknowledgement, proof_height ? core_1.Height.fromData(proof_height) : undefined, signer);
    }
    toData() {
        const { packet, acknowledgement, proof_acked, proof_height, signer } = this;
        return {
            '@type': '/ibc.core.channel.v1.MsgAcknowledgement',
            packet: packet === null || packet === void 0 ? void 0 : packet.toData(),
            acknowledgement,
            proof_acked,
            proof_height: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toData(),
            signer,
        };
    }
    static fromProto(proto) {
        return new MsgAcknowledgement(proto.packet ? core_1.Packet.fromProto(proto.packet) : undefined, Buffer.from(proto.acknowledgement).toString('base64'), Buffer.from(proto.proofAcked).toString('base64'), proto.proofHeight ? core_1.Height.fromProto(proto.proofHeight) : undefined, proto.signer);
    }
    toProto() {
        const { packet, acknowledgement, proof_acked, proof_height, signer } = this;
        return tx_1.MsgAcknowledgement.fromPartial({
            packet: packet === null || packet === void 0 ? void 0 : packet.toProto(),
            acknowledgement: Buffer.from(acknowledgement, 'base64'),
            proofAcked: Buffer.from(proof_acked, 'base64'),
            proofHeight: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toProto(),
            signer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.core.channel.v1.MsgAcknowledgement',
            value: tx_1.MsgAcknowledgement.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgAcknowledgement.fromProto(tx_1.MsgAcknowledgement.decode(msgAny.value));
    }
}
exports.MsgAcknowledgement = MsgAcknowledgement;
