// @initia/initia.js/dist/core/ibc/core/channel/msgs/MsgTimeout.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgTimeout = void 0;
const json_1 = require("../../../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/core/channel/v1/tx");
const core_1 = require("../../../core");
const long_1 = __importDefault(require("long"));
/**
 * MsgTimeout receives timed-out packet
 */
class MsgTimeout extends json_1.JSONSerializable {
    /**
     * @param packet
     * @param proof_unreceived
     * @param proof_height
     * @param next_seuqnce_recv
     * @param signer signer address
     */
    constructor(packet, proof_unreceived, proof_height, next_sequence_recv, signer) {
        super();
        this.packet = packet;
        this.proof_unreceived = proof_unreceived;
        this.proof_height = proof_height;
        this.next_sequence_recv = next_sequence_recv;
        this.signer = signer;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { packet, proof_unreceived, proof_height, next_sequence_recv, signer, } = data;
        return new MsgTimeout(packet ? core_1.Packet.fromData(packet) : undefined, proof_unreceived, proof_height ? core_1.Height.fromData(proof_height) : undefined, Number.parseInt(next_sequence_recv), signer);
    }
    toData() {
        const { packet, proof_unreceived, proof_height, next_sequence_recv, signer, } = this;
        return {
            '@type': '/ibc.core.channel.v1.MsgTimeout',
            packet: packet === null || packet === void 0 ? void 0 : packet.toData(),
            proof_unreceived,
            proof_height: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toData(),
            next_sequence_recv: next_sequence_recv.toFixed(),
            signer,
        };
    }
    static fromProto(proto) {
        return new MsgTimeout(proto.packet ? core_1.Packet.fromProto(proto.packet) : undefined, Buffer.from(proto.proofUnreceived).toString('base64'), proto.proofHeight ? core_1.Height.fromProto(proto.proofHeight) : undefined, proto.nextSequenceRecv.toNumber(), proto.signer);
    }
    toProto() {
        const { packet, proof_unreceived, proof_height, next_sequence_recv, signer, } = this;
        return tx_1.MsgTimeout.fromPartial({
            packet: packet === null || packet === void 0 ? void 0 : packet.toProto(),
            proofUnreceived: Buffer.from(proof_unreceived, 'base64'),
            proofHeight: proof_height === null || proof_height === void 0 ? void 0 : proof_height.toProto(),
            nextSequenceRecv: long_1.default.fromNumber(next_sequence_recv),
            signer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.core.channel.v1.MsgTimeout',
            value: tx_1.MsgTimeout.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgTimeout.fromProto(tx_1.MsgTimeout.decode(msgAny.value));
    }
}
exports.MsgTimeout = MsgTimeout;
