// @initia/initia.js/dist/core/ibc/core/channel/msgs/MsgUpdateIbcChannelParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateIbcChannelParams = void 0;
const json_1 = require("../../../../../util/json");
const IbcChannelParams_1 = require("../IbcChannelParams");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/core/channel/v1/tx");
class MsgUpdateIbcChannelParams extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param params params defines the channel parameters to update
     */
    constructor(authority, params) {
        super();
        this.authority = authority;
        this.params = params;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { authority, params } = data;
        return new MsgUpdateIbcChannelParams(authority, IbcChannelParams_1.IbcChannelParams.fromData(params));
    }
    toData() {
        const { authority, params } = this;
        return {
            '@type': '/ibc.core.channel.v1.MsgUpdateParams',
            authority,
            params: params.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateIbcChannelParams(data.authority, IbcChannelParams_1.IbcChannelParams.fromProto(data.params));
    }
    toProto() {
        const { authority, params } = this;
        return tx_1.MsgUpdateParams.fromPartial({
            authority,
            params: params.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.core.channel.v1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateIbcChannelParams.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    }
}
exports.MsgUpdateIbcChannelParams = MsgUpdateIbcChannelParams;
