// @initia/initia.js/dist/core/group/msgs/MsgLeaveGroup.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgLeaveGroup = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmos/group/v1/tx");
const long_1 = __importDefault(require("long"));
class MsgLeaveGroup extends json_1.JSONSerializable {
    /**
     * @param address the account address of the group member
     * @param group_id the unique ID of the group
     */
    constructor(address, group_id) {
        super();
        this.address = address;
        this.group_id = group_id;
    }
    static fromAmino(data) {
        const { value: { address, group_id }, } = data;
        return new MsgLeaveGroup(address, Number.parseInt(group_id));
    }
    toAmino() {
        const { address, group_id } = this;
        return {
            type: 'cosmos-sdk/group/MsgLeaveGroup',
            value: {
                address: address,
                group_id: group_id.toString(),
            },
        };
    }
    static fromData(data) {
        const { address, group_id } = data;
        return new MsgLeaveGroup(address, Number.parseInt(group_id));
    }
    toData() {
        const { address, group_id } = this;
        return {
            '@type': '/cosmos.group.v1.MsgLeaveGroup',
            address,
            group_id: group_id.toString(),
        };
    }
    static fromProto(data) {
        return new MsgLeaveGroup(data.address, data.groupId.toNumber());
    }
    toProto() {
        const { address, group_id } = this;
        return tx_1.MsgLeaveGroup.fromPartial({
            address,
            groupId: long_1.default.fromNumber(group_id),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgLeaveGroup',
            value: tx_1.MsgLeaveGroup.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgLeaveGroup.fromProto(tx_1.MsgLeaveGroup.decode(msgAny.value));
    }
}
exports.MsgLeaveGroup = MsgLeaveGroup;
