// @initia/initia.js/dist/core/group/msgs/MsgUpdateGroupMetadata.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateGroupMetadata = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmos/group/v1/tx");
const long_1 = __importDefault(require("long"));
class MsgUpdateGroupMetadata extends json_1.JSONSerializable {
    /**
     * @param admin the account address of the group admin
     * @param group_id the unique ID of the group
     * @param metadata the updated group's metadata
     */
    constructor(admin, group_id, metadata) {
        super();
        this.admin = admin;
        this.group_id = group_id;
        this.metadata = metadata;
    }
    static fromAmino(data) {
        const { value: { admin, group_id, metadata }, } = data;
        return new MsgUpdateGroupMetadata(admin, Number.parseInt(group_id), metadata);
    }
    toAmino() {
        const { admin, group_id, metadata } = this;
        return {
            type: 'cosmos-sdk/MsgUpdateGroupMetadata',
            value: {
                admin,
                group_id: group_id.toString(),
                metadata,
            },
        };
    }
    static fromData(data) {
        const { admin, group_id, metadata } = data;
        return new MsgUpdateGroupMetadata(admin, Number.parseInt(group_id), metadata);
    }
    toData() {
        const { admin, group_id, metadata } = this;
        return {
            '@type': '/cosmos.group.v1.MsgUpdateGroupMetadata',
            admin,
            group_id: group_id.toString(),
            metadata,
        };
    }
    static fromProto(data) {
        return new MsgUpdateGroupMetadata(data.admin, data.groupId.toNumber(), data.metadata);
    }
    toProto() {
        const { admin, group_id, metadata } = this;
        return tx_1.MsgUpdateGroupMetadata.fromPartial({
            admin,
            groupId: long_1.default.fromNumber(group_id),
            metadata,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgUpdateGroupMetadata',
            value: tx_1.MsgUpdateGroupMetadata.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateGroupMetadata.fromProto(tx_1.MsgUpdateGroupMetadata.decode(msgAny.value));
    }
}
exports.MsgUpdateGroupMetadata = MsgUpdateGroupMetadata;
