// @initia/initia.js/dist/core/group/msgs/MsgUpdateGroupMembers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateGroupMembers = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const GroupMember_1 = require("../GroupMember");
const tx_1 = require("@initia/initia.proto/cosmos/group/v1/tx");
const long_1 = __importDefault(require("long"));
class MsgUpdateGroupMembers extends json_1.JSONSerializable {
    /**
     * @param admin the account address of the group admin
     * @param group_id the unique ID of the group
     * @param member_updates list of members to update, set weight to 0 to remove a member
     */
    constructor(admin, group_id, member_updates) {
        super();
        this.admin = admin;
        this.group_id = group_id;
        this.member_updates = member_updates;
    }
    static fromAmino(data) {
        const { value: { admin, group_id, member_updates }, } = data;
        return new MsgUpdateGroupMembers(admin, Number.parseInt(group_id), member_updates.map(GroupMember_1.MemberRequest.fromAmino));
    }
    toAmino() {
        const { admin, group_id, member_updates } = this;
        return {
            type: 'cosmos-sdk/MsgUpdateGroupMembers',
            value: {
                admin,
                group_id: group_id.toString(),
                member_updates: member_updates.map(d => d.toAmino()),
            },
        };
    }
    static fromData(data) {
        const { admin, group_id, member_updates } = data;
        return new MsgUpdateGroupMembers(admin, Number.parseInt(group_id), member_updates.map(GroupMember_1.MemberRequest.fromData));
    }
    toData() {
        const { admin, group_id, member_updates } = this;
        return {
            '@type': '/cosmos.group.v1.MsgUpdateGroupMembers',
            admin,
            group_id: group_id.toString(),
            member_updates: member_updates.map(d => d.toData()),
        };
    }
    static fromProto(data) {
        return new MsgUpdateGroupMembers(data.admin, data.groupId.toNumber(), data.memberUpdates.map(GroupMember_1.MemberRequest.fromProto));
    }
    toProto() {
        const { admin, group_id, member_updates } = this;
        return tx_1.MsgUpdateGroupMembers.fromPartial({
            admin,
            groupId: long_1.default.fromNumber(group_id),
            memberUpdates: member_updates.map(d => d.toProto()),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgUpdateGroupMembers',
            value: tx_1.MsgUpdateGroupMembers.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateGroupMembers.fromProto(tx_1.MsgUpdateGroupMembers.decode(msgAny.value));
    }
}
exports.MsgUpdateGroupMembers = MsgUpdateGroupMembers;
