// @initia/initia.js/dist/core/group/msgs/MsgUpdateGroupPolicyMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateGroupPolicyMetadata = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmos/group/v1/tx");
class MsgUpdateGroupPolicyMetadata extends json_1.JSONSerializable {
    /**
     * @param admin the account address of the group admin
     * @param group_policy_address the account address of group policy
     * @param metadata the group policy metadata to be updated
     */
    constructor(admin, group_policy_address, metadata) {
        super();
        this.admin = admin;
        this.group_policy_address = group_policy_address;
        this.metadata = metadata;
    }
    static fromAmino(data) {
        const { value: { admin, group_policy_address, metadata }, } = data;
        return new MsgUpdateGroupPolicyMetadata(admin, group_policy_address, metadata);
    }
    toAmino() {
        const { admin, group_policy_address, metadata } = this;
        return {
            type: 'cosmos-sdk/MsgUpdateGroupPolicyMetadata',
            value: {
                admin,
                group_policy_address,
                metadata,
            },
        };
    }
    static fromData(data) {
        const { admin, group_policy_address, metadata } = data;
        return new MsgUpdateGroupPolicyMetadata(admin, group_policy_address, metadata);
    }
    toData() {
        const { admin, group_policy_address, metadata } = this;
        return {
            '@type': '/cosmos.group.v1.MsgUpdateGroupPolicyMetadata',
            admin,
            group_policy_address,
            metadata,
        };
    }
    static fromProto(data) {
        return new MsgUpdateGroupPolicyMetadata(data.admin, data.groupPolicyAddress, data.metadata);
    }
    toProto() {
        const { admin, group_policy_address, metadata } = this;
        return tx_1.MsgUpdateGroupPolicyMetadata.fromPartial({
            admin,
            groupPolicyAddress: group_policy_address,
            metadata,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgUpdateGroupPolicyMetadata',
            value: tx_1.MsgUpdateGroupPolicyMetadata.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateGroupPolicyMetadata.fromProto(tx_1.MsgUpdateGroupPolicyMetadata.decode(msgAny.value));
    }
}
exports.MsgUpdateGroupPolicyMetadata = MsgUpdateGroupPolicyMetadata;
