// @initia/initia.js/dist/core/group/msgs/MsgWithdrawProposal.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgWithdrawProposal = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmos/group/v1/tx");
const long_1 = __importDefault(require("long"));
class MsgWithdrawProposal extends json_1.JSONSerializable {
    /**
     * @param proposal_id the unique ID of the proposal
     * @param address the admin of the group policy or one of the proposer of the proposal
     */
    constructor(proposal_id, address) {
        super();
        this.proposal_id = proposal_id;
        this.address = address;
    }
    static fromAmino(data) {
        const { value: { proposal_id, address }, } = data;
        return new MsgWithdrawProposal(Number.parseInt(proposal_id), address);
    }
    toAmino() {
        const { proposal_id, address } = this;
        return {
            type: 'cosmos-sdk/group/MsgWithdrawProposal',
            value: {
                proposal_id: proposal_id.toString(),
                address,
            },
        };
    }
    static fromData(data) {
        const { proposal_id, address } = data;
        return new MsgWithdrawProposal(Number.parseInt(proposal_id), address);
    }
    toData() {
        const { proposal_id, address } = this;
        return {
            '@type': '/cosmos.group.v1.MsgWithdrawProposal',
            proposal_id: proposal_id.toString(),
            address,
        };
    }
    static fromProto(data) {
        return new MsgWithdrawProposal(data.proposalId.toNumber(), data.address);
    }
    toProto() {
        const { proposal_id, address } = this;
        return tx_1.MsgWithdrawProposal.fromPartial({
            proposalId: long_1.default.fromNumber(proposal_id),
            address,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgWithdrawProposal',
            value: tx_1.MsgWithdrawProposal.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgWithdrawProposal.fromProto(tx_1.MsgWithdrawProposal.decode(msgAny.value));
    }
}
exports.MsgWithdrawProposal = MsgWithdrawProposal;
