// @initia/initia.js/dist/core/authz/authorizations/GenericAuthorization.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenericAuthorization = void 0;
const json_1 = require("../../../util/json");
const authz_1 = require("@initia/initia.proto/cosmos/authz/v1beta1/authz");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
class GenericAuthorization extends json_1.JSONSerializable {
    constructor(msg) {
        super();
        this.msg = msg;
    }
    static fromAmino(data) {
        return new GenericAuthorization(data.value.msg);
    }
    toAmino() {
        const { msg } = this;
        return {
            type: 'cosmos-sdk/GenericAuthorization',
            value: {
                msg,
            },
        };
    }
    static fromData(data) {
        return new GenericAuthorization(data.msg);
    }
    toData() {
        const { msg } = this;
        return {
            '@type': '/cosmos.authz.v1beta1.GenericAuthorization',
            msg,
        };
    }
    static fromProto(data) {
        return new GenericAuthorization(data.msg);
    }
    toProto() {
        return authz_1.GenericAuthorization.fromPartial({
            msg: this.msg,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.authz.v1beta1.GenericAuthorization',
            value: authz_1.GenericAuthorization.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return GenericAuthorization.fromProto(authz_1.GenericAuthorization.decode(msgAny.value));
    }
}
exports.GenericAuthorization = GenericAuthorization;
