// @initia/initia.js/dist/core/authz/authorizations/PublishAuthorization.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PublishAuthorization = void 0;
const json_1 = require("../../../util/json");
const authz_1 = require("@initia/initia.proto/initia/move/v1/authz");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
class PublishAuthorization extends json_1.JSONSerializable {
    constructor(module_names) {
        super();
        this.module_names = module_names;
    }
    static fromAmino(data) {
        return new PublishAuthorization(data.value.module_names);
    }
    toAmino() {
        return {
            type: 'move/PublishAuthorization',
            value: { module_names: this.module_names },
        };
    }
    static fromData(data) {
        return new PublishAuthorization(data.module_names);
    }
    toData() {
        return {
            '@type': '/initia.move.v1.PublishAuthorization',
            module_names: this.module_names,
        };
    }
    static fromProto(proto) {
        return new PublishAuthorization(proto.moduleNames);
    }
    toProto() {
        return authz_1.PublishAuthorization.fromPartial({
            moduleNames: this.module_names,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/initia.move.v1.PublishAuthorization',
            value: authz_1.PublishAuthorization.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return PublishAuthorization.fromProto(authz_1.PublishAuthorization.decode(msgAny.value));
    }
}
exports.PublishAuthorization = PublishAuthorization;
