// @initia/initia.js/dist/core/authz/authorizations/SendAuthorization.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendAuthorization = void 0;
const json_1 = require("../../../util/json");
const Coins_1 = require("../../Coins");
const authz_1 = require("@initia/initia.proto/cosmos/bank/v1beta1/authz");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
class SendAuthorization extends json_1.JSONSerializable {
    constructor(spend_limit, allow_list) {
        super();
        this.allow_list = allow_list;
        this.spend_limit = new Coins_1.Coins(spend_limit);
    }
    static fromAmino(data) {
        return new SendAuthorization(Coins_1.Coins.fromAmino(data.value.spend_limit), data.value.allow_list);
    }
    toAmino() {
        const { spend_limit, allow_list } = this;
        return {
            type: 'cosmos-sdk/SendAuthorization',
            value: {
                spend_limit: spend_limit.toAmino(),
                allow_list,
            },
        };
    }
    static fromData(data) {
        return new SendAuthorization(Coins_1.Coins.fromData(data.spend_limit), data.allow_list);
    }
    toData() {
        const { spend_limit, allow_list } = this;
        return {
            '@type': '/cosmos.bank.v1beta1.SendAuthorization',
            spend_limit: spend_limit.toAmino(),
            allow_list,
        };
    }
    static fromProto(proto) {
        return new SendAuthorization(Coins_1.Coins.fromProto(proto.spendLimit), proto.allowList);
    }
    toProto() {
        return authz_1.SendAuthorization.fromPartial({
            spendLimit: this.spend_limit.toProto(),
            allowList: this.allow_list,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.bank.v1beta1.SendAuthorization',
            value: authz_1.SendAuthorization.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return SendAuthorization.fromProto(authz_1.SendAuthorization.decode(msgAny.value));
    }
}
exports.SendAuthorization = SendAuthorization;
