// @initia/initia.js/dist/core/consensus/ABCIParams.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ABCIParams = void 0;
const json_1 = require("../../util/json");
const params_1 = require("@initia/initia.proto/tendermint/types/params");
const long_1 = __importDefault(require("long"));
class ABCIParams extends json_1.JSONSerializable {
    /**
     * @param vote_extensions_enable_height
     */
    constructor(vote_extensions_enable_height) {
        super();
        this.vote_extensions_enable_height = vote_extensions_enable_height;
    }
    static fromAmino(data) {
        return new ABCIParams(Number.parseInt(data.vote_extensions_enable_height));
    }
    toAmino() {
        return {
            vote_extensions_enable_height: this.vote_extensions_enable_height.toString(),
        };
    }
    static fromData(data) {
        return new ABCIParams(Number.parseInt(data.vote_extensions_enable_height));
    }
    toData() {
        return {
            vote_extensions_enable_height: this.vote_extensions_enable_height.toString(),
        };
    }
    static fromProto(data) {
        return new ABCIParams(data.voteExtensionsEnableHeight.toNumber());
    }
    toProto() {
        return params_1.ABCIParams.fromPartial({
            voteExtensionsEnableHeight: long_1.default.fromNumber(this.vote_extensions_enable_height),
        });
    }
}
exports.ABCIParams = ABCIParams;
