// @initia/initia.js/dist/core/consensus/ValidatorParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidatorParams = void 0;
const json_1 = require("../../util/json");
const params_1 = require("@initia/initia.proto/tendermint/types/params");
class ValidatorParams extends json_1.JSONSerializable {
    /**
     * @param pub_key_types
     */
    constructor(pub_key_types) {
        super();
        this.pub_key_types = pub_key_types;
    }
    static fromAmino(data) {
        return new ValidatorParams(data.pub_key_types);
    }
    toAmino() {
        return {
            pub_key_types: this.pub_key_types,
        };
    }
    static fromData(data) {
        return new ValidatorParams(data.pub_key_types);
    }
    toData() {
        return {
            pub_key_types: this.pub_key_types,
        };
    }
    static fromProto(data) {
        return new ValidatorParams(data.pubKeyTypes);
    }
    toProto() {
        return params_1.ValidatorParams.fromPartial({
            pubKeyTypes: this.pub_key_types,
        });
    }
}
exports.ValidatorParams = ValidatorParams;
