// @initia/initia.js/dist/core/ibc/applications/fetchprice/msgs/MsgActivate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgActivate = void 0;
const json_1 = require("../../../../../util/json");
const Duration_1 = require("../../../../Duration");
const tx_1 = require("@initia/initia.proto/ibc/applications/fetchprice/v1/tx");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
class MsgActivate extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param source_port
     * @param source_channel
     * @param timeout_duration
     */
    constructor(authority, source_port, source_channel, timeout_duration) {
        super();
        this.authority = authority;
        this.source_port = source_port;
        this.source_channel = source_channel;
        this.timeout_duration = timeout_duration;
    }
    static fromAmino(data) {
        const { value: { authority, source_port, source_channel, timeout_duration }, } = data;
        return new MsgActivate(authority, source_port, source_channel, Duration_1.Duration.fromAmino(timeout_duration));
    }
    toAmino() {
        const { authority, source_port, source_channel, timeout_duration } = this;
        return {
            type: 'fetchprice/MsgActivate',
            value: {
                authority,
                source_port,
                source_channel,
                timeout_duration: timeout_duration.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, source_port, source_channel, timeout_duration } = data;
        return new MsgActivate(authority, source_port, source_channel, Duration_1.Duration.fromData(timeout_duration));
    }
    toData() {
        const { authority, source_port, source_channel, timeout_duration } = this;
        return {
            '@type': '/ibc.applications.fetchprice.v1.MsgActivate',
            authority,
            source_port,
            source_channel,
            timeout_duration: timeout_duration.toData(),
        };
    }
    static fromProto(data) {
        return new MsgActivate(data.authority, data.sourcePort, data.sourceChannel, Duration_1.Duration.fromProto(data.timeoutDuration));
    }
    toProto() {
        const { authority, source_port, source_channel, timeout_duration } = this;
        return tx_1.MsgActivate.fromPartial({
            authority,
            sourcePort: source_port,
            sourceChannel: source_channel,
            timeoutDuration: timeout_duration.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.applications.fetchprice.v1.MsgActivate',
            value: tx_1.MsgActivate.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgActivate.fromProto(tx_1.MsgActivate.decode(msgAny.value));
    }
}
exports.MsgActivate = MsgActivate;
