// @initia/initia.js/dist/core/ibc/applications/fetchprice/msgs/MsgDeactivate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgDeactivate = void 0;
const json_1 = require("../../../../../util/json");
const tx_1 = require("@initia/initia.proto/ibc/applications/fetchprice/v1/tx");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
class MsgDeactivate extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     */
    constructor(authority) {
        super();
        this.authority = authority;
    }
    static fromAmino(data) {
        return new MsgDeactivate(data.value.authority);
    }
    toAmino() {
        return {
            type: 'fetchprice/MsgDeactivate',
            value: {
                authority: this.authority,
            },
        };
    }
    static fromData(data) {
        return new MsgDeactivate(data.authority);
    }
    toData() {
        return {
            '@type': '/ibc.applications.fetchprice.v1.MsgDeactivate',
            authority: this.authority,
        };
    }
    static fromProto(data) {
        return new MsgDeactivate(data.authority);
    }
    toProto() {
        return tx_1.MsgDeactivate.fromPartial({
            authority: this.authority,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.applications.fetchprice.v1.MsgDeactivate',
            value: tx_1.MsgDeactivate.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgDeactivate.fromProto(tx_1.MsgDeactivate.decode(msgAny.value));
    }
}
exports.MsgDeactivate = MsgDeactivate;
