// @initia/initia.js/dist/core/ibc/applications/fetchprice/msgs/MsgUpdateIbcFetchpriceParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateIbcFetchpriceParams = void 0;
const json_1 = require("../../../../../util/json");
const IbcFetchpriceParams_1 = require("../IbcFetchpriceParams");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/applications/fetchprice/v1/tx");
class MsgUpdateIbcFetchpriceParams extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param params params defines the parameters to update
     */
    constructor(authority, params) {
        super();
        this.authority = authority;
        this.params = params;
    }
    static fromAmino(data) {
        const { value: { authority, params }, } = data;
        return new MsgUpdateIbcFetchpriceParams(authority, IbcFetchpriceParams_1.IbcFetchpriceParams.fromAmino(params));
    }
    toAmino() {
        const { authority, params } = this;
        return {
            type: 'fetchprice/MsgUpdateParams',
            value: {
                authority,
                params: params.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, params } = data;
        return new MsgUpdateIbcFetchpriceParams(authority, IbcFetchpriceParams_1.IbcFetchpriceParams.fromData(params));
    }
    toData() {
        const { authority, params } = this;
        return {
            '@type': '/ibc.applications.fetchprice.v1.MsgUpdateParams',
            authority,
            params: params.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateIbcFetchpriceParams(data.authority, IbcFetchpriceParams_1.IbcFetchpriceParams.fromProto(data.params));
    }
    toProto() {
        const { authority, params } = this;
        return tx_1.MsgUpdateParams.fromPartial({
            authority,
            params: params.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.applications.fetchprice.v1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateIbcFetchpriceParams.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    }
}
exports.MsgUpdateIbcFetchpriceParams = MsgUpdateIbcFetchpriceParams;
