// @initia/initia.js/dist/core/gov/Deposit.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Deposit = void 0;
const Coins_1 = require("../Coins");
const json_1 = require("../../util/json");
const gov_1 = require("@initia/initia.proto/cosmos/gov/v1/gov");
const long_1 = __importDefault(require("long"));
/**
 * Defines an amount deposited by an account address to an active proposal
 */
class Deposit extends json_1.JSONSerializable {
    /**
     * @param proposal_id the unique id of the proposal
     * @param depositor the deposit addresses from the proposals
     * @param amount amount to be deposited by depositor
     */
    constructor(proposal_id, depositor, amount) {
        super();
        this.proposal_id = proposal_id;
        this.depositor = depositor;
        this.amount = new Coins_1.Coins(amount);
    }
    static fromAmino(data) {
        const { proposal_id, depositor, amount } = data;
        return new Deposit(Number.parseInt(proposal_id), depositor, Coins_1.Coins.fromAmino(amount));
    }
    toAmino() {
        const { proposal_id, depositor, amount } = this;
        return {
            proposal_id: proposal_id.toString(),
            depositor,
            amount: amount.toAmino(),
        };
    }
    static fromData(data) {
        const { proposal_id, depositor, amount } = data;
        return new Deposit(Number.parseInt(proposal_id), depositor, Coins_1.Coins.fromData(amount));
    }
    toData() {
        const { proposal_id, depositor, amount } = this;
        return {
            proposal_id: proposal_id.toString(),
            depositor,
            amount: amount.toData(),
        };
    }
    static fromProto(data) {
        return new Deposit(data.proposalId.toNumber(), data.depositor, Coins_1.Coins.fromProto(data.amount));
    }
    toProto() {
        const { proposal_id, depositor, amount } = this;
        return gov_1.Deposit.fromPartial({
            proposalId: long_1.default.fromNumber(proposal_id),
            depositor: depositor,
            amount: amount.toProto(),
        });
    }
}
exports.Deposit = Deposit;
