// @initia/initia.js/dist/core/auth/msgs/MsgUpdateAuthParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateAuthParams = void 0;
const json_1 = require("../../../util/json");
const AuthParams_1 = require("../AuthParams");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmos/auth/v1beta1/tx");
class MsgUpdateAuthParams extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param params params defines the x/auth parameters to update
     */
    constructor(authority, params) {
        super();
        this.authority = authority;
        this.params = params;
    }
    static fromAmino(data) {
        const { value: { authority, params }, } = data;
        return new MsgUpdateAuthParams(authority, AuthParams_1.AuthParams.fromAmino(params));
    }
    toAmino() {
        const { authority, params } = this;
        return {
            type: 'cosmos-sdk/x/auth/MsgUpdateParams',
            value: {
                authority,
                params: params.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, params } = data;
        return new MsgUpdateAuthParams(authority, AuthParams_1.AuthParams.fromData(params));
    }
    toData() {
        const { authority, params } = this;
        return {
            '@type': '/cosmos.auth.v1beta1.MsgUpdateParams',
            authority,
            params: params.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateAuthParams(data.authority, AuthParams_1.AuthParams.fromProto(data.params));
    }
    toProto() {
        const { authority, params } = this;
        return tx_1.MsgUpdateParams.fromPartial({
            authority,
            params: params.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.auth.v1beta1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateAuthParams.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    }
}
exports.MsgUpdateAuthParams = MsgUpdateAuthParams;
