// @initia/initia.js/dist/core/slashing/msgs/MsgUpdateSlashingParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateSlashingParams = void 0;
const json_1 = require("../../../util/json");
const SlashingParams_1 = require("../SlashingParams");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmos/slashing/v1beta1/tx");
class MsgUpdateSlashingParams extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param params params defines the x/slashing parameters to update
     */
    constructor(authority, params) {
        super();
        this.authority = authority;
        this.params = params;
    }
    static fromAmino(data) {
        const { value: { authority, params }, } = data;
        return new MsgUpdateSlashingParams(authority, SlashingParams_1.SlashingParams.fromAmino(params));
    }
    toAmino() {
        const { authority, params } = this;
        return {
            type: 'cosmos-sdk/x/slashing/MsgUpdateParams',
            value: {
                authority,
                params: params.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, params } = data;
        return new MsgUpdateSlashingParams(authority, SlashingParams_1.SlashingParams.fromData(params));
    }
    toData() {
        const { authority, params } = this;
        return {
            '@type': '/cosmos.slashing.v1beta1.MsgUpdateParams',
            authority,
            params: params.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateSlashingParams(data.authority, SlashingParams_1.SlashingParams.fromProto(data.params));
    }
    toProto() {
        const { authority, params } = this;
        return tx_1.MsgUpdateParams.fromPartial({
            authority,
            params: params.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.slashing.v1beta1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateSlashingParams.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    }
}
exports.MsgUpdateSlashingParams = MsgUpdateSlashingParams;
