// @initia/initia.js/dist/core/mstaking/Delegation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Delegation = void 0;
const json_1 = require("../../util/json");
const Coins_1 = require("../Coins");
const staking_1 = require("@initia/initia.proto/initia/mstaking/v1/staking");
/**
 * Stores information about the status of a delegation between a delegator and validator, fetched from the blockchain.
 */
class Delegation extends json_1.JSONSerializable {
    /**
     * @param delegator_address 	delegator's account address
     * @param validator_address 	validator's operator address
     * @param shares 	delegator's shares
     * @param balance balance of the delegation
     */
    constructor(delegator_address, validator_address, shares, balance) {
        super();
        this.delegator_address = delegator_address;
        this.validator_address = validator_address;
        this.shares = new Coins_1.Coins(shares);
        this.balance = new Coins_1.Coins(balance);
    }
    static fromAmino(data) {
        const { delegation: { delegator_address, validator_address, shares }, balance, } = data;
        return new Delegation(delegator_address, validator_address, Coins_1.Coins.fromAmino(shares), Coins_1.Coins.fromAmino(balance));
    }
    toAmino() {
        const { delegator_address, validator_address, shares, balance } = this;
        return {
            delegation: {
                delegator_address,
                validator_address,
                shares: shares.toAmino(),
            },
            balance: balance.toAmino(),
        };
    }
    static fromData(data) {
        const { delegation: { delegator_address, validator_address, shares }, balance, } = data;
        return new Delegation(delegator_address, validator_address, Coins_1.Coins.fromData(shares), Coins_1.Coins.fromData(balance));
    }
    toData() {
        const { delegator_address, validator_address, shares, balance } = this;
        return {
            delegation: {
                delegator_address,
                validator_address,
                shares: shares.toData(),
            },
            balance: balance.toData(),
        };
    }
    static fromProto(proto) {
        const delegationProto = proto.delegation;
        return new Delegation(delegationProto.delegatorAddress, delegationProto.validatorAddress, Coins_1.Coins.fromProto(delegationProto.shares), Coins_1.Coins.fromProto(proto.balance));
    }
    toProto() {
        const { delegator_address, validator_address, shares, balance } = this;
        return staking_1.DelegationResponse.fromPartial({
            delegation: staking_1.Delegation.fromPartial({
                delegatorAddress: delegator_address,
                validatorAddress: validator_address,
                shares: shares.toProto(),
            }),
            balance: balance.toProto(),
        });
    }
}
exports.Delegation = Delegation;
