// @initia/initia.js/dist/core/mstaking/UnbondingDelegation.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnbondingDelegation = void 0;
const json_1 = require("../../util/json");
const Coins_1 = require("../Coins");
const staking_1 = require("@initia/initia.proto/initia/mstaking/v1/staking");
const long_1 = __importDefault(require("long"));
/**
 * When a delegator decides to take out their funds from the mstaking pool, they must
 * unbond their tokens which takes an amount of time specified by `unbonding_time`
 * parameter in the mstaking module.
 *
 * An unbonding delegation is implemented through creating [[UnbondingDelegation.Entry]]
 * objects, limited by the max_entry parameter in the mstaking module params. You cannot
 * initiate unbonds more times than the amount of entries permitted. Entries are cleared
 * when their unbonding periods are completed and the funds are returned to the
 * delegator's account balance to be spent freely.
 */
class UnbondingDelegation extends json_1.JSONSerializable {
    constructor(delegator_address, validator_address, entries) {
        super();
        this.delegator_address = delegator_address;
        this.validator_address = validator_address;
        this.entries = entries;
    }
    static fromAmino(data) {
        const { delegator_address, validator_address, entries } = data;
        return new UnbondingDelegation(delegator_address, validator_address, entries.map(e => UnbondingDelegation.Entry.fromAmino(e)));
    }
    toAmino() {
        const { delegator_address, validator_address, entries } = this;
        return {
            delegator_address,
            validator_address,
            entries: entries.map(e => e.toAmino()),
        };
    }
    static fromData(data) {
        const { delegator_address, validator_address, entries } = data;
        return new UnbondingDelegation(delegator_address, validator_address, entries.map(e => UnbondingDelegation.Entry.fromData(e)));
    }
    toData() {
        const { delegator_address, validator_address, entries } = this;
        return {
            delegator_address,
            validator_address,
            entries: entries.map(e => e.toData()),
        };
    }
    toProto() {
        const { delegator_address, validator_address, entries } = this;
        return staking_1.UnbondingDelegation.fromPartial({
            delegatorAddress: delegator_address,
            entries: entries.map(e => e.toProto()),
            validatorAddress: validator_address,
        });
    }
    static fromProto(proto) {
        return new UnbondingDelegation(proto.delegatorAddress, proto.validatorAddress, proto.entries.map(e => UnbondingDelegation.Entry.fromProto(e)));
    }
}
exports.UnbondingDelegation = UnbondingDelegation;
(function (UnbondingDelegation) {
    class Entry extends json_1.JSONSerializable {
        /**
         * Note that the size of the undelegation is `initial_balance - balance`
         * @param initial_balance balance of delegation prior to initiating unbond
         * @param balance balance of delegation after initiating unbond
         * @param creation_height height of blockchain when entry was created
         * @param completion_time time when unbonding will be completed
         */
        constructor(initial_balance, balance, creation_height, completion_time) {
            super();
            this.creation_height = creation_height;
            this.completion_time = completion_time;
            this.initial_balance = new Coins_1.Coins(initial_balance);
            this.balance = new Coins_1.Coins(balance);
        }
        toAmino() {
            return {
                initial_balance: this.initial_balance.toAmino(),
                balance: this.balance.toAmino(),
                creation_height: this.creation_height.toFixed(),
                completion_time: this.completion_time.toISOString(),
            };
        }
        static fromAmino(data) {
            const { initial_balance, balance, creation_height, completion_time } = data;
            return new Entry(Coins_1.Coins.fromAmino(initial_balance), Coins_1.Coins.fromAmino(balance), Number.parseInt(creation_height), new Date(completion_time));
        }
        toData() {
            return {
                initial_balance: this.initial_balance.toData(),
                balance: this.balance.toData(),
                creation_height: this.creation_height.toFixed(),
                completion_time: this.completion_time.toISOString(),
            };
        }
        static fromData(data) {
            const { initial_balance, balance, creation_height, completion_time } = data;
            return new Entry(Coins_1.Coins.fromData(initial_balance), Coins_1.Coins.fromData(balance), Number.parseInt(creation_height), new Date(completion_time));
        }
        toProto() {
            const { initial_balance, balance, creation_height, completion_time } = this;
            return staking_1.UnbondingDelegationEntry.fromPartial({
                balance: balance.toProto(),
                completionTime: completion_time,
                creationHeight: long_1.default.fromNumber(creation_height),
                initialBalance: initial_balance.toProto(),
            });
        }
        static fromProto(proto) {
            return new Entry(Coins_1.Coins.fromProto(proto.initialBalance), Coins_1.Coins.fromProto(proto.balance), proto.creationHeight.toNumber(), proto.completionTime);
        }
    }
    UnbondingDelegation.Entry = Entry;
})(UnbondingDelegation || (exports.UnbondingDelegation = UnbondingDelegation = {}));
