// @initia/initia.js/dist/core/oracle/msgs/MsgAddCurrencyPairs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgAddCurrencyPairs = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/slinky/oracle/v1/tx");
const CurrencyPair_1 = require("../CurrencyPair");
class MsgAddCurrencyPairs extends json_1.JSONSerializable {
    /**
     * @param authority
     * @param currency_pairs set of CurrencyPairs to be added to the module
     */
    constructor(authority, currency_pairs) {
        super();
        this.authority = authority;
        this.currency_pairs = currency_pairs;
    }
    static fromAmino(data) {
        const { value: { authority, currency_pairs }, } = data;
        return new MsgAddCurrencyPairs(authority, currency_pairs.map(CurrencyPair_1.CurrencyPair.fromAmino));
    }
    toAmino() {
        const { authority, currency_pairs } = this;
        return {
            type: 'slinky/x/oracle/MsgAddCurrencyPairs',
            value: {
                authority,
                currency_pairs: currency_pairs.map(msg => msg.toAmino()),
            },
        };
    }
    static fromData(data) {
        const { authority, currency_pairs } = data;
        return new MsgAddCurrencyPairs(authority, currency_pairs.map(CurrencyPair_1.CurrencyPair.fromData));
    }
    toData() {
        const { authority, currency_pairs } = this;
        return {
            '@type': '/slinky.oracle.v1.MsgAddCurrencyPairs',
            authority,
            currency_pairs: currency_pairs.map(msg => msg.toData()),
        };
    }
    static fromProto(data) {
        return new MsgAddCurrencyPairs(data.authority, data.currencyPairs.map(CurrencyPair_1.CurrencyPair.fromProto));
    }
    toProto() {
        const { authority, currency_pairs } = this;
        return tx_1.MsgAddCurrencyPairs.fromPartial({
            authority,
            currencyPairs: currency_pairs.map(c => c.toProto()),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/slinky.oracle.v1.MsgAddCurrencyPairs',
            value: tx_1.MsgAddCurrencyPairs.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgAddCurrencyPairs.fromProto(tx_1.MsgAddCurrencyPairs.decode(msgAny.value));
    }
}
exports.MsgAddCurrencyPairs = MsgAddCurrencyPairs;
