// @initia/initia.js/dist/core/oracle/msgs/MsgRemoveCurrencyPairs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgRemoveCurrencyPairs = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/slinky/oracle/v1/tx");
class MsgRemoveCurrencyPairs extends json_1.JSONSerializable {
    /**
     * @param authority
     * @param currency_pair_ids the stringified representation of a currency-pairs (base/quote) to be removed from the module's state
     */
    constructor(authority, currency_pair_ids) {
        super();
        this.authority = authority;
        this.currency_pair_ids = currency_pair_ids;
    }
    static fromAmino(data) {
        const { value: { authority, currency_pair_ids }, } = data;
        return new MsgRemoveCurrencyPairs(authority, currency_pair_ids);
    }
    toAmino() {
        const { authority, currency_pair_ids } = this;
        return {
            type: 'slinky/x/oracle/MsgSetCurrencyPairs',
            value: {
                authority,
                currency_pair_ids,
            },
        };
    }
    static fromData(data) {
        const { authority, currency_pair_ids } = data;
        return new MsgRemoveCurrencyPairs(authority, currency_pair_ids);
    }
    toData() {
        const { authority, currency_pair_ids } = this;
        return {
            '@type': '/slinky.oracle.v1.MsgRemoveCurrencyPairs',
            authority,
            currency_pair_ids,
        };
    }
    static fromProto(data) {
        return new MsgRemoveCurrencyPairs(data.authority, data.currencyPairIds);
    }
    toProto() {
        const { authority, currency_pair_ids } = this;
        return tx_1.MsgRemoveCurrencyPairs.fromPartial({
            authority,
            currencyPairIds: currency_pair_ids,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/slinky.oracle.v1.MsgRemoveCurrencyPairs',
            value: tx_1.MsgRemoveCurrencyPairs.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgRemoveCurrencyPairs.fromProto(tx_1.MsgRemoveCurrencyPairs.decode(msgAny.value));
    }
}
exports.MsgRemoveCurrencyPairs = MsgRemoveCurrencyPairs;
