// @initia/initia.proto/noble/forwarding/v1/genesis.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenesisState_TotalForwardedEntry = exports.GenesisState_NumOfForwardsEntry = exports.GenesisState_NumOfAccountsEntry = exports.GenesisState = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "noble.forwarding.v1";
function createBaseGenesisState() {
    return { numOfAccounts: new Map(), numOfForwards: new Map(), totalForwarded: new Map() };
}
exports.GenesisState = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        message.numOfAccounts.forEach((value, key) => {
            exports.GenesisState_NumOfAccountsEntry.encode({ key: key, value }, writer.uint32(10).fork()).ldelim();
        });
        message.numOfForwards.forEach((value, key) => {
            exports.GenesisState_NumOfForwardsEntry.encode({ key: key, value }, writer.uint32(18).fork()).ldelim();
        });
        message.totalForwarded.forEach((value, key) => {
            exports.GenesisState_TotalForwardedEntry.encode({ key: key, value }, writer.uint32(26).fork()).ldelim();
        });
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    const entry1 = exports.GenesisState_NumOfAccountsEntry.decode(reader, reader.uint32());
                    if (entry1.value !== undefined) {
                        message.numOfAccounts.set(entry1.key, entry1.value);
                    }
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    const entry2 = exports.GenesisState_NumOfForwardsEntry.decode(reader, reader.uint32());
                    if (entry2.value !== undefined) {
                        message.numOfForwards.set(entry2.key, entry2.value);
                    }
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    const entry3 = exports.GenesisState_TotalForwardedEntry.decode(reader, reader.uint32());
                    if (entry3.value !== undefined) {
                        message.totalForwarded.set(entry3.key, entry3.value);
                    }
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            numOfAccounts: isObject(object.numOfAccounts)
                ? Object.entries(object.numOfAccounts).reduce((acc, [key, value]) => {
                    acc.set(key, long_1.default.fromValue(value));
                    return acc;
                }, new Map())
                : new Map(),
            numOfForwards: isObject(object.numOfForwards)
                ? Object.entries(object.numOfForwards).reduce((acc, [key, value]) => {
                    acc.set(key, long_1.default.fromValue(value));
                    return acc;
                }, new Map())
                : new Map(),
            totalForwarded: isObject(object.totalForwarded)
                ? Object.entries(object.totalForwarded).reduce((acc, [key, value]) => {
                    acc.set(key, String(value));
                    return acc;
                }, new Map())
                : new Map(),
        };
    },
    toJSON(message) {
        var _a, _b, _c;
        const obj = {};
        if ((_a = message.numOfAccounts) === null || _a === void 0 ? void 0 : _a.size) {
            obj.numOfAccounts = {};
            message.numOfAccounts.forEach((v, k) => {
                obj.numOfAccounts[k] = v.toString();
            });
        }
        if ((_b = message.numOfForwards) === null || _b === void 0 ? void 0 : _b.size) {
            obj.numOfForwards = {};
            message.numOfForwards.forEach((v, k) => {
                obj.numOfForwards[k] = v.toString();
            });
        }
        if ((_c = message.totalForwarded) === null || _c === void 0 ? void 0 : _c.size) {
            obj.totalForwarded = {};
            message.totalForwarded.forEach((v, k) => {
                obj.totalForwarded[k] = v;
            });
        }
        return obj;
    },
    create(base) {
        return exports.GenesisState.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseGenesisState();
        message.numOfAccounts = (() => {
            var _a;
            const m = new Map();
            ((_a = object.numOfAccounts) !== null && _a !== void 0 ? _a : new Map()).forEach((value, key) => {
                if (value !== undefined) {
                    m.set(key, long_1.default.fromValue(value));
                }
            });
            return m;
        })();
        message.numOfForwards = (() => {
            var _a;
            const m = new Map();
            ((_a = object.numOfForwards) !== null && _a !== void 0 ? _a : new Map()).forEach((value, key) => {
                if (value !== undefined) {
                    m.set(key, long_1.default.fromValue(value));
                }
            });
            return m;
        })();
        message.totalForwarded = (() => {
            var _a;
            const m = new Map();
            ((_a = object.totalForwarded) !== null && _a !== void 0 ? _a : new Map()).forEach((value, key) => {
                if (value !== undefined) {
                    m.set(key, globalThis.String(value));
                }
            });
            return m;
        })();
        return message;
    },
};
function createBaseGenesisState_NumOfAccountsEntry() {
    return { key: "", value: long_1.default.UZERO };
}
exports.GenesisState_NumOfAccountsEntry = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (!message.value.isZero()) {
            writer.uint32(16).uint64(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState_NumOfAccountsEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.value = reader.uint64();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            value: isSet(object.value) ? long_1.default.fromValue(object.value) : long_1.default.UZERO,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (!message.value.isZero()) {
            obj.value = (message.value || long_1.default.UZERO).toString();
        }
        return obj;
    },
    create(base) {
        return exports.GenesisState_NumOfAccountsEntry.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGenesisState_NumOfAccountsEntry();
        message.key = (_a = object.key) !== null && _a !== void 0 ? _a : "";
        message.value =
            object.value !== undefined && object.value !== null ? long_1.default.fromValue(object.value) : long_1.default.UZERO;
        return message;
    },
};
function createBaseGenesisState_NumOfForwardsEntry() {
    return { key: "", value: long_1.default.UZERO };
}
exports.GenesisState_NumOfForwardsEntry = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (!message.value.isZero()) {
            writer.uint32(16).uint64(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState_NumOfForwardsEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.value = reader.uint64();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            value: isSet(object.value) ? long_1.default.fromValue(object.value) : long_1.default.UZERO,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (!message.value.isZero()) {
            obj.value = (message.value || long_1.default.UZERO).toString();
        }
        return obj;
    },
    create(base) {
        return exports.GenesisState_NumOfForwardsEntry.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGenesisState_NumOfForwardsEntry();
        message.key = (_a = object.key) !== null && _a !== void 0 ? _a : "";
        message.value =
            object.value !== undefined && object.value !== null ? long_1.default.fromValue(object.value) : long_1.default.UZERO;
        return message;
    },
};
function createBaseGenesisState_TotalForwardedEntry() {
    return { key: "", value: "" };
}
exports.GenesisState_TotalForwardedEntry = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== "") {
            writer.uint32(18).string(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState_TotalForwardedEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.value = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            value: isSet(object.value) ? globalThis.String(object.value) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.value !== "") {
            obj.value = message.value;
        }
        return obj;
    },
    create(base) {
        return exports.GenesisState_TotalForwardedEntry.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseGenesisState_TotalForwardedEntry();
        message.key = (_a = object.key) !== null && _a !== void 0 ? _a : "";
        message.value = (_b = object.value) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isObject(value) {
    return typeof value === "object" && value !== null;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
