// @initia/initia.proto/celestia/blob/v1/params.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "celestia.blob.v1";
function createBaseParams() {
    return { gasPerBlobByte: 0, govMaxSquareSize: long_1.default.UZERO };
}
exports.Params = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.gasPerBlobByte !== 0) {
            writer.uint32(8).uint32(message.gasPerBlobByte);
        }
        if (!message.govMaxSquareSize.isZero()) {
            writer.uint32(16).uint64(message.govMaxSquareSize);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.gasPerBlobByte = reader.uint32();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.govMaxSquareSize = reader.uint64();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            gasPerBlobByte: isSet(object.gasPerBlobByte) ? globalThis.Number(object.gasPerBlobByte) : 0,
            govMaxSquareSize: isSet(object.govMaxSquareSize) ? long_1.default.fromValue(object.govMaxSquareSize) : long_1.default.UZERO,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.gasPerBlobByte !== 0) {
            obj.gasPerBlobByte = Math.round(message.gasPerBlobByte);
        }
        if (!message.govMaxSquareSize.isZero()) {
            obj.govMaxSquareSize = (message.govMaxSquareSize || long_1.default.UZERO).toString();
        }
        return obj;
    },
    create(base) {
        return exports.Params.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseParams();
        message.gasPerBlobByte = (_a = object.gasPerBlobByte) !== null && _a !== void 0 ? _a : 0;
        message.govMaxSquareSize =
            object.govMaxSquareSize !== undefined && object.govMaxSquareSize !== null
                ? long_1.default.fromValue(object.govMaxSquareSize)
                : long_1.default.UZERO;
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
