// @initia/initia.proto/celestia/blob/v1/tx.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcWebError = exports.GrpcWebImpl = exports.MsgPayForBlobsDesc = exports.MsgDesc = exports.MsgClientImpl = exports.MsgPayForBlobsResponse = exports.MsgPayForBlobs = exports.protobufPackage = void 0;
/* eslint-disable */
const grpc_web_1 = require("@improbable-eng/grpc-web");
const browser_headers_1 = require("browser-headers");
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "celestia.blob.v1";
function createBaseMsgPayForBlobs() {
    return { signer: "", namespaces: [], blobSizes: [], shareCommitments: [], shareVersions: [] };
}
exports.MsgPayForBlobs = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.signer !== "") {
            writer.uint32(10).string(message.signer);
        }
        for (const v of message.namespaces) {
            writer.uint32(18).bytes(v);
        }
        writer.uint32(26).fork();
        for (const v of message.blobSizes) {
            writer.uint32(v);
        }
        writer.ldelim();
        for (const v of message.shareCommitments) {
            writer.uint32(34).bytes(v);
        }
        writer.uint32(66).fork();
        for (const v of message.shareVersions) {
            writer.uint32(v);
        }
        writer.ldelim();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMsgPayForBlobs();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.signer = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.namespaces.push(reader.bytes());
                    continue;
                case 3:
                    if (tag === 24) {
                        message.blobSizes.push(reader.uint32());
                        continue;
                    }
                    if (tag === 26) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.blobSizes.push(reader.uint32());
                        }
                        continue;
                    }
                    break;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.shareCommitments.push(reader.bytes());
                    continue;
                case 8:
                    if (tag === 64) {
                        message.shareVersions.push(reader.uint32());
                        continue;
                    }
                    if (tag === 66) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.shareVersions.push(reader.uint32());
                        }
                        continue;
                    }
                    break;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            signer: isSet(object.signer) ? globalThis.String(object.signer) : "",
            namespaces: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.namespaces)
                ? object.namespaces.map((e) => bytesFromBase64(e))
                : [],
            blobSizes: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.blobSizes)
                ? object.blobSizes.map((e) => globalThis.Number(e))
                : [],
            shareCommitments: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.shareCommitments)
                ? object.shareCommitments.map((e) => bytesFromBase64(e))
                : [],
            shareVersions: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.shareVersions)
                ? object.shareVersions.map((e) => globalThis.Number(e))
                : [],
        };
    },
    toJSON(message) {
        var _a, _b, _c, _d;
        const obj = {};
        if (message.signer !== "") {
            obj.signer = message.signer;
        }
        if ((_a = message.namespaces) === null || _a === void 0 ? void 0 : _a.length) {
            obj.namespaces = message.namespaces.map((e) => base64FromBytes(e));
        }
        if ((_b = message.blobSizes) === null || _b === void 0 ? void 0 : _b.length) {
            obj.blobSizes = message.blobSizes.map((e) => Math.round(e));
        }
        if ((_c = message.shareCommitments) === null || _c === void 0 ? void 0 : _c.length) {
            obj.shareCommitments = message.shareCommitments.map((e) => base64FromBytes(e));
        }
        if ((_d = message.shareVersions) === null || _d === void 0 ? void 0 : _d.length) {
            obj.shareVersions = message.shareVersions.map((e) => Math.round(e));
        }
        return obj;
    },
    create(base) {
        return exports.MsgPayForBlobs.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e;
        const message = createBaseMsgPayForBlobs();
        message.signer = (_a = object.signer) !== null && _a !== void 0 ? _a : "";
        message.namespaces = ((_b = object.namespaces) === null || _b === void 0 ? void 0 : _b.map((e) => e)) || [];
        message.blobSizes = ((_c = object.blobSizes) === null || _c === void 0 ? void 0 : _c.map((e) => e)) || [];
        message.shareCommitments = ((_d = object.shareCommitments) === null || _d === void 0 ? void 0 : _d.map((e) => e)) || [];
        message.shareVersions = ((_e = object.shareVersions) === null || _e === void 0 ? void 0 : _e.map((e) => e)) || [];
        return message;
    },
};
function createBaseMsgPayForBlobsResponse() {
    return {};
}
exports.MsgPayForBlobsResponse = {
    encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMsgPayForBlobsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.MsgPayForBlobsResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseMsgPayForBlobsResponse();
        return message;
    },
};
class MsgClientImpl {
    constructor(rpc) {
        this.rpc = rpc;
        this.PayForBlobs = this.PayForBlobs.bind(this);
    }
    PayForBlobs(request, metadata) {
        return this.rpc.unary(exports.MsgPayForBlobsDesc, exports.MsgPayForBlobs.fromPartial(request), metadata);
    }
}
exports.MsgClientImpl = MsgClientImpl;
exports.MsgDesc = { serviceName: "celestia.blob.v1.Msg" };
exports.MsgPayForBlobsDesc = {
    methodName: "PayForBlobs",
    service: exports.MsgDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.MsgPayForBlobs.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.MsgPayForBlobsResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
class GrpcWebImpl {
    constructor(host, options) {
        this.host = host;
        this.options = options;
    }
    unary(methodDesc, _request, metadata) {
        var _a;
        const request = Object.assign(Object.assign({}, _request), methodDesc.requestType);
        const maybeCombinedMetadata = metadata && this.options.metadata
            ? new browser_headers_1.BrowserHeaders(Object.assign(Object.assign({}, (_a = this.options) === null || _a === void 0 ? void 0 : _a.metadata.headersMap), metadata === null || metadata === void 0 ? void 0 : metadata.headersMap))
            : metadata !== null && metadata !== void 0 ? metadata : this.options.metadata;
        return new Promise((resolve, reject) => {
            var _a;
            grpc_web_1.grpc.unary(methodDesc, Object.assign(Object.assign({ request, host: this.host, metadata: maybeCombinedMetadata !== null && maybeCombinedMetadata !== void 0 ? maybeCombinedMetadata : {} }, (this.options.transport !== undefined ? { transport: this.options.transport } : {})), { debug: (_a = this.options.debug) !== null && _a !== void 0 ? _a : false, onEnd: function (response) {
                    if (response.status === grpc_web_1.grpc.Code.OK) {
                        resolve(response.message.toObject());
                    }
                    else {
                        const err = new GrpcWebError(response.statusMessage, response.status, response.trailers);
                        reject(err);
                    }
                } }));
        });
    }
}
exports.GrpcWebImpl = GrpcWebImpl;
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(globalThis.String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
class GrpcWebError extends globalThis.Error {
    constructor(message, code, metadata) {
        super(message);
        this.code = code;
        this.metadata = metadata;
    }
}
exports.GrpcWebError = GrpcWebError;
