// @initia/initia.proto/initia/move/v1/authz.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExecuteAuthorization = exports.PublishAuthorization = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const types_1 = require("./types");
exports.protobufPackage = "initia.move.v1";
function createBasePublishAuthorization() {
    return { moduleNames: [] };
}
exports.PublishAuthorization = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.moduleNames) {
            writer.uint32(10).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePublishAuthorization();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.moduleNames.push(reader.string());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            moduleNames: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.moduleNames)
                ? object.moduleNames.map((e) => globalThis.String(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.moduleNames) === null || _a === void 0 ? void 0 : _a.length) {
            obj.moduleNames = message.moduleNames;
        }
        return obj;
    },
    create(base) {
        return exports.PublishAuthorization.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBasePublishAuthorization();
        message.moduleNames = ((_a = object.moduleNames) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        return message;
    },
};
function createBaseExecuteAuthorization() {
    return { items: [] };
}
exports.ExecuteAuthorization = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.items) {
            types_1.ExecuteAuthorizationItem.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseExecuteAuthorization();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.items.push(types_1.ExecuteAuthorizationItem.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            items: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.items)
                ? object.items.map((e) => types_1.ExecuteAuthorizationItem.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.items) === null || _a === void 0 ? void 0 : _a.length) {
            obj.items = message.items.map((e) => types_1.ExecuteAuthorizationItem.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.ExecuteAuthorization.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseExecuteAuthorization();
        message.items = ((_a = object.items) === null || _a === void 0 ? void 0 : _a.map((e) => types_1.ExecuteAuthorizationItem.fromPartial(e))) || [];
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
