// @initia/initia.js/dist/core/ibc/core/client/Height.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Height = void 0;
const client_1 = require("@initia/initia.proto/ibc/core/client/v1/client");
const long_1 = __importDefault(require("long"));
const json_1 = require("../../../../util/json");
/**
 * Height is a monotonically increasing data type
 * that can be compared against another Height for the purposes of updating and
 * freezing clients
 *
 * Normally the RevisionHeight is incremented at each height while keeping
 * RevisionNumber the same. However some consensus algorithms may choose to
 * reset the height in certain conditions e.g. hard forks, state-machine
 * breaking changes In these cases, the RevisionNumber is incremented so that
 * height continues to be monitonically increasing even as the RevisionHeight
 * gets reset
 */
class Height extends json_1.JSONSerializable {
    /**
     * @param revision_number the revision that the client is currently on
     * @param revision_height the height within the given revision
     */
    constructor(revision_number, revision_height) {
        super();
        this.revision_number = revision_number;
        this.revision_height = revision_height;
    }
    static fromAmino(data) {
        const { revision_number, revision_height } = data;
        return new Height(parseInt(revision_number !== null && revision_number !== void 0 ? revision_number : '0'), parseInt(revision_height !== null && revision_height !== void 0 ? revision_height : '0'));
    }
    toAmino() {
        const { revision_number, revision_height } = this;
        const res = {
            revision_number: revision_number > 0 ? revision_number.toFixed() : undefined,
            revision_height: revision_height > 0 ? revision_height.toFixed() : undefined,
        };
        return res;
    }
    static fromData(data) {
        const { revision_number, revision_height } = data;
        return new Height(Number.parseInt(revision_number), Number.parseInt(revision_height));
    }
    toData() {
        const { revision_number, revision_height } = this;
        const res = {
            revision_number: revision_number.toFixed(),
            revision_height: revision_height.toFixed(),
        };
        return res;
    }
    static fromProto(proto) {
        return new Height(proto.revisionNumber.toNumber(), proto.revisionHeight.toNumber());
    }
    toProto() {
        const { revision_number, revision_height } = this;
        return client_1.Height.fromPartial({
            revisionNumber: long_1.default.fromNumber(revision_number),
            revisionHeight: long_1.default.fromNumber(revision_height),
        });
    }
}
exports.Height = Height;
