// @initia/initia.js/dist/core/ibc/core/client/IbcClientParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IbcClientParams = void 0;
const json_1 = require("../../../../util/json");
const client_1 = require("@initia/initia.proto/ibc/core/client/v1/client");
class IbcClientParams extends json_1.JSONSerializable {
    /**
     * @param allowed_clients the list of allowed client state types.
     */
    constructor(allowed_clients) {
        super();
        this.allowed_clients = allowed_clients;
    }
    static fromAmino(data) {
        const { allowed_clients } = data;
        return new IbcClientParams(allowed_clients);
    }
    toAmino() {
        const { allowed_clients } = this;
        const res = {
            allowed_clients: allowed_clients,
        };
        return res;
    }
    static fromData(data) {
        const { allowed_clients } = data;
        return new IbcClientParams(allowed_clients);
    }
    toData() {
        const { allowed_clients } = this;
        const res = {
            allowed_clients,
        };
        return res;
    }
    static fromProto(proto) {
        return new IbcClientParams(proto.allowedClients);
    }
    toProto() {
        const { allowed_clients } = this;
        return client_1.Params.fromPartial({
            allowedClients: allowed_clients,
        });
    }
}
exports.IbcClientParams = IbcClientParams;
