// @initia/initia.js/dist/core/ibc/core/channel/Channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Channel = void 0;
const channel_1 = require("@initia/initia.proto/ibc/core/channel/v1/channel");
const json_1 = require("../../../../util/json");
const ChannelCounterparty_1 = require("./ChannelCounterparty");
/**
 * Channel is a monotonically increasing data type
 * that can be compared against another Channel for the purposes of updating and
 * freezing clients
 *
 * Normally the RevisionChannel is incremented at each height while keeping
 * RevisionNumber the same. However some consensus algorithms may choose to
 * reset the height in certain conditions e.g. hard forks, state-machine
 * breaking changes In these cases, the RevisionNumber is incremented so that
 * height continues to be monitonically increasing even as the RevisionChannel
 * gets reset
 */
class Channel extends json_1.JSONSerializable {
    /**
     * @param state current state of the channel end
     * @param ordering  whether the channel is ordered or unordered
     * @param counterparty counterparty channel end
     * @param connection_hops list of connection identifiers, in order, along which packets sent on this channel will travel
     * @param version opaque channel version, which is agreed upon during the handshake
     */
    constructor(state, ordering, counterparty, connection_hops, version) {
        super();
        this.state = state;
        this.ordering = ordering;
        this.counterparty = counterparty;
        this.connection_hops = connection_hops;
        this.version = version;
    }
    static fromAmino(data) {
        const { state, ordering, counterparty, connection_hops, version } = data;
        return new Channel(state, ordering, counterparty ? ChannelCounterparty_1.ChannelCounterparty.fromAmino(counterparty) : undefined, connection_hops, version);
    }
    toAmino() {
        const { state, ordering, counterparty, connection_hops, version } = this;
        const res = {
            state,
            ordering,
            counterparty: counterparty === null || counterparty === void 0 ? void 0 : counterparty.toAmino(),
            connection_hops,
            version,
        };
        return res;
    }
    static fromData(data) {
        const { state, ordering, counterparty, connection_hops, version } = data;
        return new Channel(state, ordering, counterparty ? ChannelCounterparty_1.ChannelCounterparty.fromData(counterparty) : undefined, connection_hops, version);
    }
    toData() {
        const { state, ordering, counterparty, connection_hops, version } = this;
        const res = {
            state,
            ordering,
            counterparty: counterparty === null || counterparty === void 0 ? void 0 : counterparty.toData(),
            connection_hops,
            version,
        };
        return res;
    }
    static fromProto(proto) {
        return new Channel(proto.state, proto.ordering, proto.counterparty
            ? ChannelCounterparty_1.ChannelCounterparty.fromProto(proto.counterparty)
            : undefined, proto.connectionHops, proto.version);
    }
    toProto() {
        const { state, ordering, counterparty, connection_hops, version } = this;
        return channel_1.Channel.fromPartial({
            state,
            ordering,
            counterparty: counterparty === null || counterparty === void 0 ? void 0 : counterparty.toProto(),
            connectionHops: connection_hops,
            version,
        });
    }
}
exports.Channel = Channel;
