// @initia/initia.js/dist/core/ibc/core/channel/ChannelCounterparty.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelCounterparty = void 0;
const json_1 = require("../../../../util/json");
const channel_1 = require("@initia/initia.proto/ibc/core/channel/v1/channel");
/** ChannelCounterparty defines a channel end ChannelCounterparty */
class ChannelCounterparty extends json_1.JSONSerializable {
    /**
     * @param port_id port on the ChannelCounterparty chain which owns the other end of the channel.
     * @param channel_id channel end on the ChannelCounterparty chain
     */
    constructor(port_id, channel_id) {
        super();
        this.port_id = port_id;
        this.channel_id = channel_id;
    }
    static fromAmino(data) {
        const { port_id, channel_id } = data;
        return new ChannelCounterparty(port_id, channel_id);
    }
    toAmino() {
        const { port_id, channel_id } = this;
        const res = {
            port_id,
            channel_id,
        };
        return res;
    }
    static fromData(data) {
        const { port_id, channel_id } = data;
        return new ChannelCounterparty(port_id, channel_id);
    }
    toData() {
        const { port_id, channel_id } = this;
        const res = {
            port_id,
            channel_id,
        };
        return res;
    }
    static fromProto(proto) {
        return new ChannelCounterparty(proto.portId, proto.channelId);
    }
    toProto() {
        const { port_id, channel_id } = this;
        return channel_1.Counterparty.fromPartial({
            portId: port_id,
            channelId: channel_id,
        });
    }
}
exports.ChannelCounterparty = ChannelCounterparty;
