// @initia/initia.js/dist/core/ibc/core/channel/IbcChannelParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IbcChannelParams = void 0;
const json_1 = require("../../../../util/json");
const channel_1 = require("@initia/initia.proto/ibc/core/channel/v1/channel");
const Timeout_1 = require("./Timeout");
class IbcChannelParams extends json_1.JSONSerializable {
    /**
     * @param upgrade_timeout the relative timeout after which channel upgrades will time out
     */
    constructor(upgrade_timeout) {
        super();
        this.upgrade_timeout = upgrade_timeout;
    }
    static fromAmino(data) {
        return new IbcChannelParams(Timeout_1.Timeout.fromAmino(data.upgrade_timeout));
    }
    toAmino() {
        return {
            upgrade_timeout: this.upgrade_timeout.toAmino(),
        };
    }
    static fromData(data) {
        return new IbcChannelParams(Timeout_1.Timeout.fromData(data.upgrade_timeout));
    }
    toData() {
        return {
            upgrade_timeout: this.upgrade_timeout.toData(),
        };
    }
    static fromProto(data) {
        return new IbcChannelParams(Timeout_1.Timeout.fromProto(data.upgradeTimeout));
    }
    toProto() {
        return channel_1.Params.fromPartial({
            upgradeTimeout: this.upgrade_timeout.toProto(),
        });
    }
}
exports.IbcChannelParams = IbcChannelParams;
