// @initia/initia.js/dist/core/ibc/core/channel/Timeout.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Timeout = void 0;
const json_1 = require("../../../../util/json");
const channel_1 = require("@initia/initia.proto/ibc/core/channel/v1/channel");
const client_1 = require("../client");
const long_1 = __importDefault(require("long"));
class Timeout extends json_1.JSONSerializable {
    /**
     * @param height block height after which the packet or upgrade times out
     * @param timestamp block timestamp (in nanoseconds) after which the packet or upgrade times out
     */
    constructor(height, timestamp) {
        super();
        this.height = height;
        this.timestamp = timestamp;
    }
    static fromAmino(data) {
        const { height, timestamp } = data;
        return new Timeout(client_1.Height.fromAmino(height), Number.parseInt(timestamp));
    }
    toAmino() {
        const { height, timestamp } = this;
        return {
            height: height.toAmino(),
            timestamp: timestamp.toString(),
        };
    }
    static fromData(data) {
        const { height, timestamp } = data;
        return new Timeout(client_1.Height.fromData(height), Number.parseInt(timestamp));
    }
    toData() {
        const { height, timestamp } = this;
        return {
            height: height.toData(),
            timestamp: timestamp.toString(),
        };
    }
    static fromProto(proto) {
        return new Timeout(client_1.Height.fromProto(proto.height), proto.timestamp.toNumber());
    }
    toProto() {
        const { height, timestamp } = this;
        return channel_1.Timeout.fromPartial({
            height: height.toProto(),
            timestamp: long_1.default.fromNumber(timestamp),
        });
    }
}
exports.Timeout = Timeout;
