// @initia/initia.js/dist/core/ophost/BatchInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchInfo = void 0;
const json_1 = require("../../util/json");
const types_1 = require("@initia/opinit.proto/opinit/ophost/v1/types");
class BatchInfo extends json_1.JSONSerializable {
    /**
     * @param submitter the address of the batch submitter
     * @param chain the target chain
     */
    constructor(submitter, chain) {
        super();
        this.submitter = submitter;
        this.chain = chain;
    }
    static fromAmino(data) {
        const { submitter, chain } = data;
        return new BatchInfo(submitter, chain);
    }
    toAmino() {
        const { submitter, chain } = this;
        return { submitter, chain };
    }
    static fromData(data) {
        const { submitter, chain } = data;
        return new BatchInfo(submitter, chain);
    }
    toData() {
        const { submitter, chain } = this;
        return { submitter, chain };
    }
    static fromProto(data) {
        return new BatchInfo(data.submitter, data.chain);
    }
    toProto() {
        const { submitter, chain } = this;
        return types_1.BatchInfo.fromPartial({
            submitter,
            chain,
        });
    }
}
exports.BatchInfo = BatchInfo;
