// @initia/initia.js/dist/core/ophost/OphostParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OphostParams = void 0;
const json_1 = require("../../util/json");
const Coins_1 = require("../Coins");
const types_1 = require("@initia/opinit.proto/opinit/ophost/v1/types");
class OphostParams extends json_1.JSONSerializable {
    /**
     * @param registration_fee the amount to be paid by l2 creator
     */
    constructor(registration_fee) {
        super();
        this.registration_fee = new Coins_1.Coins(registration_fee);
    }
    static fromAmino(data) {
        return new OphostParams(Coins_1.Coins.fromAmino(data.value.registration_fee));
    }
    toAmino() {
        return {
            type: 'ophost/Params',
            value: {
                registration_fee: this.registration_fee.toAmino(),
            },
        };
    }
    static fromData(data) {
        return new OphostParams(Coins_1.Coins.fromData(data.registration_fee));
    }
    toData() {
        return {
            '@type': '/opinit.ophost.v1.Params',
            registration_fee: this.registration_fee.toData(),
        };
    }
    static fromProto(data) {
        return new OphostParams(Coins_1.Coins.fromProto(data.registrationFee));
    }
    toProto() {
        return types_1.Params.fromPartial({
            registrationFee: this.registration_fee.toProto(),
        });
    }
}
exports.OphostParams = OphostParams;
