// @initia/initia.js/dist/client/lcd/api/DistributionAPI.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DistributionAPI = void 0;
const BaseAPI_1 = require("./BaseAPI");
const core_1 = require("../../../core");
class DistributionAPI extends BaseAPI_1.BaseAPI {
    /**
     * Gets a delegator's rewards.
     * @param delegator delegator's account address
     */
    rewards(delegator, params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const rewardsData = yield this.c.get(`/initia/distribution/v1/delegators/${delegator}/rewards`, params);
            const rewards = {};
            for (const reward of rewardsData.rewards) {
                rewards[reward.validator_address] = reward.reward.map(pool => ({
                    denom: pool.denom,
                    coins: core_1.Coins.fromData(pool.dec_coins),
                }));
            }
            return {
                rewards,
                total: rewardsData.total.map(pool => ({
                    denom: pool.denom,
                    coins: core_1.Coins.fromData(pool.dec_coins),
                })),
            };
        });
    }
    /**
     * Gets a delegator's rewards by validator.
     * @param delegator delegator's account address
     * @param validator validator's account address
     */
    rewardsByValidator(delegator, validator, params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/initia/distribution/v1/delegators/${delegator}/rewards/${validator}`, params)
                .then(d => d.rewards.map(pool => ({
                denom: pool.denom,
                coins: core_1.Coins.fromData(pool.dec_coins),
            })));
        });
    }
    /**
     * Gets a validator's rewards.
     * @param validator validator's account address
     */
    validatorRewards(validator, params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/initia/distribution/v1/validators/${validator}/outstanding_rewards`, params)
                .then(d => d.rewards.rewards.map(pool => ({
                denom: pool.denom,
                coins: core_1.Coins.fromData(pool.dec_coins),
            })));
        });
    }
    /**
     * Gets a validator's commissions.
     * @param validator validator's operator address
     */
    validatorCommission(validator, params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/initia/distribution/v1/validators/${validator}/commission`, params)
                .then(d => d.commission.commissions.map(pool => ({
                denom: pool.denom,
                coins: core_1.Coins.fromData(pool.dec_coins),
            })));
        });
    }
    /**
     * Gets the withdraw address of a delegator, the address to which rewards are withdrawn.
     * @param delegator
     */
    withdrawAddress(delegator, params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/cosmos/distribution/v1beta1/delegators/${delegator}/withdraw_address`, params)
                .then(d => d.withdraw_address);
        });
    }
    /**
     * Gets the current value of the community pool.
     */
    communityPool(params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/cosmos/distribution/v1beta1/community_pool`, params)
                .then(d => core_1.Coins.fromData(d.pool));
        });
    }
    /**
     * Gets the current distribution parameters.
     */
    parameters(params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/initia/distribution/v1/params`, params)
                .then(({ params: d }) => core_1.DistributionParams.fromData(d));
        });
    }
}
exports.DistributionAPI = DistributionAPI;
