// @initia/initia.js/dist/client/lcd/api/ForwardingAPI.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForwardingAPI = void 0;
const core_1 = require("../../../core");
const BaseAPI_1 = require("./BaseAPI");
class ForwardingAPI extends BaseAPI_1.BaseAPI {
    address(channel, recipient, params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/noble/forwarding/v1/address/${channel}/${recipient}`, params)
                .then(d => d.address);
        });
    }
    stats(channel, params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/noble/forwarding/v1/stats/${channel}`, params)
                .then(d => ({
                num_of_accounts: parseInt(d.num_of_accounts),
                num_of_forwards: parseInt(d.num_of_forwards),
                total_forwarded: core_1.Coin.fromData(d.total_forwarded),
            }));
        });
    }
}
exports.ForwardingAPI = ForwardingAPI;
