// @initia/initia.js/dist/client/lcd/api/IbcAPI.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IbcAPI = void 0;
const BaseAPI_1 = require("./BaseAPI");
const core_1 = require("../../../core");
class IbcAPI extends BaseAPI_1.BaseAPI {
    /**
     * query all the IBC channels of a chain
     */
    channels(params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/ibc/core/channel/v1/channels`, params)
                .then(d => [d.channels.map(core_1.Channel.fromData), d.pagination]);
        });
    }
    /**
     * query the information of the port at given channel
     * @param channel_id channel identifier
     * @param port_id port name
     */
    port(channel_id, port_id, params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/ibc/core/channel/v1/channels/${channel_id}/ports/${port_id}`, params)
                .then(d => {
                return {
                    channel: core_1.Channel.fromData(d.channel),
                    proof: d.proof,
                    proof_height: core_1.Height.fromData(d.proof_height),
                };
            });
        });
    }
    /**
     *  query all the IBC connections of a chain
     */
    connections(params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/ibc/core/connection/v1/connections`, params)
                .then(d => [
                d.connections.map(core_1.IdentifiedConnection.fromData),
                d.pagination,
            ]);
        });
    }
    /**
     * query an IBC connection end
     * @param connection_id connection unique identifier
     */
    connection(connection_id, params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/ibc/core/connection/v1/connections/${connection_id}`, params)
                .then(d => core_1.IdentifiedConnection.fromData(d.connection));
        });
    }
    /**
     * query all the channels associated with a connection end
     * @param connection_id connection unique identifier
     */
    connectionChannels(connection_id, params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/ibc/core/channel/v1/connections/${connection_id}/channels`, params)
                .then(d => [
                d.channels.map(core_1.Channel.fromData),
                core_1.Height.fromData(d.height),
                d.pagination,
            ]);
        });
    }
    /**
     * Gets the current transfer application parameters.
     */
    parameters(params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/ibc/client/v1/params`, params)
                .then(d => core_1.IbcClientParams.fromData(d.params));
        });
    }
    /**
     * query all the IBC light clients of a chain
     */
    clientStates(params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/ibc/core/client/v1/client_states`, params)
                .then(d => [
                d.client_states.map(core_1.IdentifiedClientState.fromData),
                d.pagination,
            ]);
        });
    }
    /**
     * query an IBC light client
     * @param client_id client state unique identifier
     * @returns
     */
    clientState(client_id, params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/ibc/core/client/v1/client_states/${client_id}`, params)
                .then(d => core_1.IdentifiedClientState.fromData(d.client_state));
        });
    }
    /**
     * query the status of an IBC light client
     * @param client_id client state unique identifier
     * @returns
     */
    clientStatus(client_id, params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/ibc/core/client/v1/client_status/${client_id}`, params)
                .then();
        });
    }
    /**
     * query all the consensus state associated with a given client
     * @param client_id client identifier
     * @returns
     */
    consensusStates(client_id, params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/ibc/core/client/v1/consensus_states/${client_id}`, params)
                .then();
        });
    }
    /**
     * query the height of every consensus states associated with a given client
     * @param client_id client identifier
     * @returns
     */
    consensusStateHeights(client_id, params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/ibc/core/client/v1/consensus_states/${client_id}/heights`, params)
                .then(d => [
                d.consensus_state_heights.map(core_1.Height.fromData),
                d.pagination,
            ]);
        });
    }
}
exports.IbcAPI = IbcAPI;
