// @initia/initia.js/dist/client/lcd/api/IbcTransferAPI.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IbcTransferAPI = void 0;
const BaseAPI_1 = require("./BaseAPI");
const core_1 = require("../../../core");
class IbcTransferAPI extends BaseAPI_1.BaseAPI {
    /** Gets a denomTrace for the hash or denom */
    denomTrace(hash) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/ibc/apps/transfer/v1/denom_traces/${hash}`)
                .then(d => core_1.DenomTrace.fromData(d.denom_trace));
        });
    }
    /** Gets a list of denomTraces */
    denomTraces(params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/ibc/apps/transfer/v1/denom_traces`, params)
                .then(d => [d.denom_traces.map(core_1.DenomTrace.fromData), d.pagination]);
        });
    }
    /** Gets a denomination hash information */
    denomHash(trace) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.c
                .get(`/ibc/apps/transfer/v1/denom_hashes/${trace}`)
                .then(d => d.hash);
        });
    }
    /**
     * Gets the current transfer application parameters.
     */
    parameters(params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/ibc/apps/transfer/v1/params`, params)
                .then(({ params: d }) => ({
                send_enabled: d.send_enabled,
                receive_enabled: d.receive_enabled,
            }));
        });
    }
    /** Gets the escrow address for a particular port and channel id */
    escrowAddress(channel_id, port_id, params = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.c
                .get(`/ibc/apps/transfer/v1/channels/${channel_id}/ports/${port_id}/escrow_address`, params)
                .then(d => d.escrow_address);
        });
    }
}
exports.IbcTransferAPI = IbcTransferAPI;
