// @initia/initia.js/dist/core/crisis/msgs/MsgVerifyInvariant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgVerifyInvariant = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmos/crisis/v1beta1/tx");
/**
 * MsgVerifyInvariant represents a message to verify a particular invariance.
 */
class MsgVerifyInvariant extends json_1.JSONSerializable {
    /**
     * @param sender sender's address
     * @param invariantModuleName module name to verify invariant
     * @param invariantRoute route to verify
     */
    constructor(sender, invariantModuleName, invariantRoute) {
        super();
        this.sender = sender;
        this.invariantModuleName = invariantModuleName;
        this.invariantRoute = invariantRoute;
    }
    static fromAmino(data) {
        const { value: { sender, invariantModuleName, invariantRoute }, } = data;
        return new MsgVerifyInvariant(sender, invariantModuleName, invariantRoute);
    }
    toAmino() {
        throw new Error('MsgVerifyInvarant is not allowed to send');
    }
    static fromData(data) {
        const { sender, invariantModuleName, invariantRoute } = data;
        return new MsgVerifyInvariant(sender, invariantModuleName, invariantRoute);
    }
    toData() {
        const { sender, invariantModuleName, invariantRoute } = this;
        return {
            '@type': '/cosmos.crisis.v1beta1.MsgVerifyInvariant',
            sender,
            invariantModuleName,
            invariantRoute,
        };
    }
    static fromProto(proto) {
        return new MsgVerifyInvariant(proto.sender, proto.invariantModuleName, proto.invariantRoute);
    }
    toProto() {
        throw new Error('MsgVerifyInvarant is not allowed to send');
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.crisis.v1beta1.MsgVerifyInvariant',
            value: tx_1.MsgVerifyInvariant.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgVerifyInvariant.fromProto(tx_1.MsgVerifyInvariant.decode(msgAny.value));
    }
}
exports.MsgVerifyInvariant = MsgVerifyInvariant;
