// @initia/initia.proto/cosmos/app/runtime/v1alpha1/module.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StoreKeyConfig = exports.Module = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "cosmos.app.runtime.v1alpha1";
function createBaseModule() {
    return {
        appName: "",
        beginBlockers: [],
        endBlockers: [],
        initGenesis: [],
        exportGenesis: [],
        overrideStoreKeys: [],
        orderMigrations: [],
        precommiters: [],
        prepareCheckStaters: [],
    };
}
exports.Module = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.appName !== "") {
            writer.uint32(10).string(message.appName);
        }
        for (const v of message.beginBlockers) {
            writer.uint32(18).string(v);
        }
        for (const v of message.endBlockers) {
            writer.uint32(26).string(v);
        }
        for (const v of message.initGenesis) {
            writer.uint32(34).string(v);
        }
        for (const v of message.exportGenesis) {
            writer.uint32(42).string(v);
        }
        for (const v of message.overrideStoreKeys) {
            exports.StoreKeyConfig.encode(v, writer.uint32(50).fork()).ldelim();
        }
        for (const v of message.orderMigrations) {
            writer.uint32(58).string(v);
        }
        for (const v of message.precommiters) {
            writer.uint32(66).string(v);
        }
        for (const v of message.prepareCheckStaters) {
            writer.uint32(74).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseModule();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.appName = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.beginBlockers.push(reader.string());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.endBlockers.push(reader.string());
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.initGenesis.push(reader.string());
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.exportGenesis.push(reader.string());
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.overrideStoreKeys.push(exports.StoreKeyConfig.decode(reader, reader.uint32()));
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.orderMigrations.push(reader.string());
                    continue;
                case 8:
                    if (tag !== 66) {
                        break;
                    }
                    message.precommiters.push(reader.string());
                    continue;
                case 9:
                    if (tag !== 74) {
                        break;
                    }
                    message.prepareCheckStaters.push(reader.string());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            appName: isSet(object.appName) ? globalThis.String(object.appName) : "",
            beginBlockers: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.beginBlockers)
                ? object.beginBlockers.map((e) => globalThis.String(e))
                : [],
            endBlockers: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.endBlockers)
                ? object.endBlockers.map((e) => globalThis.String(e))
                : [],
            initGenesis: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.initGenesis)
                ? object.initGenesis.map((e) => globalThis.String(e))
                : [],
            exportGenesis: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.exportGenesis)
                ? object.exportGenesis.map((e) => globalThis.String(e))
                : [],
            overrideStoreKeys: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.overrideStoreKeys)
                ? object.overrideStoreKeys.map((e) => exports.StoreKeyConfig.fromJSON(e))
                : [],
            orderMigrations: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.orderMigrations)
                ? object.orderMigrations.map((e) => globalThis.String(e))
                : [],
            precommiters: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.precommiters)
                ? object.precommiters.map((e) => globalThis.String(e))
                : [],
            prepareCheckStaters: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.prepareCheckStaters)
                ? object.prepareCheckStaters.map((e) => globalThis.String(e))
                : [],
        };
    },
    toJSON(message) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const obj = {};
        if (message.appName !== "") {
            obj.appName = message.appName;
        }
        if ((_a = message.beginBlockers) === null || _a === void 0 ? void 0 : _a.length) {
            obj.beginBlockers = message.beginBlockers;
        }
        if ((_b = message.endBlockers) === null || _b === void 0 ? void 0 : _b.length) {
            obj.endBlockers = message.endBlockers;
        }
        if ((_c = message.initGenesis) === null || _c === void 0 ? void 0 : _c.length) {
            obj.initGenesis = message.initGenesis;
        }
        if ((_d = message.exportGenesis) === null || _d === void 0 ? void 0 : _d.length) {
            obj.exportGenesis = message.exportGenesis;
        }
        if ((_e = message.overrideStoreKeys) === null || _e === void 0 ? void 0 : _e.length) {
            obj.overrideStoreKeys = message.overrideStoreKeys.map((e) => exports.StoreKeyConfig.toJSON(e));
        }
        if ((_f = message.orderMigrations) === null || _f === void 0 ? void 0 : _f.length) {
            obj.orderMigrations = message.orderMigrations;
        }
        if ((_g = message.precommiters) === null || _g === void 0 ? void 0 : _g.length) {
            obj.precommiters = message.precommiters;
        }
        if ((_h = message.prepareCheckStaters) === null || _h === void 0 ? void 0 : _h.length) {
            obj.prepareCheckStaters = message.prepareCheckStaters;
        }
        return obj;
    },
    create(base) {
        return exports.Module.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        const message = createBaseModule();
        message.appName = (_a = object.appName) !== null && _a !== void 0 ? _a : "";
        message.beginBlockers = ((_b = object.beginBlockers) === null || _b === void 0 ? void 0 : _b.map((e) => e)) || [];
        message.endBlockers = ((_c = object.endBlockers) === null || _c === void 0 ? void 0 : _c.map((e) => e)) || [];
        message.initGenesis = ((_d = object.initGenesis) === null || _d === void 0 ? void 0 : _d.map((e) => e)) || [];
        message.exportGenesis = ((_e = object.exportGenesis) === null || _e === void 0 ? void 0 : _e.map((e) => e)) || [];
        message.overrideStoreKeys = ((_f = object.overrideStoreKeys) === null || _f === void 0 ? void 0 : _f.map((e) => exports.StoreKeyConfig.fromPartial(e))) || [];
        message.orderMigrations = ((_g = object.orderMigrations) === null || _g === void 0 ? void 0 : _g.map((e) => e)) || [];
        message.precommiters = ((_h = object.precommiters) === null || _h === void 0 ? void 0 : _h.map((e) => e)) || [];
        message.prepareCheckStaters = ((_j = object.prepareCheckStaters) === null || _j === void 0 ? void 0 : _j.map((e) => e)) || [];
        return message;
    },
};
function createBaseStoreKeyConfig() {
    return { moduleName: "", kvStoreKey: "" };
}
exports.StoreKeyConfig = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.moduleName !== "") {
            writer.uint32(10).string(message.moduleName);
        }
        if (message.kvStoreKey !== "") {
            writer.uint32(18).string(message.kvStoreKey);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStoreKeyConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.moduleName = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.kvStoreKey = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            moduleName: isSet(object.moduleName) ? globalThis.String(object.moduleName) : "",
            kvStoreKey: isSet(object.kvStoreKey) ? globalThis.String(object.kvStoreKey) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.moduleName !== "") {
            obj.moduleName = message.moduleName;
        }
        if (message.kvStoreKey !== "") {
            obj.kvStoreKey = message.kvStoreKey;
        }
        return obj;
    },
    create(base) {
        return exports.StoreKeyConfig.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseStoreKeyConfig();
        message.moduleName = (_a = object.moduleName) !== null && _a !== void 0 ? _a : "";
        message.kvStoreKey = (_b = object.kvStoreKey) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
