// @initia/initia.proto/cosmos/staking/module/v1/module.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Module = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "cosmos.staking.module.v1";
function createBaseModule() {
    return { hooksOrder: [], authority: "", bech32PrefixValidator: "", bech32PrefixConsensus: "" };
}
exports.Module = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.hooksOrder) {
            writer.uint32(10).string(v);
        }
        if (message.authority !== "") {
            writer.uint32(18).string(message.authority);
        }
        if (message.bech32PrefixValidator !== "") {
            writer.uint32(26).string(message.bech32PrefixValidator);
        }
        if (message.bech32PrefixConsensus !== "") {
            writer.uint32(34).string(message.bech32PrefixConsensus);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseModule();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.hooksOrder.push(reader.string());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.authority = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.bech32PrefixValidator = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.bech32PrefixConsensus = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            hooksOrder: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.hooksOrder)
                ? object.hooksOrder.map((e) => globalThis.String(e))
                : [],
            authority: isSet(object.authority) ? globalThis.String(object.authority) : "",
            bech32PrefixValidator: isSet(object.bech32PrefixValidator)
                ? globalThis.String(object.bech32PrefixValidator)
                : "",
            bech32PrefixConsensus: isSet(object.bech32PrefixConsensus)
                ? globalThis.String(object.bech32PrefixConsensus)
                : "",
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.hooksOrder) === null || _a === void 0 ? void 0 : _a.length) {
            obj.hooksOrder = message.hooksOrder;
        }
        if (message.authority !== "") {
            obj.authority = message.authority;
        }
        if (message.bech32PrefixValidator !== "") {
            obj.bech32PrefixValidator = message.bech32PrefixValidator;
        }
        if (message.bech32PrefixConsensus !== "") {
            obj.bech32PrefixConsensus = message.bech32PrefixConsensus;
        }
        return obj;
    },
    create(base) {
        return exports.Module.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseModule();
        message.hooksOrder = ((_a = object.hooksOrder) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        message.authority = (_b = object.authority) !== null && _b !== void 0 ? _b : "";
        message.bech32PrefixValidator = (_c = object.bech32PrefixValidator) !== null && _c !== void 0 ? _c : "";
        message.bech32PrefixConsensus = (_d = object.bech32PrefixConsensus) !== null && _d !== void 0 ? _d : "";
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
