// @initia/initia.js/dist/core/ophost/msgs/MsgUpdateOphostParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateOphostParams = void 0;
const json_1 = require("../../../util/json");
const OphostParams_1 = require("../OphostParams");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/opinit.proto/opinit/ophost/v1/tx");
class MsgUpdateOphostParams extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param params params defines the x/move parameters to update
     */
    constructor(authority, params) {
        super();
        this.authority = authority;
        this.params = params;
    }
    static fromAmino(data) {
        const { value: { authority, params }, } = data;
        return new MsgUpdateOphostParams(authority, OphostParams_1.OphostParams.fromAmino(params));
    }
    toAmino() {
        const { authority, params } = this;
        return {
            type: 'ophost/MsgUpdateParams',
            value: {
                authority,
                params: params.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, params } = data;
        return new MsgUpdateOphostParams(authority, OphostParams_1.OphostParams.fromData(params));
    }
    toData() {
        const { authority, params } = this;
        return {
            '@type': '/opinit.ophost.v1.MsgUpdateParams',
            authority,
            params: params.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateOphostParams(data.authority, OphostParams_1.OphostParams.fromProto(data.params));
    }
    toProto() {
        const { authority, params } = this;
        return tx_1.MsgUpdateParams.fromPartial({
            authority,
            params: params.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/opinit.ophost.v1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateOphostParams.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    }
}
exports.MsgUpdateOphostParams = MsgUpdateOphostParams;
