// @initia/initia.js/dist/core/ophost/msgs/MsgUpdateProposer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateProposer = void 0;
const json_1 = require("../../../util/json");
const tx_1 = require("@initia/opinit.proto/opinit/ophost/v1/tx");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const long_1 = __importDefault(require("long"));
class MsgUpdateProposer extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param bridge_id
     * @param new_proposer
     */
    constructor(authority, bridge_id, new_proposer) {
        super();
        this.authority = authority;
        this.bridge_id = bridge_id;
        this.new_proposer = new_proposer;
    }
    static fromAmino(data) {
        const { value: { authority, bridge_id, new_proposer }, } = data;
        return new MsgUpdateProposer(authority, Number.parseInt(bridge_id), new_proposer);
    }
    toAmino() {
        const { authority, bridge_id, new_proposer } = this;
        return {
            type: 'ophost/MsgUpdateProposer',
            value: {
                authority,
                bridge_id: bridge_id.toString(),
                new_proposer,
            },
        };
    }
    static fromData(data) {
        const { authority, bridge_id, new_proposer } = data;
        return new MsgUpdateProposer(authority, Number.parseInt(bridge_id), new_proposer);
    }
    toData() {
        const { authority, bridge_id, new_proposer } = this;
        return {
            '@type': '/opinit.ophost.v1.MsgUpdateProposer',
            authority,
            bridge_id: bridge_id.toString(),
            new_proposer,
        };
    }
    static fromProto(data) {
        return new MsgUpdateProposer(data.authority, data.bridgeId.toNumber(), data.newProposer);
    }
    toProto() {
        const { authority, bridge_id, new_proposer } = this;
        return tx_1.MsgUpdateProposer.fromPartial({
            authority,
            bridgeId: long_1.default.fromNumber(bridge_id),
            newProposer: new_proposer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/opinit.ophost.v1.MsgUpdateProposer',
            value: tx_1.MsgUpdateProposer.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateProposer.fromProto(tx_1.MsgUpdateProposer.decode(msgAny.value));
    }
}
exports.MsgUpdateProposer = MsgUpdateProposer;
