// @initia/initia.js/dist/core/ophost/msgs/MsgUpdateChallenger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateChallenger = void 0;
const json_1 = require("../../../util/json");
const tx_1 = require("@initia/opinit.proto/opinit/ophost/v1/tx");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const long_1 = __importDefault(require("long"));
class MsgUpdateChallenger extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param bridge_id
     * @param new_challenger
     */
    constructor(authority, bridge_id, new_challenger) {
        super();
        this.authority = authority;
        this.bridge_id = bridge_id;
        this.new_challenger = new_challenger;
    }
    static fromAmino(data) {
        const { value: { authority, bridge_id, new_challenger }, } = data;
        return new MsgUpdateChallenger(authority, Number.parseInt(bridge_id), new_challenger);
    }
    toAmino() {
        const { authority, bridge_id, new_challenger } = this;
        return {
            type: 'ophost/MsgUpdateChallenger',
            value: {
                authority,
                bridge_id: bridge_id.toString(),
                new_challenger,
            },
        };
    }
    static fromData(data) {
        const { authority, bridge_id, new_challenger } = data;
        return new MsgUpdateChallenger(authority, Number.parseInt(bridge_id), new_challenger);
    }
    toData() {
        const { authority, bridge_id, new_challenger } = this;
        return {
            '@type': '/opinit.ophost.v1.MsgUpdateChallenger',
            authority,
            bridge_id: bridge_id.toString(),
            new_challenger,
        };
    }
    static fromProto(data) {
        return new MsgUpdateChallenger(data.authority, data.bridgeId.toNumber(), data.newChallenger);
    }
    toProto() {
        const { authority, bridge_id, new_challenger } = this;
        return tx_1.MsgUpdateChallenger.fromPartial({
            authority,
            bridgeId: long_1.default.fromNumber(bridge_id),
            newChallenger: new_challenger,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/opinit.ophost.v1.MsgUpdateChallenger',
            value: tx_1.MsgUpdateChallenger.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateChallenger.fromProto(tx_1.MsgUpdateChallenger.decode(msgAny.value));
    }
}
exports.MsgUpdateChallenger = MsgUpdateChallenger;
