// @initia/initia.js/dist/core/wasm/authorizations/AllowAllMessagesFilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllowAllMessagesFilter = void 0;
const json_1 = require("../../../util/json");
const authz_1 = require("@initia/initia.proto/cosmwasm/wasm/v1/authz");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
class AllowAllMessagesFilter extends json_1.JSONSerializable {
    constructor() {
        super();
    }
    static fromAmino(_) {
        _;
        return new AllowAllMessagesFilter();
    }
    toAmino() {
        return {
            type: 'wasm/AllowAllMessagesFilter',
            value: {},
        };
    }
    static fromData(_) {
        _;
        return new AllowAllMessagesFilter();
    }
    toData() {
        return {
            '@type': '/cosmwasm.wasm.v1.AllowAllMessagesFilter',
        };
    }
    static fromProto(_) {
        _;
        return new AllowAllMessagesFilter();
    }
    toProto() {
        return authz_1.AllowAllMessagesFilter.fromPartial({});
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.AllowAllMessagesFilter',
            value: authz_1.AllowAllMessagesFilter.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return AllowAllMessagesFilter.fromProto(authz_1.AllowAllMessagesFilter.decode(msgAny.value));
    }
}
exports.AllowAllMessagesFilter = AllowAllMessagesFilter;
