// @initia/initia.js/dist/core/wasm/authorizations/CodeGrant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeGrant = void 0;
const json_1 = require("../../../util/json");
const authz_1 = require("@initia/initia.proto/cosmwasm/wasm/v1/authz");
const AccessConfig_1 = require("../AccessConfig");
class CodeGrant extends json_1.JSONSerializable {
    constructor(code_hash, instantiate_permission) {
        super();
        this.code_hash = code_hash;
        this.instantiate_permission = instantiate_permission;
    }
    static fromAmino(data) {
        const { code_hash, instantiate_permission } = data;
        return new CodeGrant(code_hash, instantiate_permission
            ? AccessConfig_1.AccessConfig.fromAmino(instantiate_permission)
            : undefined);
    }
    toAmino() {
        const { code_hash, instantiate_permission } = this;
        return {
            code_hash,
            instantiate_permission: instantiate_permission === null || instantiate_permission === void 0 ? void 0 : instantiate_permission.toAmino(),
        };
    }
    static fromData(data) {
        const { code_hash, instantiate_permission } = data;
        return new CodeGrant(code_hash, instantiate_permission
            ? AccessConfig_1.AccessConfig.fromData(instantiate_permission)
            : undefined);
    }
    toData() {
        const { code_hash, instantiate_permission } = this;
        return {
            code_hash,
            instantiate_permission: instantiate_permission === null || instantiate_permission === void 0 ? void 0 : instantiate_permission.toData(),
        };
    }
    static fromProto(data) {
        return new CodeGrant(Buffer.from(data.codeHash).toString('base64'), data.instantiatePermission
            ? AccessConfig_1.AccessConfig.fromProto(data.instantiatePermission)
            : undefined);
    }
    toProto() {
        const { code_hash, instantiate_permission } = this;
        return authz_1.CodeGrant.fromPartial({
            codeHash: Buffer.from(code_hash, 'base64'),
            instantiatePermission: instantiate_permission === null || instantiate_permission === void 0 ? void 0 : instantiate_permission.toProto(),
        });
    }
}
exports.CodeGrant = CodeGrant;
