// @initia/initia.js/dist/core/wasm/authorizations/ContractMigrationAuthorization.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractMigrationAuthorization = void 0;
const json_1 = require("../../../util/json");
const authz_1 = require("@initia/initia.proto/cosmwasm/wasm/v1/authz");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const ContractGrant_1 = require("./ContractGrant");
class ContractMigrationAuthorization extends json_1.JSONSerializable {
    constructor(grants) {
        super();
        this.grants = grants;
    }
    static fromAmino(data) {
        return new ContractMigrationAuthorization(data.value.grants.map(ContractGrant_1.ContractGrant.fromAmino));
    }
    toAmino() {
        return {
            type: 'wasm/ContractMigrationAuthorization',
            value: { grants: this.grants.map(grant => grant.toAmino()) },
        };
    }
    static fromData(data) {
        return new ContractMigrationAuthorization(data.grants.map(ContractGrant_1.ContractGrant.fromData));
    }
    toData() {
        return {
            '@type': '/cosmwasm.wasm.v1.ContractMigrationAuthorization',
            grants: this.grants.map(grant => grant.toData()),
        };
    }
    static fromProto(data) {
        return new ContractMigrationAuthorization(data.grants.map(ContractGrant_1.ContractGrant.fromProto));
    }
    toProto() {
        return authz_1.ContractMigrationAuthorization.fromPartial({
            grants: this.grants.map(grant => grant.toProto()),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.ContractMigrationAuthorization',
            value: authz_1.ContractMigrationAuthorization.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return ContractMigrationAuthorization.fromProto(authz_1.ContractMigrationAuthorization.decode(msgAny.value));
    }
}
exports.ContractMigrationAuthorization = ContractMigrationAuthorization;
