// @initia/initia.js/dist/core/wasm/authorizations/StoreCodeAuthorization.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StoreCodeAuthorization = void 0;
const json_1 = require("../../../util/json");
const authz_1 = require("@initia/initia.proto/cosmwasm/wasm/v1/authz");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const CodeGrant_1 = require("./CodeGrant");
class StoreCodeAuthorization extends json_1.JSONSerializable {
    constructor(grants) {
        super();
        this.grants = grants;
    }
    static fromAmino(data) {
        return new StoreCodeAuthorization(data.value.grants.map(CodeGrant_1.CodeGrant.fromAmino));
    }
    toAmino() {
        return {
            type: 'wasm/StoreCodeAuthorization',
            value: { grants: this.grants.map(grant => grant.toAmino()) },
        };
    }
    static fromData(data) {
        return new StoreCodeAuthorization(data.grants.map(CodeGrant_1.CodeGrant.fromData));
    }
    toData() {
        return {
            '@type': '/cosmwasm.wasm.v1.StoreCodeAuthorization',
            grants: this.grants.map(grant => grant.toData()),
        };
    }
    static fromProto(data) {
        return new StoreCodeAuthorization(data.grants.map(CodeGrant_1.CodeGrant.fromProto));
    }
    toProto() {
        return authz_1.StoreCodeAuthorization.fromPartial({
            grants: this.grants.map(grant => grant.toProto()),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.StoreCodeAuthorization',
            value: authz_1.StoreCodeAuthorization.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return StoreCodeAuthorization.fromProto(authz_1.StoreCodeAuthorization.decode(msgAny.value));
    }
}
exports.StoreCodeAuthorization = StoreCodeAuthorization;
