// @initia/initia.js/dist/core/distribution/RewardWeight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RewardWeight = void 0;
const json_1 = require("../../util/json");
const distribution_1 = require("@initia/initia.proto/initia/distribution/v1/distribution");
class RewardWeight extends json_1.JSONSerializable {
    /**
     * @param denom
     * @param weight
     */
    constructor(denom, weight) {
        super();
        this.denom = denom;
        this.weight = weight;
    }
    static fromAmino(data) {
        const { denom, weight } = data;
        return new RewardWeight(denom, weight);
    }
    toAmino() {
        const { denom, weight } = this;
        return {
            denom,
            weight,
        };
    }
    static fromData(data) {
        const { denom, weight } = data;
        return new RewardWeight(denom, weight);
    }
    toData() {
        const { denom, weight } = this;
        return {
            denom,
            weight,
        };
    }
    static fromProto(data) {
        return new RewardWeight(data.denom, data.weight);
    }
    toProto() {
        const { denom, weight } = this;
        return distribution_1.RewardWeight.fromPartial({
            denom,
            weight,
        });
    }
}
exports.RewardWeight = RewardWeight;
