// @initia/initia.proto/initia/gov/v1/genesis.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenesisState = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const gov_1 = require("../../../cosmos/gov/v1/gov");
const gov_2 = require("./gov");
exports.protobufPackage = "initia.gov.v1";
function createBaseGenesisState() {
    return {
        startingProposalId: long_1.default.UZERO,
        deposits: [],
        votes: [],
        proposals: [],
        params: undefined,
        constitution: "",
    };
}
exports.GenesisState = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (!message.startingProposalId.isZero()) {
            writer.uint32(8).uint64(message.startingProposalId);
        }
        for (const v of message.deposits) {
            gov_1.Deposit.encode(v, writer.uint32(18).fork()).ldelim();
        }
        for (const v of message.votes) {
            gov_1.Vote.encode(v, writer.uint32(26).fork()).ldelim();
        }
        for (const v of message.proposals) {
            gov_2.Proposal.encode(v, writer.uint32(34).fork()).ldelim();
        }
        if (message.params !== undefined) {
            gov_2.Params.encode(message.params, writer.uint32(42).fork()).ldelim();
        }
        if (message.constitution !== "") {
            writer.uint32(50).string(message.constitution);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.startingProposalId = reader.uint64();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.deposits.push(gov_1.Deposit.decode(reader, reader.uint32()));
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.votes.push(gov_1.Vote.decode(reader, reader.uint32()));
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.proposals.push(gov_2.Proposal.decode(reader, reader.uint32()));
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.params = gov_2.Params.decode(reader, reader.uint32());
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.constitution = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            startingProposalId: isSet(object.startingProposalId)
                ? long_1.default.fromValue(object.startingProposalId)
                : long_1.default.UZERO,
            deposits: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.deposits)
                ? object.deposits.map((e) => gov_1.Deposit.fromJSON(e))
                : [],
            votes: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.votes) ? object.votes.map((e) => gov_1.Vote.fromJSON(e)) : [],
            proposals: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.proposals)
                ? object.proposals.map((e) => gov_2.Proposal.fromJSON(e))
                : [],
            params: isSet(object.params) ? gov_2.Params.fromJSON(object.params) : undefined,
            constitution: isSet(object.constitution) ? globalThis.String(object.constitution) : "",
        };
    },
    toJSON(message) {
        var _a, _b, _c;
        const obj = {};
        if (!message.startingProposalId.isZero()) {
            obj.startingProposalId = (message.startingProposalId || long_1.default.UZERO).toString();
        }
        if ((_a = message.deposits) === null || _a === void 0 ? void 0 : _a.length) {
            obj.deposits = message.deposits.map((e) => gov_1.Deposit.toJSON(e));
        }
        if ((_b = message.votes) === null || _b === void 0 ? void 0 : _b.length) {
            obj.votes = message.votes.map((e) => gov_1.Vote.toJSON(e));
        }
        if ((_c = message.proposals) === null || _c === void 0 ? void 0 : _c.length) {
            obj.proposals = message.proposals.map((e) => gov_2.Proposal.toJSON(e));
        }
        if (message.params !== undefined) {
            obj.params = gov_2.Params.toJSON(message.params);
        }
        if (message.constitution !== "") {
            obj.constitution = message.constitution;
        }
        return obj;
    },
    create(base) {
        return exports.GenesisState.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseGenesisState();
        message.startingProposalId =
            object.startingProposalId !== undefined && object.startingProposalId !== null
                ? long_1.default.fromValue(object.startingProposalId)
                : long_1.default.UZERO;
        message.deposits = ((_a = object.deposits) === null || _a === void 0 ? void 0 : _a.map((e) => gov_1.Deposit.fromPartial(e))) || [];
        message.votes = ((_b = object.votes) === null || _b === void 0 ? void 0 : _b.map((e) => gov_1.Vote.fromPartial(e))) || [];
        message.proposals = ((_c = object.proposals) === null || _c === void 0 ? void 0 : _c.map((e) => gov_2.Proposal.fromPartial(e))) || [];
        message.params =
            object.params !== undefined && object.params !== null ? gov_2.Params.fromPartial(object.params) : undefined;
        message.constitution = (_d = object.constitution) !== null && _d !== void 0 ? _d : "";
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
