// @initia/initia.js/dist/core/ibc/core/connection/ConnectionCounterparty.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectionCounterparty = void 0;
const json_1 = require("../../../../util/json");
const connection_1 = require("@initia/initia.proto/ibc/core/connection/v1/connection");
const MerklePrefix_1 = require("../commitment/MerklePrefix");
/** ConnectionCounterparty defines the ConnectionCounterparty chain associated with a connection end */
class ConnectionCounterparty extends json_1.JSONSerializable {
    /**
     * @param client_id identifies the client on the ConnectionCounterparty chain associated with a given connection.
     * @param connection_id identifies the connection end on the ConnectionCounterparty chain associated with a given connection.
     * @param prefix commitment merkle prefix of the ConnectionCounterparty chain.
     */
    constructor(client_id, connection_id, prefix) {
        super();
        this.client_id = client_id;
        this.connection_id = connection_id;
        this.prefix = prefix;
    }
    static fromAmino(data) {
        const { client_id, connection_id, prefix } = data;
        return new ConnectionCounterparty(client_id, connection_id, prefix ? MerklePrefix_1.MerklePrefix.fromAmino(prefix) : undefined);
    }
    toAmino() {
        const { client_id, connection_id, prefix } = this;
        const res = {
            client_id,
            connection_id,
            prefix,
        };
        return res;
    }
    static fromData(data) {
        const { client_id, connection_id, prefix } = data;
        return new ConnectionCounterparty(client_id, connection_id, prefix ? MerklePrefix_1.MerklePrefix.fromData(prefix) : undefined);
    }
    toData() {
        const { client_id, connection_id, prefix } = this;
        const res = {
            client_id,
            connection_id,
            prefix: prefix === null || prefix === void 0 ? void 0 : prefix.toData(),
        };
        return res;
    }
    static fromProto(proto) {
        return new ConnectionCounterparty(proto.clientId, proto.connectionId, proto.prefix ? MerklePrefix_1.MerklePrefix.fromProto(proto.prefix) : undefined);
    }
    toProto() {
        const { client_id, connection_id, prefix } = this;
        return connection_1.Counterparty.fromPartial({
            clientId: client_id,
            connectionId: connection_id,
            prefix: prefix === null || prefix === void 0 ? void 0 : prefix.toProto(),
        });
    }
}
exports.ConnectionCounterparty = ConnectionCounterparty;
