// @initia/initia.js/dist/core/ibc/core/connection/IbcVersion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IbcVersion = void 0;
const connection_1 = require("@initia/initia.proto/ibc/core/connection/v1/connection");
const json_1 = require("../../../../util/json");
/*
 * IbcVersion defines the versioning scheme used to negotiate the IBC verison in the connection handshake.
 */
class IbcVersion extends json_1.JSONSerializable {
    /**
     * @param identifier unique version identifier
     * @param features list of features compatible with the specified identifier
     */
    constructor(identifier, features) {
        super();
        this.identifier = identifier;
        this.features = features;
    }
    static fromAmino(data) {
        const { identifier, features } = data;
        return new IbcVersion(identifier, features);
    }
    toAmino() {
        const { identifier, features } = this;
        const res = {
            identifier,
            features,
        };
        return res;
    }
    static fromData(data) {
        const { identifier, features } = data;
        return new IbcVersion(identifier, features);
    }
    toData() {
        const { identifier, features } = this;
        const res = {
            identifier,
            features,
        };
        return res;
    }
    static fromProto(proto) {
        return new IbcVersion(proto.identifier, proto.features);
    }
    toProto() {
        const { identifier, features } = this;
        return connection_1.Version.fromPartial({ identifier, features });
    }
}
exports.IbcVersion = IbcVersion;
