// @initia/initia.js/dist/core/ibc/applications/transfer/FungibleTokenPacketData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FungibleTokenPacketData = void 0;
const packet_1 = require("@initia/initia.proto/ibc/applications/transfer/v2/packet");
const json_1 = require("../../../../util/json");
/**
 *  FungibleTokenPacketData defines a struct for the packet payload
 * See FungibleTokenPacketData spec:
 * https://github.com/cosmos/ibc/tree/master/spec/app/ics-020-fungible-token-transfer#data-structures
 */
class FungibleTokenPacketData extends json_1.JSONSerializable {
    /**
     * @param denom the token denomination to be transferred
     * @param amount the token amount to be transferred
     * @param sender the sender address
     * @param receiver the recipient address on the destination chain
     * @param memo optional memo
     */
    constructor(denom, amount, sender, receiver, memo) {
        super();
        this.denom = denom;
        this.amount = amount;
        this.sender = sender;
        this.receiver = receiver;
        this.memo = memo;
    }
    static fromAmino(data) {
        const { denom, amount, sender, receiver, memo } = data;
        return new FungibleTokenPacketData(denom, amount, sender, receiver, memo);
    }
    toAmino() {
        const { denom, amount, sender, receiver, memo } = this;
        const res = {
            denom,
            amount,
            sender,
            receiver,
            memo,
        };
        return res;
    }
    static fromData(data) {
        const { denom, amount, sender, receiver, memo } = data;
        return new FungibleTokenPacketData(denom, amount, sender, receiver, memo);
    }
    toData() {
        const { denom, amount, sender, receiver, memo } = this;
        const res = {
            denom,
            amount,
            sender,
            receiver,
            memo,
        };
        return res;
    }
    static fromProto(proto) {
        return new FungibleTokenPacketData(proto.denom, proto.amount, proto.sender, proto.receiver, proto.memo);
    }
    toProto() {
        const { denom, amount, sender, receiver, memo } = this;
        return packet_1.FungibleTokenPacketData.fromPartial({
            denom,
            amount,
            sender,
            receiver,
            memo,
        });
    }
}
exports.FungibleTokenPacketData = FungibleTokenPacketData;
