// @initia/initia.js/dist/core/ibc/applications/transfer/IbcTransferParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IbcTransferParams = void 0;
const json_1 = require("../../../../util/json");
const transfer_1 = require("@initia/initia.proto/ibc/applications/transfer/v1/transfer");
class IbcTransferParams extends json_1.JSONSerializable {
    /**
     * @param send_enabled enables or disables all cross-chain token transfers from this chain
     * @param receive_enabled enables or disables all cross-chain token transfers to this chain
     */
    constructor(send_enabled, receive_enabled) {
        super();
        this.send_enabled = send_enabled;
        this.receive_enabled = receive_enabled;
    }
    static fromAmino(data) {
        const { send_enabled, receive_enabled } = data;
        return new IbcTransferParams(send_enabled, receive_enabled);
    }
    toAmino() {
        const { send_enabled, receive_enabled } = this;
        return {
            send_enabled,
            receive_enabled,
        };
    }
    static fromData(data) {
        const { send_enabled, receive_enabled } = data;
        return new IbcTransferParams(send_enabled, receive_enabled);
    }
    toData() {
        const { send_enabled, receive_enabled } = this;
        return {
            send_enabled,
            receive_enabled,
        };
    }
    static fromProto(data) {
        return new IbcTransferParams(data.sendEnabled, data.receiveEnabled);
    }
    toProto() {
        const { send_enabled, receive_enabled } = this;
        return transfer_1.Params.fromPartial({
            sendEnabled: send_enabled,
            receiveEnabled: receive_enabled,
        });
    }
}
exports.IbcTransferParams = IbcTransferParams;
