// @initia/initia.js/dist/core/group/policies/PercentageDecisionPolicy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PercentageDecisionPolicy = void 0;
const json_1 = require("../../../util/json");
const DecisionPolicyWindows_1 = require("./DecisionPolicyWindows");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const types_1 = require("@initia/initia.proto/cosmos/group/v1/types");
class PercentageDecisionPolicy extends json_1.JSONSerializable {
    /**
     * @param percentage the minimum percentage of the weighted sum of `YES` votes must meet for a proposal to succeed
     * @param windows the different windows for voting and execution
     */
    constructor(percentage, windows) {
        super();
        this.percentage = percentage;
        this.windows = windows;
    }
    static fromAmino(data) {
        const { value: { percentage, windows }, } = data;
        return new PercentageDecisionPolicy(percentage, DecisionPolicyWindows_1.DecisionPolicyWindows.fromAmino(windows));
    }
    toAmino() {
        const { percentage, windows } = this;
        return {
            type: 'cosmos-sdk/PercentageDecisionPolicy',
            value: {
                percentage,
                windows: windows.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { percentage, windows } = data;
        return new PercentageDecisionPolicy(percentage, DecisionPolicyWindows_1.DecisionPolicyWindows.fromData(windows));
    }
    toData() {
        const { percentage, windows } = this;
        return {
            '@type': '/cosmos.group.v1.PercentageDecisionPolicy',
            percentage,
            windows: windows.toData(),
        };
    }
    static fromProto(data) {
        return new PercentageDecisionPolicy(data.percentage, DecisionPolicyWindows_1.DecisionPolicyWindows.fromProto(data.windows));
    }
    toProto() {
        const { percentage, windows } = this;
        return types_1.PercentageDecisionPolicy.fromPartial({
            percentage,
            windows: windows.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.PercentageDecisionPolicy',
            value: types_1.PercentageDecisionPolicy.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return PercentageDecisionPolicy.fromProto(types_1.PercentageDecisionPolicy.decode(msgAny.value));
    }
}
exports.PercentageDecisionPolicy = PercentageDecisionPolicy;
