// @initia/initia.js/dist/core/bank/BankParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BankParams = void 0;
const json_1 = require("../../util/json");
const SendEnabled_1 = require("./SendEnabled");
const bank_1 = require("@initia/initia.proto/cosmos/bank/v1beta1/bank");
class BankParams extends json_1.JSONSerializable {
    /**
     * @param send_enabled
     * @param default_send_enabled
     */
    constructor(send_enabled, default_send_enabled) {
        super();
        this.send_enabled = send_enabled;
        this.default_send_enabled = default_send_enabled;
    }
    static fromAmino(data) {
        const { value: { send_enabled, default_send_enabled }, } = data;
        return new BankParams(send_enabled.map(SendEnabled_1.SendEnabled.fromAmino), default_send_enabled);
    }
    toAmino() {
        const { send_enabled, default_send_enabled } = this;
        return {
            type: 'cosmos-sdk/x/bank/Params',
            value: {
                send_enabled: send_enabled.map(d => d.toAmino()),
                default_send_enabled,
            },
        };
    }
    static fromData(data) {
        const { send_enabled, default_send_enabled } = data;
        return new BankParams(send_enabled.map(SendEnabled_1.SendEnabled.fromData), default_send_enabled);
    }
    toData() {
        const { send_enabled, default_send_enabled } = this;
        return {
            '@type': '/cosmos.bank.v1beta1.Params',
            send_enabled: send_enabled.map(d => d.toData()),
            default_send_enabled,
        };
    }
    static fromProto(data) {
        return new BankParams(data.sendEnabled.map(SendEnabled_1.SendEnabled.fromProto), data.defaultSendEnabled);
    }
    toProto() {
        const { send_enabled, default_send_enabled } = this;
        return bank_1.Params.fromPartial({
            sendEnabled: send_enabled.map(d => d.toProto()),
            defaultSendEnabled: default_send_enabled,
        });
    }
}
exports.BankParams = BankParams;
