// @initia/initia.js/dist/core/bank/SendEnabled.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendEnabled = void 0;
const json_1 = require("../../util/json");
const bank_1 = require("@initia/initia.proto/cosmos/bank/v1beta1/bank");
class SendEnabled extends json_1.JSONSerializable {
    /**
     * @param denom
     * @param enabled
     */
    constructor(denom, enabled) {
        super();
        this.denom = denom;
        this.enabled = enabled;
    }
    static fromAmino(data) {
        const { denom, enabled } = data;
        return new SendEnabled(denom, enabled);
    }
    toAmino() {
        const { denom, enabled } = this;
        return { denom, enabled };
    }
    static fromData(data) {
        const { denom, enabled } = data;
        return new SendEnabled(denom, enabled);
    }
    toData() {
        const { denom, enabled } = this;
        return { denom, enabled };
    }
    static fromProto(data) {
        return new SendEnabled(data.denom, data.enabled);
    }
    toProto() {
        const { denom, enabled } = this;
        return bank_1.SendEnabled.fromPartial({
            denom,
            enabled,
        });
    }
}
exports.SendEnabled = SendEnabled;
