// @initia/initia.js/dist/core/bank/msgs/MsgSetDenomMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgSetDenomMetadata = void 0;
const json_1 = require("../../../util/json");
const DenomMetadata_1 = require("../DenomMetadata");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/initia/bank/v1/tx");
class MsgSetDenomMetadata extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param metadata the x/bank denom metadata to update
     */
    constructor(authority, metadata) {
        super();
        this.authority = authority;
        this.metadata = metadata;
    }
    static fromAmino(data) {
        const { value: { authority, metadata }, } = data;
        return new MsgSetDenomMetadata(authority, DenomMetadata_1.DenomMetadata.fromAmino(metadata));
    }
    toAmino() {
        const { authority, metadata } = this;
        return {
            type: 'bank/MsgSetDenomMetadata',
            value: {
                authority,
                metadata: metadata.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, metadata } = data;
        return new MsgSetDenomMetadata(authority, DenomMetadata_1.DenomMetadata.fromData(metadata));
    }
    toData() {
        const { authority, metadata } = this;
        return {
            '@type': '/initia.bank.v1.MsgSetDenomMetadata',
            authority,
            metadata: metadata.toData(),
        };
    }
    static fromProto(data) {
        return new MsgSetDenomMetadata(data.authority, DenomMetadata_1.DenomMetadata.fromProto(data.metadata));
    }
    toProto() {
        const { authority, metadata } = this;
        return tx_1.MsgSetDenomMetadata.fromPartial({
            authority,
            metadata: metadata.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/initia.bank.v1.MsgSetDenomMetadata',
            value: tx_1.MsgSetDenomMetadata.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgSetDenomMetadata.fromProto(tx_1.MsgSetDenomMetadata.decode(msgAny.value));
    }
}
exports.MsgSetDenomMetadata = MsgSetDenomMetadata;
