// @initia/initia.js/dist/core/bank/msgs/MsgSend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgSend = void 0;
const Coins_1 = require("../../Coins");
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmos/bank/v1beta1/tx");
/**
 * A basic message for sending [[Coins]] between Initia accounts.
 */
class MsgSend extends json_1.JSONSerializable {
    /**
     * @param from_address sender's address
     * @param to_address recipient's address
     * @param amount value of the transaction
     */
    constructor(from_address, to_address, amount) {
        super();
        this.from_address = from_address;
        this.to_address = to_address;
        this.amount = new Coins_1.Coins(amount);
    }
    static fromAmino(data) {
        const { value: { from_address, to_address, amount }, } = data;
        return new MsgSend(from_address, to_address, Coins_1.Coins.fromAmino(amount));
    }
    toAmino() {
        const { from_address, to_address, amount } = this;
        return {
            type: 'cosmos-sdk/MsgSend',
            value: {
                from_address,
                to_address,
                amount: amount.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { from_address, to_address, amount } = data;
        return new MsgSend(from_address, to_address, Coins_1.Coins.fromData(amount));
    }
    toData() {
        const { from_address, to_address, amount } = this;
        return {
            '@type': '/cosmos.bank.v1beta1.MsgSend',
            from_address,
            to_address,
            amount: amount.toData(),
        };
    }
    static fromProto(proto) {
        return new MsgSend(proto.fromAddress, proto.toAddress, Coins_1.Coins.fromProto(proto.amount));
    }
    toProto() {
        const { from_address, to_address, amount } = this;
        return tx_1.MsgSend.fromPartial({
            fromAddress: from_address,
            toAddress: to_address,
            amount: amount.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.bank.v1beta1.MsgSend',
            value: tx_1.MsgSend.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgSend.fromProto(tx_1.MsgSend.decode(msgAny.value));
    }
}
exports.MsgSend = MsgSend;
