// @initia/initia.js/dist/core/mstaking/msgs/MsgDelegate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgDelegate = void 0;
const Coins_1 = require("../../Coins");
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/initia/mstaking/v1/tx");
/**
 * A delegator can submit this message to send more Initia to be staked through a
 * validator delegate.
 */
class MsgDelegate extends json_1.JSONSerializable {
    /**
     *
     * @param delegator_address delegator's account address
     * @param validator_address validator's operator address
     * @param amount amount of INIT to be sent for delegation
     */
    constructor(delegator_address, validator_address, amount) {
        super();
        this.delegator_address = delegator_address;
        this.validator_address = validator_address;
        this.amount = new Coins_1.Coins(amount);
    }
    static fromAmino(data) {
        const { value: { delegator_address, validator_address, amount }, } = data;
        return new MsgDelegate(delegator_address, validator_address, Coins_1.Coins.fromAmino(amount));
    }
    toAmino() {
        const { delegator_address, validator_address, amount } = this;
        return {
            type: 'mstaking/MsgDelegate',
            value: {
                delegator_address,
                validator_address,
                amount: amount.toAmino(),
            },
        };
    }
    static fromProto(proto) {
        return new MsgDelegate(proto.delegatorAddress, proto.validatorAddress, Coins_1.Coins.fromProto(proto.amount));
    }
    toProto() {
        const { delegator_address, validator_address, amount } = this;
        return tx_1.MsgDelegate.fromPartial({
            amount: amount.toProto(),
            delegatorAddress: delegator_address,
            validatorAddress: validator_address,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/initia.mstaking.v1.MsgDelegate',
            value: tx_1.MsgDelegate.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgDelegate.fromProto(tx_1.MsgDelegate.decode(msgAny.value));
    }
    static fromData(data) {
        const { delegator_address, validator_address, amount } = data;
        return new MsgDelegate(delegator_address, validator_address, Coins_1.Coins.fromData(amount));
    }
    toData() {
        const { delegator_address, validator_address, amount } = this;
        return {
            '@type': '/initia.mstaking.v1.MsgDelegate',
            delegator_address,
            validator_address,
            amount: amount.toData(),
        };
    }
}
exports.MsgDelegate = MsgDelegate;
