// @initia/initia.js/dist/core/mstaking/msgs/MsgUndelegate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUndelegate = void 0;
const Coins_1 = require("../../Coins");
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/initia/mstaking/v1/tx");
/**
 * A delegator can undelegate an amount of bonded Initia, and will begin the unbonding
 * process for those funds. The unbonding process takes 21 days to complete, during
 * which the Initia cannot be transacted or swapped.
 */
class MsgUndelegate extends json_1.JSONSerializable {
    /**
     * @param delegator_address delegator's account address
     * @param validator_address validator's operator address
     * @param amount INIT to be undelegated
     */
    constructor(delegator_address, validator_address, amount) {
        super();
        this.delegator_address = delegator_address;
        this.validator_address = validator_address;
        this.amount = new Coins_1.Coins(amount);
    }
    static fromAmino(data) {
        const { value: { delegator_address, validator_address, amount }, } = data;
        return new MsgUndelegate(delegator_address, validator_address, Coins_1.Coins.fromAmino(amount));
    }
    toAmino() {
        const { delegator_address, validator_address, amount } = this;
        return {
            type: 'mstaking/MsgUndelegate',
            value: {
                delegator_address,
                validator_address,
                amount: amount.toAmino(),
            },
        };
    }
    static fromProto(proto) {
        return new MsgUndelegate(proto.delegatorAddress, proto.validatorAddress, Coins_1.Coins.fromProto(proto.amount));
    }
    toProto() {
        const { delegator_address, validator_address, amount } = this;
        return tx_1.MsgUndelegate.fromPartial({
            amount: amount.toProto(),
            delegatorAddress: delegator_address,
            validatorAddress: validator_address,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/initia.mstaking.v1.MsgUndelegate',
            value: tx_1.MsgUndelegate.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUndelegate.fromProto(tx_1.MsgUndelegate.decode(msgAny.value));
    }
    static fromData(data) {
        const { delegator_address, validator_address, amount } = data;
        return new MsgUndelegate(delegator_address, validator_address, Coins_1.Coins.fromData(amount));
    }
    toData() {
        const { delegator_address, validator_address, amount } = this;
        return {
            '@type': '/initia.mstaking.v1.MsgUndelegate',
            delegator_address,
            validator_address,
            amount: amount.toData(),
        };
    }
}
exports.MsgUndelegate = MsgUndelegate;
