// @initia/initia.js/dist/core/mstaking/msgs/MsgUpdateMstakingParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateMstakingParams = void 0;
const json_1 = require("../../../util/json");
const MstakingParams_1 = require("../MstakingParams");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/initia/mstaking/v1/tx");
class MsgUpdateMstakingParams extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param params params defines the x/mstaking parameters to update
     */
    constructor(authority, params) {
        super();
        this.authority = authority;
        this.params = params;
    }
    static fromAmino(data) {
        const { value: { authority, params }, } = data;
        return new MsgUpdateMstakingParams(authority, MstakingParams_1.MstakingParams.fromAmino(params));
    }
    toAmino() {
        const { authority, params } = this;
        return {
            type: 'mstaking/MsgUpdateParams',
            value: {
                authority,
                params: params.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, params } = data;
        return new MsgUpdateMstakingParams(authority, MstakingParams_1.MstakingParams.fromData(params));
    }
    toData() {
        const { authority, params } = this;
        return {
            '@type': '/initia.mstaking.v1.MsgUpdateParams',
            authority,
            params: params.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateMstakingParams(data.authority, MstakingParams_1.MstakingParams.fromProto(data.params));
    }
    toProto() {
        const { authority, params } = this;
        return tx_1.MsgUpdateParams.fromPartial({
            authority,
            params: params.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/initia.mstaking.v1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateMstakingParams.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    }
}
exports.MsgUpdateMstakingParams = MsgUpdateMstakingParams;
