// @initia/initia.proto/cosmos/group/module/v1/module.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Module = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const duration_1 = require("../../../../google/protobuf/duration");
exports.protobufPackage = "cosmos.group.module.v1";
function createBaseModule() {
    return { maxExecutionPeriod: undefined, maxMetadataLen: long_1.default.UZERO };
}
exports.Module = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.maxExecutionPeriod !== undefined) {
            duration_1.Duration.encode(message.maxExecutionPeriod, writer.uint32(10).fork()).ldelim();
        }
        if (!message.maxMetadataLen.isZero()) {
            writer.uint32(16).uint64(message.maxMetadataLen);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseModule();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.maxExecutionPeriod = duration_1.Duration.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.maxMetadataLen = reader.uint64();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            maxExecutionPeriod: isSet(object.maxExecutionPeriod)
                ? duration_1.Duration.fromJSON(object.maxExecutionPeriod)
                : undefined,
            maxMetadataLen: isSet(object.maxMetadataLen) ? long_1.default.fromValue(object.maxMetadataLen) : long_1.default.UZERO,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.maxExecutionPeriod !== undefined) {
            obj.maxExecutionPeriod = duration_1.Duration.toJSON(message.maxExecutionPeriod);
        }
        if (!message.maxMetadataLen.isZero()) {
            obj.maxMetadataLen = (message.maxMetadataLen || long_1.default.UZERO).toString();
        }
        return obj;
    },
    create(base) {
        return exports.Module.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseModule();
        message.maxExecutionPeriod =
            object.maxExecutionPeriod !== undefined && object.maxExecutionPeriod !== null
                ? duration_1.Duration.fromPartial(object.maxExecutionPeriod)
                : undefined;
        message.maxMetadataLen =
            object.maxMetadataLen !== undefined && object.maxMetadataLen !== null
                ? long_1.default.fromValue(object.maxMetadataLen)
                : long_1.default.UZERO;
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
