// @initia/initia.js/dist/core/auction/msgs/MsgUpdateAuctionParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateAuctionParams = void 0;
const json_1 = require("../../../util/json");
const AuctionParams_1 = require("../AuctionParams");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/sdk/auction/v1/tx");
class MsgUpdateAuctionParams extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param params params defines the x/auth parameters to update
     */
    constructor(authority, params) {
        super();
        this.authority = authority;
        this.params = params;
    }
    static fromAmino(data) {
        const { value: { authority, params }, } = data;
        return new MsgUpdateAuctionParams(authority, AuctionParams_1.AuctionParams.fromAmino(params));
    }
    toAmino() {
        const { authority, params } = this;
        return {
            type: 'block-sdk/x/auction/MsgUpdateParams',
            value: {
                authority,
                params: params.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, params } = data;
        return new MsgUpdateAuctionParams(authority, AuctionParams_1.AuctionParams.fromData(params));
    }
    toData() {
        const { authority, params } = this;
        return {
            '@type': '/sdk.auction.v1.MsgUpdateParams',
            authority,
            params: params.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateAuctionParams(data.authority, AuctionParams_1.AuctionParams.fromProto(data.params));
    }
    toProto() {
        const { authority, params } = this;
        return tx_1.MsgUpdateParams.fromPartial({
            authority,
            params: params.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/sdk.auction.v1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateAuctionParams.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    }
}
exports.MsgUpdateAuctionParams = MsgUpdateAuctionParams;
