// @initia/initia.js/dist/core/ibchooks/ACL.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ACL = void 0;
const json_1 = require("../../util/json");
const types_1 = require("@initia/initia.proto/initia/ibchooks/v1/types");
class ACL extends json_1.JSONSerializable {
    /**
     * @param address
     * @param allowed
     */
    constructor(address, allowed) {
        super();
        this.address = address;
        this.allowed = allowed;
    }
    static fromAmino(data) {
        const { address, allowed } = data;
        return new ACL(address, allowed);
    }
    toAmino() {
        const { address, allowed } = this;
        return { address, allowed };
    }
    static fromData(data) {
        const { address, allowed } = data;
        return new ACL(address, allowed);
    }
    toData() {
        const { address, allowed } = this;
        return { address, allowed };
    }
    static fromProto(data) {
        return new ACL(data.address, data.allowed);
    }
    toProto() {
        const { address, allowed } = this;
        return types_1.ACL.fromPartial({
            address,
            allowed,
        });
    }
}
exports.ACL = ACL;
