// @initia/initia.js/dist/core/ibchooks/IbcHooksParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IbcHooksParams = void 0;
const json_1 = require("../../util/json");
const types_1 = require("@initia/initia.proto/initia/ibchooks/v1/types");
class IbcHooksParams extends json_1.JSONSerializable {
    /**
     * @param default_allowed
     */
    constructor(default_allowed) {
        super();
        this.default_allowed = default_allowed;
    }
    static fromAmino(data) {
        return new IbcHooksParams(data.value.default_allowed);
    }
    toAmino() {
        return {
            type: 'ibc-hooks/Params',
            value: { default_allowed: this.default_allowed },
        };
    }
    static fromData(data) {
        return new IbcHooksParams(data.default_allowed);
    }
    toData() {
        return {
            '@type': '/initia.ibchooks.v1.Params',
            default_allowed: this.default_allowed,
        };
    }
    static fromProto(data) {
        return new IbcHooksParams(data.defaultAllowed);
    }
    toProto() {
        return types_1.Params.fromPartial({
            defaultAllowed: this.default_allowed,
        });
    }
}
exports.IbcHooksParams = IbcHooksParams;
