// @initia/initia.js/dist/client/lcd/Wallet.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wallet = void 0;
const signing_1 = require("@initia/initia.proto/cosmos/tx/signing/v1beta1/signing");
class Wallet {
    constructor(lcd, key) {
        this.lcd = lcd;
        this.key = key;
        this.accAddress = key.accAddress;
    }
    setAccountAddress(accAddress) {
        this.accAddress = accAddress;
    }
    accountNumberAndSequence() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.lcd.auth.accountInfo(this.accAddress).then(d => {
                return {
                    account_number: d.getAccountNumber(),
                    sequence: d.getSequenceNumber(),
                };
            });
        });
    }
    accountNumber() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.lcd.auth.accountInfo(this.accAddress).then(d => {
                return d.getAccountNumber();
            });
        });
    }
    sequence() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.lcd.auth.accountInfo(this.accAddress).then(d => {
                return d.getSequenceNumber();
            });
        });
    }
    createTx(options) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.lcd.tx.create([
                {
                    address: this.accAddress,
                    sequenceNumber: options.sequence,
                    publicKey: this.key.publicKey,
                },
            ], options);
        });
    }
    createAndSignTx(options) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.lcd.config.chainId) {
                this.lcd.config.chainId = yield this.lcd.tendermint.chainId();
            }
            let accountNumber = options.accountNumber;
            let sequence = options.sequence;
            if (accountNumber === undefined || sequence === undefined) {
                const res = yield this.accountNumberAndSequence();
                if (accountNumber === undefined) {
                    accountNumber = res.account_number;
                }
                if (sequence === undefined) {
                    sequence = res.sequence;
                }
            }
            options.sequence = sequence;
            options.accountNumber = accountNumber;
            const tx = yield this.createTx(options);
            return this.key.signTx(tx, {
                accountNumber,
                sequence,
                chainId: this.lcd.config.chainId,
                signMode: (_a = options.signMode) !== null && _a !== void 0 ? _a : signing_1.SignMode.SIGN_MODE_DIRECT,
            });
        });
    }
}
exports.Wallet = Wallet;
