// @initia/initia.proto/cosmos/group/v1/genesis.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenesisState = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const types_1 = require("./types");
exports.protobufPackage = "cosmos.group.v1";
function createBaseGenesisState() {
    return {
        groupSeq: long_1.default.UZERO,
        groups: [],
        groupMembers: [],
        groupPolicySeq: long_1.default.UZERO,
        groupPolicies: [],
        proposalSeq: long_1.default.UZERO,
        proposals: [],
        votes: [],
    };
}
exports.GenesisState = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (!message.groupSeq.isZero()) {
            writer.uint32(8).uint64(message.groupSeq);
        }
        for (const v of message.groups) {
            types_1.GroupInfo.encode(v, writer.uint32(18).fork()).ldelim();
        }
        for (const v of message.groupMembers) {
            types_1.GroupMember.encode(v, writer.uint32(26).fork()).ldelim();
        }
        if (!message.groupPolicySeq.isZero()) {
            writer.uint32(32).uint64(message.groupPolicySeq);
        }
        for (const v of message.groupPolicies) {
            types_1.GroupPolicyInfo.encode(v, writer.uint32(42).fork()).ldelim();
        }
        if (!message.proposalSeq.isZero()) {
            writer.uint32(48).uint64(message.proposalSeq);
        }
        for (const v of message.proposals) {
            types_1.Proposal.encode(v, writer.uint32(58).fork()).ldelim();
        }
        for (const v of message.votes) {
            types_1.Vote.encode(v, writer.uint32(66).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.groupSeq = reader.uint64();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.groups.push(types_1.GroupInfo.decode(reader, reader.uint32()));
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.groupMembers.push(types_1.GroupMember.decode(reader, reader.uint32()));
                    continue;
                case 4:
                    if (tag !== 32) {
                        break;
                    }
                    message.groupPolicySeq = reader.uint64();
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.groupPolicies.push(types_1.GroupPolicyInfo.decode(reader, reader.uint32()));
                    continue;
                case 6:
                    if (tag !== 48) {
                        break;
                    }
                    message.proposalSeq = reader.uint64();
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.proposals.push(types_1.Proposal.decode(reader, reader.uint32()));
                    continue;
                case 8:
                    if (tag !== 66) {
                        break;
                    }
                    message.votes.push(types_1.Vote.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            groupSeq: isSet(object.groupSeq) ? long_1.default.fromValue(object.groupSeq) : long_1.default.UZERO,
            groups: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.groups)
                ? object.groups.map((e) => types_1.GroupInfo.fromJSON(e))
                : [],
            groupMembers: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.groupMembers)
                ? object.groupMembers.map((e) => types_1.GroupMember.fromJSON(e))
                : [],
            groupPolicySeq: isSet(object.groupPolicySeq) ? long_1.default.fromValue(object.groupPolicySeq) : long_1.default.UZERO,
            groupPolicies: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.groupPolicies)
                ? object.groupPolicies.map((e) => types_1.GroupPolicyInfo.fromJSON(e))
                : [],
            proposalSeq: isSet(object.proposalSeq) ? long_1.default.fromValue(object.proposalSeq) : long_1.default.UZERO,
            proposals: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.proposals)
                ? object.proposals.map((e) => types_1.Proposal.fromJSON(e))
                : [],
            votes: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.votes) ? object.votes.map((e) => types_1.Vote.fromJSON(e)) : [],
        };
    },
    toJSON(message) {
        var _a, _b, _c, _d, _e;
        const obj = {};
        if (!message.groupSeq.isZero()) {
            obj.groupSeq = (message.groupSeq || long_1.default.UZERO).toString();
        }
        if ((_a = message.groups) === null || _a === void 0 ? void 0 : _a.length) {
            obj.groups = message.groups.map((e) => types_1.GroupInfo.toJSON(e));
        }
        if ((_b = message.groupMembers) === null || _b === void 0 ? void 0 : _b.length) {
            obj.groupMembers = message.groupMembers.map((e) => types_1.GroupMember.toJSON(e));
        }
        if (!message.groupPolicySeq.isZero()) {
            obj.groupPolicySeq = (message.groupPolicySeq || long_1.default.UZERO).toString();
        }
        if ((_c = message.groupPolicies) === null || _c === void 0 ? void 0 : _c.length) {
            obj.groupPolicies = message.groupPolicies.map((e) => types_1.GroupPolicyInfo.toJSON(e));
        }
        if (!message.proposalSeq.isZero()) {
            obj.proposalSeq = (message.proposalSeq || long_1.default.UZERO).toString();
        }
        if ((_d = message.proposals) === null || _d === void 0 ? void 0 : _d.length) {
            obj.proposals = message.proposals.map((e) => types_1.Proposal.toJSON(e));
        }
        if ((_e = message.votes) === null || _e === void 0 ? void 0 : _e.length) {
            obj.votes = message.votes.map((e) => types_1.Vote.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.GenesisState.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e;
        const message = createBaseGenesisState();
        message.groupSeq =
            object.groupSeq !== undefined && object.groupSeq !== null
                ? long_1.default.fromValue(object.groupSeq)
                : long_1.default.UZERO;
        message.groups = ((_a = object.groups) === null || _a === void 0 ? void 0 : _a.map((e) => types_1.GroupInfo.fromPartial(e))) || [];
        message.groupMembers = ((_b = object.groupMembers) === null || _b === void 0 ? void 0 : _b.map((e) => types_1.GroupMember.fromPartial(e))) || [];
        message.groupPolicySeq =
            object.groupPolicySeq !== undefined && object.groupPolicySeq !== null
                ? long_1.default.fromValue(object.groupPolicySeq)
                : long_1.default.UZERO;
        message.groupPolicies = ((_c = object.groupPolicies) === null || _c === void 0 ? void 0 : _c.map((e) => types_1.GroupPolicyInfo.fromPartial(e))) || [];
        message.proposalSeq =
            object.proposalSeq !== undefined && object.proposalSeq !== null
                ? long_1.default.fromValue(object.proposalSeq)
                : long_1.default.UZERO;
        message.proposals = ((_d = object.proposals) === null || _d === void 0 ? void 0 : _d.map((e) => types_1.Proposal.fromPartial(e))) || [];
        message.votes = ((_e = object.votes) === null || _e === void 0 ? void 0 : _e.map((e) => types_1.Vote.fromPartial(e))) || [];
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
