// @initia/initia.js/dist/core/move/msgs/MsgDelist.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgDelist = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/initia/move/v1/tx");
class MsgDelist extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param metadata_lp Dex coin LP metadata address
     */
    constructor(authority, metadata_lp) {
        super();
        this.authority = authority;
        this.metadata_lp = metadata_lp;
    }
    static fromAmino(data) {
        const { value: { authority, metadata_lp }, } = data;
        return new MsgDelist(authority, metadata_lp);
    }
    toAmino() {
        const { authority, metadata_lp } = this;
        return {
            type: 'move/MsgDelist',
            value: {
                authority,
                metadata_lp,
            },
        };
    }
    static fromData(data) {
        const { authority, metadata_lp } = data;
        return new MsgDelist(authority, metadata_lp);
    }
    toData() {
        const { authority, metadata_lp } = this;
        return {
            '@type': '/initia.move.v1.MsgDelist',
            authority,
            metadata_lp,
        };
    }
    static fromProto(data) {
        return new MsgDelist(data.authority, data.metadataLp);
    }
    toProto() {
        const { authority, metadata_lp } = this;
        return tx_1.MsgDelist.fromPartial({
            authority,
            metadataLp: metadata_lp,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/initia.move.v1.MsgDelist',
            value: tx_1.MsgDelist.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgDelist.fromProto(tx_1.MsgDelist.decode(msgAny.value));
    }
}
exports.MsgDelist = MsgDelist;
