// @initia/initia.js/dist/core/move/msgs/MsgExecuteJSON.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgExecuteJSON = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/initia/move/v1/tx");
class MsgExecuteJSON extends json_1.JSONSerializable {
    /**
     * @param sender the actor that signed the messages
     * @param module_address the address of the module deployer
     * @param module_name the name of module to execute
     * @param function_name the name of function to execute
     * @param type_args the type arguments of a function to execute
     * @param args the arguments of a function to execute in json stringify format
     */
    constructor(sender, module_address, module_name, function_name, type_args = [], args = []) {
        super();
        this.sender = sender;
        this.module_address = module_address;
        this.module_name = module_name;
        this.function_name = function_name;
        this.type_args = type_args;
        this.args = args;
    }
    static fromAmino(data) {
        const { value: { sender, module_address, module_name, function_name, type_args, args, }, } = data;
        return new MsgExecuteJSON(sender, module_address, module_name, function_name, type_args !== null && type_args !== void 0 ? type_args : [], args !== null && args !== void 0 ? args : []);
    }
    toAmino() {
        const { sender, module_address, module_name, function_name, type_args, args, } = this;
        return {
            type: 'move/MsgExecuteJSON',
            value: {
                sender,
                module_address,
                module_name,
                function_name,
                type_args: type_args.length === 0 ? undefined : type_args,
                args: args.length === 0 ? undefined : args,
            },
        };
    }
    static fromData(data) {
        const { sender, module_address, module_name, function_name, type_args, args, } = data;
        return new MsgExecuteJSON(sender, module_address, module_name, function_name, type_args, args);
    }
    toData() {
        const { sender, module_address, module_name, function_name, type_args, args, } = this;
        return {
            '@type': '/initia.move.v1.MsgExecuteJSON',
            sender,
            module_address,
            module_name,
            function_name,
            type_args,
            args,
        };
    }
    static fromProto(data) {
        return new MsgExecuteJSON(data.sender, data.moduleAddress, data.moduleName, data.functionName, data.typeArgs, data.args);
    }
    toProto() {
        const { sender, module_address, module_name, function_name, type_args, args, } = this;
        return tx_1.MsgExecuteJSON.fromPartial({
            sender,
            moduleAddress: module_address,
            moduleName: module_name,
            functionName: function_name,
            typeArgs: type_args,
            args,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/initia.move.v1.MsgExecuteJSON',
            value: tx_1.MsgExecuteJSON.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgExecuteJSON.fromProto(tx_1.MsgExecuteJSON.decode(msgAny.value));
    }
}
exports.MsgExecuteJSON = MsgExecuteJSON;
