// @initia/initia.js/dist/core/move/msgs/MsgGovScriptJSON.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgGovScriptJSON = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/initia/move/v1/tx");
class MsgGovScriptJSON extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param sender the actor that signed the messages
     * @param code_bytes the script bytes code to execute
     * @param type_args type arguments of function to execute
     * @param args arguments of function to execute
     */
    constructor(authority, sender, code_bytes, type_args = [], args = []) {
        super();
        this.authority = authority;
        this.sender = sender;
        this.code_bytes = code_bytes;
        this.type_args = type_args;
        this.args = args;
    }
    static fromAmino(data) {
        const { value: { authority, sender, code_bytes, type_args, args }, } = data;
        return new MsgGovScriptJSON(authority, sender, code_bytes, type_args !== null && type_args !== void 0 ? type_args : [], args !== null && args !== void 0 ? args : []);
    }
    toAmino() {
        const { authority, sender, code_bytes, type_args, args } = this;
        return {
            type: 'move/MsgGovScriptJSON',
            value: {
                authority,
                sender,
                code_bytes,
                type_args: type_args.length === 0 ? undefined : type_args,
                args: args.length === 0 ? undefined : args,
            },
        };
    }
    static fromData(data) {
        const { authority, sender, code_bytes, type_args, args } = data;
        return new MsgGovScriptJSON(authority, sender, code_bytes, type_args, args);
    }
    toData() {
        const { authority, sender, code_bytes, type_args, args } = this;
        return {
            '@type': '/initia.move.v1.MsgGovScriptJSON',
            authority,
            sender,
            code_bytes,
            type_args,
            args,
        };
    }
    static fromProto(data) {
        return new MsgGovScriptJSON(data.authority, data.sender, Buffer.from(data.codeBytes).toString('base64'), data.typeArgs, data.args);
    }
    toProto() {
        const { authority, sender, code_bytes, type_args, args } = this;
        return tx_1.MsgGovScriptJSON.fromPartial({
            authority,
            sender,
            codeBytes: Buffer.from(code_bytes, 'base64'),
            typeArgs: type_args,
            args,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/initia.move.v1.MsgGovScriptJSON',
            value: tx_1.MsgGovScriptJSON.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgGovScriptJSON.fromProto(tx_1.MsgGovScriptJSON.decode(msgAny.value));
    }
}
exports.MsgGovScriptJSON = MsgGovScriptJSON;
