// @initia/initia.js/dist/core/move/msgs/MsgGovPublish.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgGovPublish = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/initia/move/v1/tx");
const types_1 = require("@initia/initia.proto/initia/move/v1/types");
class MsgGovPublish extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param sender the actor that signed the messages
     * @param code_bytes raw move module bytes code
     * @param upgrade_policy arbitrary `0`, compatible `1`, immutable `2`
     */
    constructor(authority, sender, code_bytes, upgrade_policy) {
        super();
        this.authority = authority;
        this.sender = sender;
        this.code_bytes = code_bytes;
        this.upgrade_policy = upgrade_policy;
    }
    static fromAmino(data) {
        const { value: { authority, sender, code_bytes, upgrade_policy }, } = data;
        return new MsgGovPublish(authority, sender, code_bytes, upgrade_policy ? upgrade_policy : 0);
    }
    toAmino() {
        const { authority, sender, code_bytes, upgrade_policy } = this;
        return {
            type: 'move/MsgGovPublish',
            value: {
                authority,
                sender,
                code_bytes,
                upgrade_policy: upgrade_policy === 0 ? undefined : upgrade_policy,
            },
        };
    }
    static fromData(data) {
        const { authority, sender, code_bytes, upgrade_policy } = data;
        return new MsgGovPublish(authority, sender, code_bytes, upgrade_policy);
    }
    toData() {
        const { authority, sender, code_bytes, upgrade_policy } = this;
        return {
            '@type': '/initia.move.v1.MsgGovPublish',
            authority,
            sender,
            code_bytes,
            upgrade_policy,
        };
    }
    static fromProto(proto) {
        return new MsgGovPublish(proto.authority, proto.sender, proto.codeBytes.map(code => Buffer.from(code).toString('base64')), proto.upgradePolicy);
    }
    toProto() {
        const { authority, sender, code_bytes, upgrade_policy } = this;
        return tx_1.MsgGovPublish.fromPartial({
            authority,
            sender,
            codeBytes: code_bytes.map(code => Buffer.from(code, 'base64')),
            upgradePolicy: upgrade_policy,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/initia.move.v1.MsgGovPublish',
            value: tx_1.MsgGovPublish.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgGovPublish.fromProto(tx_1.MsgGovPublish.decode(msgAny.value));
    }
}
exports.MsgGovPublish = MsgGovPublish;
(function (MsgGovPublish) {
    MsgGovPublish.Policy = types_1.UpgradePolicy;
})(MsgGovPublish || (exports.MsgGovPublish = MsgGovPublish = {}));
