// @initia/initia.js/dist/core/move/msgs/MsgWhitelist.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgWhitelist = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/initia/move/v1/tx");
class MsgWhitelist extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param metadata_lp Dex coin LP metadata address
     * @param reward_weight registered to distribution's Params
     */
    constructor(authority, metadata_lp, reward_weight) {
        super();
        this.authority = authority;
        this.metadata_lp = metadata_lp;
        this.reward_weight = reward_weight;
    }
    static fromAmino(data) {
        const { value: { authority, metadata_lp, reward_weight }, } = data;
        return new MsgWhitelist(authority, metadata_lp, reward_weight);
    }
    toAmino() {
        const { authority, metadata_lp, reward_weight } = this;
        return {
            type: 'move/MsgWhitelist',
            value: {
                authority,
                metadata_lp,
                reward_weight,
            },
        };
    }
    static fromData(data) {
        const { authority, metadata_lp, reward_weight } = data;
        return new MsgWhitelist(authority, metadata_lp, reward_weight);
    }
    toData() {
        const { authority, metadata_lp, reward_weight } = this;
        return {
            '@type': '/initia.move.v1.MsgWhitelist',
            authority,
            metadata_lp,
            reward_weight,
        };
    }
    static fromProto(data) {
        return new MsgWhitelist(data.authority, data.metadataLp, data.rewardWeight);
    }
    toProto() {
        const { authority, metadata_lp, reward_weight } = this;
        return tx_1.MsgWhitelist.fromPartial({
            authority,
            metadataLp: metadata_lp,
            rewardWeight: reward_weight,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/initia.move.v1.MsgWhitelist',
            value: tx_1.MsgWhitelist.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgWhitelist.fromProto(tx_1.MsgWhitelist.decode(msgAny.value));
    }
}
exports.MsgWhitelist = MsgWhitelist;
