// @initia/initia.js/dist/core/ibc/applications/nft-transfer/msgs/MsgUpdateIbcNftParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateIbcNftParams = void 0;
const json_1 = require("../../../../../util/json");
const IbcNftParams_1 = require("../IbcNftParams");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/applications/nft_transfer/v1/tx");
class MsgUpdateIbcNftParams extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param params params defines the parameters to update
     */
    constructor(authority, params) {
        super();
        this.authority = authority;
        this.params = params;
    }
    static fromAmino(data) {
        const { value: { authority, params }, } = data;
        return new MsgUpdateIbcNftParams(authority, IbcNftParams_1.IbcNftParams.fromAmino(params));
    }
    toAmino() {
        const { authority, params } = this;
        return {
            type: 'nft-transfer/MsgUpdateParams',
            value: {
                authority,
                params: params.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, params } = data;
        return new MsgUpdateIbcNftParams(authority, IbcNftParams_1.IbcNftParams.fromData(params));
    }
    toData() {
        const { authority, params } = this;
        return {
            '@type': '/ibc.applications.nft_transfer.v1.MsgUpdateParams',
            authority,
            params: params.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateIbcNftParams(data.authority, IbcNftParams_1.IbcNftParams.fromProto(data.params));
    }
    toProto() {
        const { authority, params } = this;
        return tx_1.MsgUpdateParams.fromPartial({
            authority,
            params: params.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.applications.nft_transfer.v1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateIbcNftParams.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    }
}
exports.MsgUpdateIbcNftParams = MsgUpdateIbcNftParams;
