// @initia/initia.js/dist/core/feegrant/allowances/BasicAllowance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasicAllowance = void 0;
const json_1 = require("../../../util/json");
const Coins_1 = require("../../Coins");
const num_1 = require("../../num");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const feegrant_1 = require("@initia/initia.proto/cosmos/feegrant/v1beta1/feegrant");
/**
 * BasicAllowance implements Allowance with a one-time grant of tokens
 * that optionally expires. The grantee can use up to SpendLimit to cover fees.
 */
class BasicAllowance extends json_1.JSONSerializable {
    /**
     * @param spend_limit spend_limit allowed to be spent as fee
     * @param expiration allowance's expiration
     */
    constructor(spend_limit, expiration) {
        super();
        this.expiration = expiration;
        let hasNotPositive = false;
        this.spend_limit = new Coins_1.Coins(spend_limit);
        this.spend_limit.map(c => {
            // isPositive() from decimal.js returns true when the amount is 0.
            // but Coins.IsAllPositive() from cosmos-sdk will return false in same case.
            // so we use lessThanorEquenTo(0) instead of isPositive() == false
            if ((0, num_1.num)(c.amount).isLessThanOrEqualTo(0)) {
                hasNotPositive = true;
            }
        });
        if (hasNotPositive) {
            throw new Error('spend_limit must be positive');
        }
    }
    static fromAmino(data) {
        const { value: { spend_limit, expiration }, } = data;
        return new BasicAllowance(Coins_1.Coins.fromAmino(spend_limit), expiration ? new Date(expiration) : undefined);
    }
    toAmino() {
        const { spend_limit, expiration } = this;
        return {
            type: 'cosmos-sdk/BasicAllowance',
            value: {
                spend_limit: spend_limit.toAmino(),
                expiration: expiration === null || expiration === void 0 ? void 0 : expiration.toISOString().replace(/\.000Z$/, 'Z'),
            },
        };
    }
    static fromData(proto) {
        const { spend_limit, expiration } = proto;
        return new BasicAllowance(Coins_1.Coins.fromData(spend_limit), expiration ? new Date(expiration) : undefined);
    }
    toData() {
        const { spend_limit, expiration } = this;
        return {
            '@type': '/cosmos.feegrant.v1beta1.BasicAllowance',
            spend_limit: spend_limit.toData(),
            expiration: expiration === null || expiration === void 0 ? void 0 : expiration.toISOString().replace(/\.000Z$/, 'Z'),
        };
    }
    static fromProto(proto) {
        return new BasicAllowance(Coins_1.Coins.fromProto(proto.spendLimit), proto.expiration);
    }
    toProto() {
        const { spend_limit, expiration } = this;
        return feegrant_1.BasicAllowance.fromPartial({
            expiration,
            spendLimit: spend_limit.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.feegrant.v1beta1.BasicAllowance',
            value: feegrant_1.BasicAllowance.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return BasicAllowance.fromProto(feegrant_1.BasicAllowance.decode(msgAny.value));
    }
}
exports.BasicAllowance = BasicAllowance;
