// @initia/initia.js/dist/core/feegrant/allowances/PeriodicAllowance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PeriodicAllowance = void 0;
const json_1 = require("../../../util/json");
const Coins_1 = require("../../Coins");
const Duration_1 = require("../../Duration");
const BasicAllowance_1 = require("./BasicAllowance");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const feegrant_1 = require("@initia/initia.proto/cosmos/feegrant/v1beta1/feegrant");
/**
 * PeriodicAllowance extends Allowance to allow for both a maximum cap,
 * as well as a limit per time period.
 */
class PeriodicAllowance extends json_1.JSONSerializable {
    /**
     * @param basic basic allowance given per period
     * @param period the time duration in which period_spend_limit coins can be spent before that allowance is reset
     * @param period_spend_limit the maximum number of coins that can be spent in the period
     * @param period_can_spend the number of coins left to be spent before the period_reset time
     * @param period_reset the time at which this period resets and a new one begins
     */
    constructor(basic, period, period_spend_limit, period_can_spend, period_reset) {
        super();
        this.basic = basic;
        this.period = period;
        this.period_reset = period_reset;
        this.period_spend_limit = new Coins_1.Coins(period_spend_limit);
        this.period_can_spend = new Coins_1.Coins(period_can_spend);
    }
    static fromAmino(data) {
        const { value: { basic, period, period_spend_limit, period_can_spend, period_reset, }, } = data;
        return new PeriodicAllowance(BasicAllowance_1.BasicAllowance.fromAmino(basic), Duration_1.Duration.fromAmino(period), Coins_1.Coins.fromAmino(period_spend_limit), Coins_1.Coins.fromAmino(period_can_spend), new Date(period_reset));
    }
    toAmino() {
        const { basic, period, period_spend_limit, period_can_spend, period_reset, } = this;
        return {
            type: 'cosmos-sdk/PeriodicAllowance',
            value: {
                basic: basic.toAmino(),
                period: period.toAmino(),
                period_spend_limit: period_spend_limit.toAmino(),
                period_can_spend: period_can_spend.toAmino(),
                period_reset: period_reset.toISOString().replace(/\.000Z$/, 'Z'),
            },
        };
    }
    static fromData(proto) {
        const { basic, period, period_spend_limit, period_can_spend, period_reset, } = proto;
        return new PeriodicAllowance(BasicAllowance_1.BasicAllowance.fromData(basic), Duration_1.Duration.fromData(period), Coins_1.Coins.fromData(period_spend_limit), Coins_1.Coins.fromData(period_can_spend), new Date(period_reset));
    }
    toData() {
        const { basic, period, period_spend_limit, period_can_spend, period_reset, } = this;
        return {
            '@type': '/cosmos.feegrant.v1beta1.PeriodicAllowance',
            basic: basic.toData(),
            period: period.toData(),
            period_spend_limit: period_spend_limit.toData(),
            period_can_spend: period_can_spend.toData(),
            period_reset: period_reset.toISOString().replace(/\.000Z$/, 'Z'),
        };
    }
    static fromProto(proto) {
        return new PeriodicAllowance(BasicAllowance_1.BasicAllowance.fromProto(proto.basic), Duration_1.Duration.fromProto(proto.period), Coins_1.Coins.fromProto(proto.periodSpendLimit), Coins_1.Coins.fromProto(proto.periodCanSpend), proto.periodReset);
    }
    toProto() {
        const { basic, period, period_spend_limit, period_can_spend, period_reset, } = this;
        return feegrant_1.PeriodicAllowance.fromPartial({
            basic: basic.toProto(),
            period: period.toProto(),
            periodCanSpend: period_can_spend.toProto(),
            periodReset: period_reset,
            periodSpendLimit: period_spend_limit.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.feegrant.v1beta1.PeriodicAllowance',
            value: feegrant_1.PeriodicAllowance.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return PeriodicAllowance.fromProto(feegrant_1.PeriodicAllowance.decode(msgAny.value));
    }
}
exports.PeriodicAllowance = PeriodicAllowance;
