// @initia/initia.js/dist/core/ibc/applications/transfer/msgs/MsgUpdateIbcTransferParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateIbcTransferParams = void 0;
const json_1 = require("../../../../../util/json");
const IbcTransferParams_1 = require("../IbcTransferParams");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/applications/transfer/v1/tx");
class MsgUpdateIbcTransferParams extends json_1.JSONSerializable {
    /**
     * @param signer the address that controls the module
     * @param params params defines the parameters to update
     */
    constructor(signer, params) {
        super();
        this.signer = signer;
        this.params = params;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { signer, params } = data;
        return new MsgUpdateIbcTransferParams(signer, IbcTransferParams_1.IbcTransferParams.fromData(params));
    }
    toData() {
        const { signer, params } = this;
        return {
            '@type': '/ibc.applications.transfer.v1.MsgUpdateParams',
            signer,
            params: params.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateIbcTransferParams(data.signer, IbcTransferParams_1.IbcTransferParams.fromProto(data.params));
    }
    toProto() {
        const { signer, params } = this;
        return tx_1.MsgUpdateParams.fromPartial({
            signer,
            params: params.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.applications.transfer.v1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateIbcTransferParams.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    }
}
exports.MsgUpdateIbcTransferParams = MsgUpdateIbcTransferParams;
