// @initia/initia.js/dist/core/distribution/msgs/MsgCommunityPoolSpend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgCommunityPoolSpend = void 0;
const json_1 = require("../../../util/json");
const Coins_1 = require("../../Coins");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmos/distribution/v1beta1/tx");
/**
 * defines a message for sending tokens from the community pool to another account
 */
class MsgCommunityPoolSpend extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param recipient recipient address
     * @param amount amount to give recipient
     */
    constructor(authority, recipient, amount) {
        super();
        this.authority = authority;
        this.recipient = recipient;
        this.amount = new Coins_1.Coins(amount);
    }
    static fromAmino(data) {
        const { value: { authority, recipient, amount }, } = data;
        return new MsgCommunityPoolSpend(authority, recipient, Coins_1.Coins.fromAmino(amount));
    }
    toAmino() {
        const { authority, recipient, amount } = this;
        return {
            type: 'cosmos-sdk/distr/MsgCommunityPoolSpend',
            value: {
                authority,
                recipient,
                amount: amount.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, recipient, amount } = data;
        return new MsgCommunityPoolSpend(authority, recipient, Coins_1.Coins.fromData(amount));
    }
    toData() {
        const { authority, recipient, amount } = this;
        return {
            '@type': '/cosmos.distribution.v1beta1.MsgCommunityPoolSpend',
            authority,
            recipient,
            amount: amount.toData(),
        };
    }
    static fromProto(proto) {
        return new MsgCommunityPoolSpend(proto.authority, proto.recipient, Coins_1.Coins.fromProto(proto.amount));
    }
    toProto() {
        const { authority, recipient, amount } = this;
        return tx_1.MsgCommunityPoolSpend.fromPartial({
            authority,
            recipient,
            amount: amount.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.distribution.v1beta1.MsgCommunityPoolSpend',
            value: tx_1.MsgCommunityPoolSpend.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgCommunityPoolSpend.fromProto(tx_1.MsgCommunityPoolSpend.decode(msgAny.value));
    }
}
exports.MsgCommunityPoolSpend = MsgCommunityPoolSpend;
