// @initia/initia.js/dist/core/distribution/msgs/MsgUpdateDistrParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateDistrParams = void 0;
const json_1 = require("../../../util/json");
const DistributionParams_1 = require("../DistributionParams");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/initia/distribution/v1/tx");
class MsgUpdateDistrParams extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param params params defines the x/distribution parameters to update
     */
    constructor(authority, params) {
        super();
        this.authority = authority;
        this.params = params;
    }
    static fromAmino(data) {
        const { value: { authority, params }, } = data;
        return new MsgUpdateDistrParams(authority, DistributionParams_1.DistributionParams.fromAmino(params));
    }
    toAmino() {
        const { authority, params } = this;
        return {
            type: 'distribution/MsgUpdateParams',
            value: {
                authority,
                params: params.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, params } = data;
        return new MsgUpdateDistrParams(authority, DistributionParams_1.DistributionParams.fromData(params));
    }
    toData() {
        const { authority, params } = this;
        return {
            '@type': '/initia.distribution.v1.MsgUpdateParams',
            authority,
            params: params.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateDistrParams(data.authority, DistributionParams_1.DistributionParams.fromProto(data.params));
    }
    toProto() {
        const { authority, params } = this;
        return tx_1.MsgUpdateParams.fromPartial({
            authority,
            params: params.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/initia.distribution.v1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateDistrParams.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    }
}
exports.MsgUpdateDistrParams = MsgUpdateDistrParams;
