// @initia/initia.js/dist/core/distribution/msgs/MsgWithdrawValidatorCommission.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgWithdrawValidatorCommission = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmos/distribution/v1beta1/tx");
/**
 * A validator can withdraw their outstanding commission rewards accrued from all
 * delegations (not including its self-delegation) into their associated account's
 * withdraw address.
 */
class MsgWithdrawValidatorCommission extends json_1.JSONSerializable {
    /**
     * @param validator_address validator's operator address
     */
    constructor(validator_address) {
        super();
        this.validator_address = validator_address;
    }
    static fromAmino(data) {
        const { value: { validator_address }, } = data;
        return new MsgWithdrawValidatorCommission(validator_address);
    }
    toAmino() {
        const { validator_address } = this;
        return {
            type: 'cosmos-sdk/MsgWithdrawValidatorCommission',
            value: {
                validator_address,
            },
        };
    }
    static fromData(proto) {
        const { validator_address } = proto;
        return new MsgWithdrawValidatorCommission(validator_address);
    }
    toData() {
        const { validator_address } = this;
        return {
            '@type': '/cosmos.distribution.v1beta1.MsgWithdrawValidatorCommission',
            validator_address,
        };
    }
    static fromProto(proto) {
        return new MsgWithdrawValidatorCommission(proto.validatorAddress);
    }
    toProto() {
        const { validator_address } = this;
        return tx_1.MsgWithdrawValidatorCommission.fromPartial({
            validatorAddress: validator_address,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.distribution.v1beta1.MsgWithdrawValidatorCommission',
            value: tx_1.MsgWithdrawValidatorCommission.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgWithdrawValidatorCommission.fromProto(tx_1.MsgWithdrawValidatorCommission.decode(msgAny.value));
    }
}
exports.MsgWithdrawValidatorCommission = MsgWithdrawValidatorCommission;
