// @initia/initia.js/dist/core/gov/msgs/MsgCancelProposal.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgCancelProposal = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmos/gov/v1/tx");
const long_1 = __importDefault(require("long"));
class MsgCancelProposal extends json_1.JSONSerializable {
    /**
     * @param proposal_id the unique id of the proposal
     * @param proposer the account address of the proposer
     */
    constructor(proposal_id, proposer) {
        super();
        this.proposal_id = proposal_id;
        this.proposer = proposer;
    }
    static fromAmino(data) {
        const { value: { proposal_id, proposer }, } = data;
        return new MsgCancelProposal(Number.parseInt(proposal_id), proposer);
    }
    toAmino() {
        const { proposal_id, proposer } = this;
        return {
            type: 'cosmos-sdk/v1/MsgCancelProposal',
            value: {
                proposal_id: proposal_id.toString(),
                proposer,
            },
        };
    }
    static fromData(data) {
        const { proposal_id, proposer } = data;
        return new MsgCancelProposal(Number.parseInt(proposal_id), proposer);
    }
    toData() {
        const { proposal_id, proposer } = this;
        return {
            '@type': '/cosmos.gov.v1.MsgCancelProposal',
            proposal_id: proposal_id.toString(),
            proposer,
        };
    }
    static fromProto(proto) {
        return new MsgCancelProposal(proto.proposalId.toNumber(), proto.proposer);
    }
    toProto() {
        const { proposal_id, proposer } = this;
        return tx_1.MsgCancelProposal.fromPartial({
            proposalId: long_1.default.fromNumber(proposal_id),
            proposer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.gov.v1.MsgCancelProposal',
            value: tx_1.MsgCancelProposal.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgCancelProposal.fromProto(tx_1.MsgCancelProposal.decode(msgAny.value));
    }
}
exports.MsgCancelProposal = MsgCancelProposal;
