// @initia/initia.js/dist/core/gov/msgs/MsgSubmitProposalLegacy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgSubmitProposalLegacy = void 0;
const Coins_1 = require("../../Coins");
const proposals_1 = require("../proposals");
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmos/gov/v1beta1/tx");
/**
 * Submit a proposal alongside an initial deposit.
 */
class MsgSubmitProposalLegacy extends json_1.JSONSerializable {
    /**
     * @param content proposal content to submit
     * @param initial_deposit deposit provided
     * @param proposer proposer's account address
     */
    constructor(content, initial_deposit, proposer) {
        super();
        this.content = content;
        this.proposer = proposer;
        this.initial_deposit = new Coins_1.Coins(initial_deposit);
    }
    static fromAmino(data) {
        const { value: { content, initial_deposit, proposer }, } = data;
        return new MsgSubmitProposalLegacy(proposals_1.Content.fromAmino(content), Coins_1.Coins.fromAmino(initial_deposit), proposer);
    }
    toAmino() {
        const { content, initial_deposit, proposer } = this;
        return {
            type: 'cosmos-sdk/MsgSubmitProposal',
            value: {
                content: content.toAmino(),
                initial_deposit: initial_deposit.toAmino(),
                proposer,
            },
        };
    }
    static fromData(data) {
        const { content, initial_deposit, proposer } = data;
        return new MsgSubmitProposalLegacy(proposals_1.Content.fromData(content), Coins_1.Coins.fromData(initial_deposit), proposer);
    }
    toData() {
        const { content, initial_deposit, proposer } = this;
        return {
            '@type': '/cosmos.gov.v1beta1.MsgSubmitProposal',
            content: content.toData(),
            initial_deposit: initial_deposit.toData(),
            proposer,
        };
    }
    static fromProto(proto) {
        return new MsgSubmitProposalLegacy(proposals_1.Content.fromProto(proto.content), Coins_1.Coins.fromProto(proto.initialDeposit), proto.proposer);
    }
    toProto() {
        const { content, initial_deposit, proposer } = this;
        return tx_1.MsgSubmitProposal.fromPartial({
            content: content.packAny(),
            initialDeposit: initial_deposit.toProto(),
            proposer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.gov.v1beta1.MsgSubmitProposal',
            value: tx_1.MsgSubmitProposal.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgSubmitProposalLegacy.fromProto(tx_1.MsgSubmitProposal.decode(msgAny.value));
    }
}
exports.MsgSubmitProposalLegacy = MsgSubmitProposalLegacy;
