// @initia/initia.js/dist/core/gov/msgs/MsgUpdateGovParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateGovParams = void 0;
const json_1 = require("../../../util/json");
const GovParams_1 = require("../GovParams");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/initia/gov/v1/tx");
class MsgUpdateGovParams extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param params params defines the x/auth parameters to update
     */
    constructor(authority, params) {
        super();
        this.authority = authority;
        this.params = params;
    }
    static fromAmino(data) {
        const { value: { authority, params }, } = data;
        return new MsgUpdateGovParams(authority, GovParams_1.GovParams.fromAmino(params));
    }
    toAmino() {
        const { authority, params } = this;
        return {
            type: 'gov/MsgUpdateParams',
            value: {
                authority,
                params: params.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, params } = data;
        return new MsgUpdateGovParams(authority, GovParams_1.GovParams.fromData(params));
    }
    toData() {
        const { authority, params } = this;
        return {
            '@type': '/initia.gov.v1.MsgUpdateParams',
            authority,
            params: params.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateGovParams(data.authority, GovParams_1.GovParams.fromProto(data.params));
    }
    toProto() {
        const { authority, params } = this;
        return tx_1.MsgUpdateParams.fromPartial({
            authority,
            params: params.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/initia.gov.v1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateGovParams.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    }
}
exports.MsgUpdateGovParams = MsgUpdateGovParams;
