// @initia/initia.js/dist/core/gov/msgs/MsgVoteLegacy.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgVoteLegacy = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmos/gov/v1beta1/tx");
const gov_1 = require("@initia/initia.proto/cosmos/gov/v1beta1/gov");
const long_1 = __importDefault(require("long"));
/**
 * Vote for a proposal
 */
class MsgVoteLegacy extends json_1.JSONSerializable {
    /**
     * @param proposal_id ID of proposal to vote on
     * @param voter voter's account address
     * @param option one of voting options
     */
    constructor(proposal_id, voter, option) {
        super();
        this.proposal_id = proposal_id;
        this.voter = voter;
        this.option = option;
    }
    static fromAmino(data) {
        const { value: { proposal_id, voter, option }, } = data;
        return new MsgVoteLegacy(Number.parseInt(proposal_id), voter, option);
    }
    toAmino() {
        const { proposal_id, voter, option } = this;
        return {
            type: 'cosmos-sdk/MsgVote',
            value: {
                proposal_id: proposal_id.toFixed(),
                voter,
                option,
            },
        };
    }
    static fromData(data) {
        const { proposal_id, voter, option } = data;
        return new MsgVoteLegacy(Number.parseInt(proposal_id), voter, option);
    }
    toData() {
        const { proposal_id, voter, option } = this;
        return {
            '@type': '/cosmos.gov.v1beta1.MsgVote',
            proposal_id: proposal_id.toFixed(),
            voter,
            option,
        };
    }
    static fromProto(proto) {
        return new MsgVoteLegacy(proto.proposalId.toNumber(), proto.voter, proto.option);
    }
    toProto() {
        const { proposal_id, voter, option } = this;
        return tx_1.MsgVote.fromPartial({
            option,
            proposalId: long_1.default.fromNumber(proposal_id),
            voter,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.gov.v1beta1.MsgVote',
            value: tx_1.MsgVote.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgVoteLegacy.fromProto(tx_1.MsgVote.decode(msgAny.value));
    }
}
exports.MsgVoteLegacy = MsgVoteLegacy;
(function (MsgVoteLegacy) {
    MsgVoteLegacy.Option = gov_1.VoteOption;
})(MsgVoteLegacy || (exports.MsgVoteLegacy = MsgVoteLegacy = {}));
