// @initia/initia.js/dist/core/gov/msgs/MsgVote.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgVote = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmos/gov/v1/tx");
const gov_1 = require("@initia/initia.proto/cosmos/gov/v1/gov");
const long_1 = __importDefault(require("long"));
/**
 * Defines a message to cast a vote
 */
class MsgVote extends json_1.JSONSerializable {
    /**
     * @param proposal_id the unique id of the proposal
     * @param voter the voter address for the proposal
     * @param option the vote option
     * @param metadata any arbitrary metadata attached to the Vote
     */
    constructor(proposal_id, voter, option, metadata) {
        super();
        this.proposal_id = proposal_id;
        this.voter = voter;
        this.option = option;
        this.metadata = metadata;
    }
    static fromAmino(data) {
        const { value: { proposal_id, voter, option, metadata }, } = data;
        return new MsgVote(Number.parseInt(proposal_id), voter, option, metadata);
    }
    toAmino() {
        const { proposal_id, voter, option, metadata } = this;
        return {
            type: 'cosmos-sdk/v1/MsgVote',
            value: {
                proposal_id: proposal_id.toString(),
                voter,
                option,
                metadata,
            },
        };
    }
    static fromData(data) {
        const { proposal_id, voter, option, metadata } = data;
        return new MsgVote(Number.parseInt(proposal_id), voter, option, metadata);
    }
    toData() {
        const { proposal_id, voter, option, metadata } = this;
        return {
            '@type': '/cosmos.gov.v1.MsgVote',
            proposal_id: proposal_id.toString(),
            voter,
            option,
            metadata,
        };
    }
    static fromProto(data) {
        return new MsgVote(data.proposalId.toNumber(), data.voter, data.option, data.metadata);
    }
    toProto() {
        const { proposal_id, voter, option, metadata } = this;
        return tx_1.MsgVote.fromPartial({
            proposalId: long_1.default.fromNumber(proposal_id),
            voter,
            option,
            metadata,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.gov.v1.MsgVote',
            value: tx_1.MsgVote.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgVote.fromProto(tx_1.MsgVote.decode(msgAny.value));
    }
}
exports.MsgVote = MsgVote;
(function (MsgVote) {
    MsgVote.Option = gov_1.VoteOption;
})(MsgVote || (exports.MsgVote = MsgVote = {}));
