// @initia/initia.js/dist/core/gov/msgs/MsgVoteWeightedLegacy.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgVoteWeightedLegacy = void 0;
const json_1 = require("../../../util/json");
const Vote_1 = require("../Vote");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmos/gov/v1beta1/tx");
const long_1 = __importDefault(require("long"));
/**
 * Weighted vote for a proposal
 */
class MsgVoteWeightedLegacy extends json_1.JSONSerializable {
    /**
     * @param proposal_id ID of proposal to vote on
     * @param voter voter's account address
     * @param option one of voting options
     */
    constructor(proposal_id, voter, options) {
        super();
        this.proposal_id = proposal_id;
        this.voter = voter;
        this.options = options;
    }
    static fromAmino(data) {
        const { value: { proposal_id, voter, options }, } = data;
        return new MsgVoteWeightedLegacy(Number.parseInt(proposal_id), voter, options.map(o => Vote_1.WeightedVoteOption.fromAmino(o)));
    }
    toAmino() {
        const { proposal_id, voter, options } = this;
        return {
            type: 'cosmos-sdk/MsgVoteWeighted',
            value: {
                proposal_id: proposal_id.toFixed(),
                voter,
                options: options.map(o => o.toAmino()),
            },
        };
    }
    static fromData(data) {
        const { proposal_id, voter, options } = data;
        return new MsgVoteWeightedLegacy(Number.parseInt(proposal_id), voter, options.map(o => Vote_1.WeightedVoteOption.fromData(o)));
    }
    toData() {
        const { proposal_id, voter, options } = this;
        return {
            '@type': '/cosmos.gov.v1beta1.MsgVoteWeighted',
            proposal_id: proposal_id.toFixed(),
            voter,
            options: options.map(o => o.toData()),
        };
    }
    static fromProto(proto) {
        return new MsgVoteWeightedLegacy(proto.proposalId.toNumber(), proto.voter, proto.options.map(o => Vote_1.WeightedVoteOption.fromProto(o)));
    }
    toProto() {
        const { proposal_id, voter, options } = this;
        return tx_1.MsgVoteWeighted.fromPartial({
            options: options.map(o => o.toProto()),
            proposalId: long_1.default.fromNumber(proposal_id),
            voter,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.gov.v1beta1.MsgVoteWeighted',
            value: tx_1.MsgVoteWeighted.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgVoteWeightedLegacy.fromProto(tx_1.MsgVoteWeighted.decode(msgAny.value));
    }
}
exports.MsgVoteWeightedLegacy = MsgVoteWeightedLegacy;
