// @initia/initia.js/dist/core/ibchooks/msgs/MsgUpdateIbcHooksParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateIbcHooksParams = void 0;
const json_1 = require("../../../util/json");
const IbcHooksParams_1 = require("../IbcHooksParams");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/initia/ibchooks/v1/tx");
class MsgUpdateIbcHooksParams extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param params params defines the x/hook parameters to update
     */
    constructor(authority, params) {
        super();
        this.authority = authority;
        this.params = params;
    }
    static fromAmino(data) {
        const { value: { authority, params }, } = data;
        return new MsgUpdateIbcHooksParams(authority, IbcHooksParams_1.IbcHooksParams.fromAmino(params));
    }
    toAmino() {
        const { authority, params } = this;
        return {
            type: 'ibchooks/MsgUpdateParams',
            value: {
                authority,
                params: params.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, params } = data;
        return new MsgUpdateIbcHooksParams(authority, IbcHooksParams_1.IbcHooksParams.fromData(params));
    }
    toData() {
        const { authority, params } = this;
        return {
            '@type': '/initia.ibchooks.v1.MsgUpdateParams',
            authority,
            params: params.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateIbcHooksParams(data.authority, IbcHooksParams_1.IbcHooksParams.fromProto(data.params));
    }
    toProto() {
        const { authority, params } = this;
        return tx_1.MsgUpdateParams.fromPartial({
            authority,
            params: params.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/initia.ibchooks.v1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateIbcHooksParams.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    }
}
exports.MsgUpdateIbcHooksParams = MsgUpdateIbcHooksParams;
