// @initia/initia.js/dist/core/ibc/applications/fee/msgs/MsgPayPacketFee.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgPayPacketFee = void 0;
const json_1 = require("../../../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/applications/fee/v1/tx");
const IbcFee_1 = require("../IbcFee");
/**
 * MsgPayPacketFee defines the request type for the PayPacketFee rpc
 * This Msg can be used to pay for a packet at the next sequence send & should be combined with the Msg that will be paid for
 */
class MsgPayPacketFee extends json_1.JSONSerializable {
    /**
     * @param fee encapsulates the recv, ack and timeout fees associated with an IBC packet
     * @param source_port_id the source port unique identifier
     * @param source_channel_id the source channel unique identifer
     * @param signer account address to refund fee if necessary
     * @param relayers optional list of relayers permitted to the receive packet fees
     */
    constructor(fee, source_port_id, source_channel_id, signer, relayers) {
        super();
        this.fee = fee;
        this.source_port_id = source_port_id;
        this.source_channel_id = source_channel_id;
        this.signer = signer;
        this.relayers = relayers;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { fee, source_port_id, source_channel_id, signer, relayers } = data;
        return new MsgPayPacketFee(fee ? IbcFee_1.IbcFee.fromData(fee) : undefined, source_port_id, source_channel_id, signer, relayers);
    }
    toData() {
        const { fee, source_port_id, source_channel_id, signer, relayers } = this;
        return {
            '@type': '/ibc.applications.fee.v1.MsgPayPacketFee',
            fee: fee === null || fee === void 0 ? void 0 : fee.toData(),
            source_port_id,
            source_channel_id,
            signer,
            relayers,
        };
    }
    static fromProto(proto) {
        return new MsgPayPacketFee(proto.fee ? IbcFee_1.IbcFee.fromProto(proto.fee) : undefined, proto.sourcePortId, proto.sourceChannelId, proto.signer, proto.relayers);
    }
    toProto() {
        const { fee, source_port_id, source_channel_id, signer, relayers } = this;
        return tx_1.MsgPayPacketFee.fromPartial({
            fee: fee === null || fee === void 0 ? void 0 : fee.toProto(),
            sourcePortId: source_port_id,
            sourceChannelId: source_channel_id,
            signer,
            relayers,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.applications.fee.v1.MsgPayPacketFee',
            value: tx_1.MsgPayPacketFee.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgPayPacketFee.fromProto(tx_1.MsgPayPacketFee.decode(msgAny.value));
    }
}
exports.MsgPayPacketFee = MsgPayPacketFee;
