// @initia/initia.js/dist/core/tokenfactory/msgs/MsgBurn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgBurn = void 0;
const json_1 = require("../../../util/json");
const Coin_1 = require("../../Coin");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/miniwasm/tokenfactory/v1/tx");
class MsgBurn extends json_1.JSONSerializable {
    /**
     * @param sender
     * @param amount
     * @param burn_from_address
     */
    constructor(sender, amount, burn_from_address) {
        super();
        this.sender = sender;
        this.amount = amount;
        this.burn_from_address = burn_from_address;
    }
    static fromAmino(data) {
        const { value: { sender, amount, burn_from_address }, } = data;
        return new MsgBurn(sender, Coin_1.Coin.fromAmino(amount), burn_from_address);
    }
    toAmino() {
        const { sender, amount, burn_from_address } = this;
        return {
            type: 'tokenfactory/MsgBurn',
            value: {
                sender,
                amount: amount.toAmino(),
                burn_from_address,
            },
        };
    }
    static fromData(data) {
        const { sender, amount, burn_from_address } = data;
        return new MsgBurn(sender, Coin_1.Coin.fromData(amount), burn_from_address);
    }
    toData() {
        const { sender, amount, burn_from_address } = this;
        return {
            '@type': '/miniwasm.tokenfactory.v1.MsgBurn',
            sender,
            amount: amount.toData(),
            burn_from_address,
        };
    }
    static fromProto(data) {
        return new MsgBurn(data.sender, Coin_1.Coin.fromProto(data.amount), data.burnFromAddress);
    }
    toProto() {
        const { sender, amount, burn_from_address } = this;
        return tx_1.MsgBurn.fromPartial({
            sender,
            amount: amount.toProto(),
            burnFromAddress: burn_from_address,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/miniwasm.tokenfactory.v1.MsgBurn',
            value: tx_1.MsgBurn.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgBurn.fromProto(tx_1.MsgBurn.decode(msgAny.value));
    }
}
exports.MsgBurn = MsgBurn;
