// @initia/initia.js/dist/core/tokenfactory/msgs/MsgUpdateTokenfactoryParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateTokenfactoryParams = void 0;
const json_1 = require("../../../util/json");
const TokenfactoryParams_1 = require("../TokenfactoryParams");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/miniwasm/tokenfactory/v1/tx");
class MsgUpdateTokenfactoryParams extends json_1.JSONSerializable {
    /**
     * @param authority the address that controls the module
     * @param params params defines the x/tokenfactory parameters to update
     */
    constructor(authority, params) {
        super();
        this.authority = authority;
        this.params = params;
    }
    static fromAmino(data) {
        const { value: { authority, params }, } = data;
        return new MsgUpdateTokenfactoryParams(authority, TokenfactoryParams_1.TokenfactoryParams.fromAmino(params));
    }
    toAmino() {
        const { authority, params } = this;
        return {
            type: 'tokenfactory/MsgUpdateParams',
            value: {
                authority,
                params: params.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { authority, params } = data;
        return new MsgUpdateTokenfactoryParams(authority, TokenfactoryParams_1.TokenfactoryParams.fromData(params));
    }
    toData() {
        const { authority, params } = this;
        return {
            '@type': '/miniwasm.tokenfactory.v1.MsgUpdateParams',
            authority,
            params: params.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateTokenfactoryParams(data.authority, TokenfactoryParams_1.TokenfactoryParams.fromProto(data.params));
    }
    toProto() {
        const { authority, params } = this;
        return tx_1.MsgUpdateParams.fromPartial({
            authority,
            params: params.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/miniwasm.tokenfactory.v1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateTokenfactoryParams.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    }
}
exports.MsgUpdateTokenfactoryParams = MsgUpdateTokenfactoryParams;
