// @initia/initia.js/dist/core/wasm/AccessConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessConfig = void 0;
const json_1 = require("../../util/json");
const types_1 = require("@initia/initia.proto/cosmwasm/wasm/v1/types");
class AccessConfig extends json_1.JSONSerializable {
    /**
     * @param permission
     * @param addresses
     */
    constructor(permission, addresses) {
        super();
        this.permission = permission;
        this.addresses = addresses;
    }
    static fromAmino(data) {
        const { permission, addresses } = data;
        return new AccessConfig((0, types_1.accessTypeFromJSON)(permission), addresses);
    }
    toAmino() {
        const { permission, addresses } = this;
        return {
            permission: (0, types_1.accessTypeToJSON)(permission),
            addresses,
        };
    }
    static fromData(data) {
        const { permission, addresses } = data;
        return new AccessConfig((0, types_1.accessTypeFromJSON)(permission), addresses);
    }
    toData() {
        const { permission, addresses } = this;
        return {
            permission: (0, types_1.accessTypeToJSON)(permission),
            addresses,
        };
    }
    static fromProto(data) {
        return new AccessConfig(data.permission, data.addresses);
    }
    toProto() {
        const { permission, addresses } = this;
        return types_1.AccessConfig.fromPartial({
            permission,
            addresses,
        });
    }
}
exports.AccessConfig = AccessConfig;
(function (AccessConfig) {
    AccessConfig.Type = types_1.AccessType;
})(AccessConfig || (exports.AccessConfig = AccessConfig = {}));
