// @initia/initia.proto/miniwasm/tokenfactory/v1/params.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const coin_1 = require("../../../cosmos/base/v1beta1/coin");
exports.protobufPackage = "miniwasm.tokenfactory.v1";
function createBaseParams() {
    return { denomCreationFee: [], denomCreationGasConsume: long_1.default.UZERO };
}
exports.Params = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.denomCreationFee) {
            coin_1.Coin.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (!message.denomCreationGasConsume.isZero()) {
            writer.uint32(16).uint64(message.denomCreationGasConsume);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.denomCreationFee.push(coin_1.Coin.decode(reader, reader.uint32()));
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.denomCreationGasConsume = reader.uint64();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            denomCreationFee: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.denomCreationFee)
                ? object.denomCreationFee.map((e) => coin_1.Coin.fromJSON(e))
                : [],
            denomCreationGasConsume: isSet(object.denomCreationGasConsume)
                ? long_1.default.fromValue(object.denomCreationGasConsume)
                : long_1.default.UZERO,
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.denomCreationFee) === null || _a === void 0 ? void 0 : _a.length) {
            obj.denomCreationFee = message.denomCreationFee.map((e) => coin_1.Coin.toJSON(e));
        }
        if (!message.denomCreationGasConsume.isZero()) {
            obj.denomCreationGasConsume = (message.denomCreationGasConsume || long_1.default.UZERO).toString();
        }
        return obj;
    },
    create(base) {
        return exports.Params.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseParams();
        message.denomCreationFee = ((_a = object.denomCreationFee) === null || _a === void 0 ? void 0 : _a.map((e) => coin_1.Coin.fromPartial(e))) || [];
        message.denomCreationGasConsume =
            object.denomCreationGasConsume !== undefined && object.denomCreationGasConsume !== null
                ? long_1.default.fromValue(object.denomCreationGasConsume)
                : long_1.default.UZERO;
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
