// @initia/initia.proto/initia/reward/v1/types.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const duration_1 = require("../../../google/protobuf/duration");
exports.protobufPackage = "initia.reward.v1";
function createBaseParams() {
    return {
        rewardDenom: "",
        dilutionPeriod: undefined,
        releaseRate: "",
        dilutionRate: "",
        releaseEnabled: false,
    };
}
exports.Params = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.rewardDenom !== "") {
            writer.uint32(10).string(message.rewardDenom);
        }
        if (message.dilutionPeriod !== undefined) {
            duration_1.Duration.encode(message.dilutionPeriod, writer.uint32(18).fork()).ldelim();
        }
        if (message.releaseRate !== "") {
            writer.uint32(26).string(message.releaseRate);
        }
        if (message.dilutionRate !== "") {
            writer.uint32(34).string(message.dilutionRate);
        }
        if (message.releaseEnabled === true) {
            writer.uint32(40).bool(message.releaseEnabled);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.rewardDenom = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.dilutionPeriod = duration_1.Duration.decode(reader, reader.uint32());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.releaseRate = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.dilutionRate = reader.string();
                    continue;
                case 5:
                    if (tag !== 40) {
                        break;
                    }
                    message.releaseEnabled = reader.bool();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            rewardDenom: isSet(object.rewardDenom) ? globalThis.String(object.rewardDenom) : "",
            dilutionPeriod: isSet(object.dilutionPeriod) ? duration_1.Duration.fromJSON(object.dilutionPeriod) : undefined,
            releaseRate: isSet(object.releaseRate) ? globalThis.String(object.releaseRate) : "",
            dilutionRate: isSet(object.dilutionRate) ? globalThis.String(object.dilutionRate) : "",
            releaseEnabled: isSet(object.releaseEnabled) ? globalThis.Boolean(object.releaseEnabled) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.rewardDenom !== "") {
            obj.rewardDenom = message.rewardDenom;
        }
        if (message.dilutionPeriod !== undefined) {
            obj.dilutionPeriod = duration_1.Duration.toJSON(message.dilutionPeriod);
        }
        if (message.releaseRate !== "") {
            obj.releaseRate = message.releaseRate;
        }
        if (message.dilutionRate !== "") {
            obj.dilutionRate = message.dilutionRate;
        }
        if (message.releaseEnabled === true) {
            obj.releaseEnabled = message.releaseEnabled;
        }
        return obj;
    },
    create(base) {
        return exports.Params.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseParams();
        message.rewardDenom = (_a = object.rewardDenom) !== null && _a !== void 0 ? _a : "";
        message.dilutionPeriod =
            object.dilutionPeriod !== undefined && object.dilutionPeriod !== null
                ? duration_1.Duration.fromPartial(object.dilutionPeriod)
                : undefined;
        message.releaseRate = (_b = object.releaseRate) !== null && _b !== void 0 ? _b : "";
        message.dilutionRate = (_c = object.dilutionRate) !== null && _c !== void 0 ? _c : "";
        message.releaseEnabled = (_d = object.releaseEnabled) !== null && _d !== void 0 ? _d : false;
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
