// @initia/initia.js/dist/core/wasm/msgs/MsgClearAdmin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgClearAdmin = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmwasm/wasm/v1/tx");
class MsgClearAdmin extends json_1.JSONSerializable {
    /**
     * @param sender the actor that signed the messages
     * @param contract the address of the smart contract
     */
    constructor(sender, contract) {
        super();
        this.sender = sender;
        this.contract = contract;
    }
    static fromAmino(data) {
        const { value: { sender, contract }, } = data;
        return new MsgClearAdmin(sender, contract);
    }
    toAmino() {
        const { sender, contract } = this;
        return {
            type: 'wasm/MsgClearAdmin',
            value: {
                sender,
                contract,
            },
        };
    }
    static fromData(data) {
        const { sender, contract } = data;
        return new MsgClearAdmin(sender, contract);
    }
    toData() {
        const { sender, contract } = this;
        return {
            '@type': '/cosmwasm.wasm.v1.MsgClearAdmin',
            sender,
            contract,
        };
    }
    static fromProto(data) {
        return new MsgClearAdmin(data.sender, data.contract);
    }
    toProto() {
        const { sender, contract } = this;
        return tx_1.MsgClearAdmin.fromPartial({
            sender,
            contract,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgClearAdmin',
            value: tx_1.MsgClearAdmin.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgClearAdmin.fromProto(tx_1.MsgClearAdmin.decode(msgAny.value));
    }
}
exports.MsgClearAdmin = MsgClearAdmin;
