// @initia/initia.js/dist/core/wasm/msgs/MsgInstantiateContractV2.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgInstantiateContractV2 = void 0;
const json_1 = require("../../../util/json");
const Coins_1 = require("../../Coins");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmwasm/wasm/v1/tx");
const long_1 = __importDefault(require("long"));
class MsgInstantiateContractV2 extends json_1.JSONSerializable {
    /**
     * @param sender the actor that signed the messages
     * @param admin an optional address that can execute migrations
     * @param code_id the reference to the stored WASM code
     * @param label optional metadata to be stored with a contract instance
     * @param msg json encoded message to be passed to the contract on instantiation
     * @param funds coins that are transferred to the contract on instantiation
     * @param salt an arbitrary value provided by the sender, can be 1 to 64
     * @param fix_msg include the msg value into the hash for the predictable address, default is false
     */
    constructor(sender, admin, code_id, label, msg, funds, salt, fix_msg) {
        super();
        this.sender = sender;
        this.admin = admin;
        this.code_id = code_id;
        this.label = label;
        this.msg = msg;
        this.salt = salt;
        this.fix_msg = fix_msg;
        this.funds = new Coins_1.Coins(funds);
    }
    static fromAmino(data) {
        const { value: { sender, admin, code_id, label, msg, funds, salt, fix_msg }, } = data;
        return new MsgInstantiateContractV2(sender, admin, Number.parseInt(code_id), label, msg, Coins_1.Coins.fromAmino(funds), salt, fix_msg);
    }
    toAmino() {
        const { sender, admin, code_id, label, msg, funds, salt, fix_msg } = this;
        return {
            type: 'wasm/MsgInstantiateContract2',
            value: {
                sender,
                admin,
                code_id: code_id.toString(),
                label,
                msg,
                funds: funds.toAmino(),
                salt,
                fix_msg,
            },
        };
    }
    static fromData(data) {
        const { sender, admin, code_id, label, msg, funds, salt, fix_msg } = data;
        return new MsgInstantiateContractV2(sender, admin, Number.parseInt(code_id), label, msg, Coins_1.Coins.fromData(funds), salt, fix_msg);
    }
    toData() {
        const { sender, admin, code_id, label, msg, funds, salt, fix_msg } = this;
        return {
            '@type': '/cosmwasm.wasm.v1.MsgInstantiateContract2',
            sender,
            admin,
            code_id: code_id.toString(),
            label,
            msg,
            funds: funds.toData(),
            salt,
            fix_msg,
        };
    }
    static fromProto(data) {
        return new MsgInstantiateContractV2(data.sender, data.admin, data.codeId.toNumber(), data.label, Buffer.from(data.msg).toString('base64'), Coins_1.Coins.fromProto(data.funds), Buffer.from(data.salt).toString('base64'), data.fixMsg);
    }
    toProto() {
        const { sender, admin, code_id, label, msg, funds, salt, fix_msg } = this;
        return tx_1.MsgInstantiateContract2.fromPartial({
            sender,
            admin,
            codeId: long_1.default.fromNumber(code_id),
            label,
            msg: Buffer.from(msg, 'base64'),
            funds: funds.toProto(),
            salt: Buffer.from(salt, 'base64'),
            fixMsg: fix_msg,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgInstantiateContract2',
            value: tx_1.MsgInstantiateContract2.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgInstantiateContractV2.fromProto(tx_1.MsgInstantiateContract2.decode(msgAny.value));
    }
}
exports.MsgInstantiateContractV2 = MsgInstantiateContractV2;
