// @initia/initia.js/dist/core/wasm/msgs/MsgStoreAndMigrateContract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgStoreAndMigrateContract = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmwasm/wasm/v1/tx");
const AccessConfig_1 = require("../AccessConfig");
class MsgStoreAndMigrateContract extends json_1.JSONSerializable {
    /**
     * @param authority the address of the governance account
     * @param wasm_byte_code can be raw or gzip compressed
     * @param instantiate_permission access control to apply on contract creation, optional
     * @param contract the address of the smart contract
     * @param msg json encoded message to be passed to the contract on migration
     */
    constructor(authority, wasm_byte_code, instantiate_permission, contract, msg) {
        super();
        this.authority = authority;
        this.wasm_byte_code = wasm_byte_code;
        this.instantiate_permission = instantiate_permission;
        this.contract = contract;
        this.msg = msg;
    }
    static fromAmino(data) {
        const { value: { authority, wasm_byte_code, instantiate_permission, contract, msg, }, } = data;
        return new MsgStoreAndMigrateContract(authority, wasm_byte_code, instantiate_permission
            ? AccessConfig_1.AccessConfig.fromAmino(instantiate_permission)
            : undefined, contract, msg);
    }
    toAmino() {
        const { authority, wasm_byte_code, instantiate_permission, contract, msg } = this;
        return {
            type: 'wasm/MsgStoreAndMigrateContract',
            value: {
                authority,
                wasm_byte_code,
                instantiate_permission: instantiate_permission === null || instantiate_permission === void 0 ? void 0 : instantiate_permission.toAmino(),
                contract,
                msg,
            },
        };
    }
    static fromData(data) {
        const { authority, wasm_byte_code, instantiate_permission, contract, msg } = data;
        return new MsgStoreAndMigrateContract(authority, wasm_byte_code, instantiate_permission
            ? AccessConfig_1.AccessConfig.fromData(instantiate_permission)
            : undefined, contract, msg);
    }
    toData() {
        const { authority, wasm_byte_code, instantiate_permission, contract, msg } = this;
        return {
            '@type': '/cosmwasm.wasm.v1.MsgStoreAndMigrateContract',
            authority,
            wasm_byte_code,
            instantiate_permission: instantiate_permission === null || instantiate_permission === void 0 ? void 0 : instantiate_permission.toData(),
            contract,
            msg,
        };
    }
    static fromProto(data) {
        return new MsgStoreAndMigrateContract(data.authority, Buffer.from(data.wasmByteCode).toString('base64'), data.instantiatePermission
            ? AccessConfig_1.AccessConfig.fromProto(data.instantiatePermission)
            : undefined, data.contract, Buffer.from(data.msg).toString('base64'));
    }
    toProto() {
        const { authority, wasm_byte_code, instantiate_permission, contract, msg } = this;
        return tx_1.MsgStoreAndMigrateContract.fromPartial({
            authority,
            wasmByteCode: Buffer.from(wasm_byte_code, 'base64'),
            instantiatePermission: instantiate_permission === null || instantiate_permission === void 0 ? void 0 : instantiate_permission.toProto(),
            contract,
            msg: Buffer.from(msg, 'base64'),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgStoreAndMigrateContract',
            value: tx_1.MsgStoreAndMigrateContract.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgStoreAndMigrateContract.fromProto(tx_1.MsgStoreAndMigrateContract.decode(msgAny.value));
    }
}
exports.MsgStoreAndMigrateContract = MsgStoreAndMigrateContract;
