// @initia/initia.js/dist/core/wasm/msgs/MsgStoreCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgStoreCode = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmwasm/wasm/v1/tx");
const AccessConfig_1 = require("../AccessConfig");
class MsgStoreCode extends json_1.JSONSerializable {
    /**
     * @param sender the actor that signed the messages
     * @param wasm_byte_code can be raw or gzip compressed
     * @param instantiate_permission access control to apply on contract creation, optional
     */
    constructor(sender, wasm_byte_code, instantiate_permission) {
        super();
        this.sender = sender;
        this.wasm_byte_code = wasm_byte_code;
        this.instantiate_permission = instantiate_permission;
    }
    static fromAmino(data) {
        const { value: { sender, wasm_byte_code, instantiate_permission }, } = data;
        return new MsgStoreCode(sender, wasm_byte_code, instantiate_permission
            ? AccessConfig_1.AccessConfig.fromAmino(instantiate_permission)
            : undefined);
    }
    toAmino() {
        const { sender, wasm_byte_code, instantiate_permission } = this;
        return {
            type: 'wasm/MsgStoreCode',
            value: {
                sender,
                wasm_byte_code,
                instantiate_permission: instantiate_permission === null || instantiate_permission === void 0 ? void 0 : instantiate_permission.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { sender, wasm_byte_code, instantiate_permission } = data;
        return new MsgStoreCode(sender, wasm_byte_code, instantiate_permission
            ? AccessConfig_1.AccessConfig.fromData(instantiate_permission)
            : undefined);
    }
    toData() {
        const { sender, wasm_byte_code, instantiate_permission } = this;
        return {
            '@type': '/cosmwasm.wasm.v1.MsgStoreCode',
            sender,
            wasm_byte_code,
            instantiate_permission: instantiate_permission === null || instantiate_permission === void 0 ? void 0 : instantiate_permission.toData(),
        };
    }
    static fromProto(data) {
        return new MsgStoreCode(data.sender, Buffer.from(data.wasmByteCode).toString('base64'), data.instantiatePermission
            ? AccessConfig_1.AccessConfig.fromProto(data.instantiatePermission)
            : undefined);
    }
    toProto() {
        const { sender, wasm_byte_code, instantiate_permission } = this;
        return tx_1.MsgStoreCode.fromPartial({
            sender,
            wasmByteCode: Buffer.from(wasm_byte_code, 'base64'),
            instantiatePermission: instantiate_permission === null || instantiate_permission === void 0 ? void 0 : instantiate_permission.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgStoreCode',
            value: tx_1.MsgStoreCode.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgStoreCode.fromProto(tx_1.MsgStoreCode.decode(msgAny.value));
    }
}
exports.MsgStoreCode = MsgStoreCode;
