// @initia/initia.js/dist/core/wasm/msgs/MsgSudoContract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgSudoContract = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmwasm/wasm/v1/tx");
class MsgSudoContract extends json_1.JSONSerializable {
    /**
     * @param authority the address of the governance account
     * @param contract the address of the smart contract
     * @param msg json encoded message to be passed to the contract as sudo
     */
    constructor(authority, contract, msg) {
        super();
        this.authority = authority;
        this.contract = contract;
        this.msg = msg;
    }
    static fromAmino(data) {
        const { value: { authority, contract, msg }, } = data;
        return new MsgSudoContract(authority, contract, msg);
    }
    toAmino() {
        const { authority, contract, msg } = this;
        return {
            type: 'wasm/MsgSudoContract',
            value: {
                authority,
                contract,
                msg,
            },
        };
    }
    static fromData(data) {
        const { authority, contract, msg } = data;
        return new MsgSudoContract(authority, contract, msg);
    }
    toData() {
        const { authority, contract, msg } = this;
        return {
            '@type': '/cosmwasm.wasm.v1.MsgSudoContract',
            authority,
            contract,
            msg,
        };
    }
    static fromProto(data) {
        return new MsgSudoContract(data.authority, data.contract, Buffer.from(data.msg).toString('base64'));
    }
    toProto() {
        const { authority, contract, msg } = this;
        return tx_1.MsgSudoContract.fromPartial({
            authority,
            contract,
            msg: Buffer.from(msg, 'base64'),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgSudoContract',
            value: tx_1.MsgSudoContract.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgSudoContract.fromProto(tx_1.MsgSudoContract.decode(msgAny.value));
    }
}
exports.MsgSudoContract = MsgSudoContract;
