// @initia/initia.js/dist/core/wasm/msgs/MsgUpdateInstantiateConfig.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateInstantiateConfig = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmwasm/wasm/v1/tx");
const AccessConfig_1 = require("../AccessConfig");
const long_1 = __importDefault(require("long"));
class MsgUpdateInstantiateConfig extends json_1.JSONSerializable {
    /**
     * @param sender the actor that signed the messages
     * @param code_id references the stored WASM code
     * @param new_instantiate_permission the new access control
     */
    constructor(sender, code_id, new_instantiate_permission) {
        super();
        this.sender = sender;
        this.code_id = code_id;
        this.new_instantiate_permission = new_instantiate_permission;
    }
    static fromAmino(data) {
        const { value: { sender, code_id, new_instantiate_permission }, } = data;
        return new MsgUpdateInstantiateConfig(sender, Number.parseInt(code_id), AccessConfig_1.AccessConfig.fromAmino(new_instantiate_permission));
    }
    toAmino() {
        const { sender, code_id, new_instantiate_permission } = this;
        return {
            type: 'wasm/MsgUpdateInstantiateConfig',
            value: {
                sender,
                code_id: code_id.toString(),
                new_instantiate_permission: new_instantiate_permission.toAmino(),
            },
        };
    }
    static fromData(data) {
        const { sender, code_id, new_instantiate_permission } = data;
        return new MsgUpdateInstantiateConfig(sender, Number.parseInt(code_id), AccessConfig_1.AccessConfig.fromData(new_instantiate_permission));
    }
    toData() {
        const { sender, code_id, new_instantiate_permission } = this;
        return {
            '@type': '/cosmwasm.wasm.v1.MsgUpdateInstantiateConfig',
            sender,
            code_id: code_id.toString(),
            new_instantiate_permission: new_instantiate_permission.toData(),
        };
    }
    static fromProto(data) {
        return new MsgUpdateInstantiateConfig(data.sender, data.codeId.toNumber(), AccessConfig_1.AccessConfig.fromProto(data.newInstantiatePermission));
    }
    toProto() {
        const { sender, code_id, new_instantiate_permission } = this;
        return tx_1.MsgUpdateInstantiateConfig.fromPartial({
            sender,
            codeId: long_1.default.fromNumber(code_id),
            newInstantiatePermission: new_instantiate_permission.toProto(),
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgUpdateInstantiateConfig',
            value: tx_1.MsgUpdateInstantiateConfig.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateInstantiateConfig.fromProto(tx_1.MsgUpdateInstantiateConfig.decode(msgAny.value));
    }
}
exports.MsgUpdateInstantiateConfig = MsgUpdateInstantiateConfig;
