// @initia/initia.js/dist/core/wasm/msgs/MsgUpdateContractLabel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateContractLabel = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/cosmwasm/wasm/v1/tx");
class MsgUpdateContractLabel extends json_1.JSONSerializable {
    /**
     * @param sender the actor that signed the messages
     * @param new_label string to be set
     * @param contract the address of the smart contract
     */
    constructor(sender, new_label, contract) {
        super();
        this.sender = sender;
        this.new_label = new_label;
        this.contract = contract;
    }
    static fromAmino(data) {
        const { value: { sender, new_label, contract }, } = data;
        return new MsgUpdateContractLabel(sender, new_label, contract);
    }
    toAmino() {
        const { sender, new_label, contract } = this;
        return {
            type: 'wasm/MsgUpdateContractLabel',
            value: {
                sender,
                new_label,
                contract,
            },
        };
    }
    static fromData(data) {
        const { sender, new_label, contract } = data;
        return new MsgUpdateContractLabel(sender, new_label, contract);
    }
    toData() {
        const { sender, new_label, contract } = this;
        return {
            '@type': '/cosmwasm.wasm.v1.MsgUpdateContractLabel',
            sender,
            new_label,
            contract,
        };
    }
    static fromProto(data) {
        return new MsgUpdateContractLabel(data.sender, data.newLabel, data.contract);
    }
    toProto() {
        const { sender, new_label, contract } = this;
        return tx_1.MsgUpdateContractLabel.fromPartial({
            sender,
            newLabel: new_label,
            contract,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgUpdateContractLabel',
            value: tx_1.MsgUpdateContractLabel.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateContractLabel.fromProto(tx_1.MsgUpdateContractLabel.decode(msgAny.value));
    }
}
exports.MsgUpdateContractLabel = MsgUpdateContractLabel;
