// @initia/initia.js/dist/core/oracle/CurrencyPair.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CurrencyPair = void 0;
const json_1 = require("../../util/json");
const genesis_1 = require("@initia/initia.proto/slinky/oracle/v1/genesis");
class CurrencyPair extends json_1.JSONSerializable {
    /**
     * @param Base
     * @param Quote
     */
    constructor(Base, Quote) {
        super();
        this.Base = Base;
        this.Quote = Quote;
    }
    toString() {
        return `${this.Base}/${this.Quote}`;
    }
    static fromAmino(data) {
        return new CurrencyPair(data.Base, data.Quote);
    }
    toAmino() {
        return {
            Base: this.Base,
            Quote: this.Quote,
        };
    }
    static fromData(data) {
        return new CurrencyPair(data.Base, data.Quote);
    }
    toData() {
        return {
            Base: this.Base,
            Quote: this.Quote,
        };
    }
    static fromProto(proto) {
        return new CurrencyPair(proto.Base, proto.Quote);
    }
    toProto() {
        return genesis_1.CurrencyPair.fromPartial({
            Base: this.Base,
            Quote: this.Quote,
        });
    }
}
exports.CurrencyPair = CurrencyPair;
