// @initia/initia.js/dist/core/forwarding/msgs/MsgClearForwardingAccount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgClearForwardingAccount = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/noble/forwarding/v1/tx");
class MsgClearForwardingAccount extends json_1.JSONSerializable {
    /**
     * @param signer
     * @param address
     */
    constructor(signer, address) {
        super();
        this.signer = signer;
        this.address = address;
    }
    static fromAmino(data) {
        const { value: { signer, address }, } = data;
        return new MsgClearForwardingAccount(signer, address);
    }
    toAmino() {
        const { signer, address } = this;
        return {
            type: 'noble/forwarding/ClearAccount',
            value: {
                signer,
                address,
            },
        };
    }
    static fromData(data) {
        const { signer, address } = data;
        return new MsgClearForwardingAccount(signer, address);
    }
    toData() {
        const { signer, address } = this;
        return {
            '@type': '/noble.forwarding.v1.MsgClearAccount',
            signer,
            address,
        };
    }
    static fromProto(data) {
        return new MsgClearForwardingAccount(data.signer, data.address);
    }
    toProto() {
        const { signer, address } = this;
        return tx_1.MsgClearAccount.fromPartial({
            signer,
            address,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/noble.forwarding.v1.MsgClearAccount',
            value: tx_1.MsgClearAccount.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgClearForwardingAccount.fromProto(tx_1.MsgClearAccount.decode(msgAny.value));
    }
}
exports.MsgClearForwardingAccount = MsgClearForwardingAccount;
