// @initia/initia.js/dist/core/intertx/msgs/MsgRegisterAccount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgRegisterAccount = void 0;
const json_1 = require("../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/initia/intertx/v1/tx");
class MsgRegisterAccount extends json_1.JSONSerializable {
    /**
     * @param owner
     * @param connection_id
     * @param version
     */
    constructor(owner, connection_id, version) {
        super();
        this.owner = owner;
        this.connection_id = connection_id;
        this.version = version;
    }
    static fromAmino(data) {
        const { value: { owner, connection_id, version }, } = data;
        return new MsgRegisterAccount(owner, connection_id, version);
    }
    toAmino() {
        const { owner, connection_id, version } = this;
        return {
            type: 'intertx/MsgRegisterAccount',
            value: { owner, connection_id, version },
        };
    }
    static fromData(data) {
        const { owner, connection_id, version } = data;
        return new MsgRegisterAccount(owner, connection_id, version);
    }
    toData() {
        const { owner, connection_id, version } = this;
        return {
            '@type': '/initia.intertx.v1.MsgRegisterAccount',
            owner,
            connection_id,
            version,
        };
    }
    static fromProto(proto) {
        return new MsgRegisterAccount(proto.owner, proto.connectionId, proto.version);
    }
    toProto() {
        const { owner, connection_id, version } = this;
        return tx_1.MsgRegisterAccount.fromPartial({
            owner,
            connectionId: connection_id,
            version,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/initia.intertx.v1.MsgRegisterAccount',
            value: tx_1.MsgRegisterAccount.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgRegisterAccount.fromProto(tx_1.MsgRegisterAccount.decode(msgAny.value));
    }
}
exports.MsgRegisterAccount = MsgRegisterAccount;
