// @initia/initia.js/dist/core/ibc/core/client/msgs/MsgCreateClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgCreateClient = void 0;
const json_1 = require("../../../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/core/client/v1/tx");
/**
 * MsgCreateClient defines a message to create an IBC client
 */
class MsgCreateClient extends json_1.JSONSerializable {
    /**
     * @param client_state light client state
     * @param consensus_state consensus state associated with the client that corresponds to a given
     * @param signer signer address
     */
    constructor(client_state, consensus_state, signer) {
        super();
        this.client_state = client_state;
        this.consensus_state = consensus_state;
        this.signer = signer;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { client_state, consensus_state, signer } = data;
        return new MsgCreateClient(client_state, consensus_state, signer);
    }
    toData() {
        const { client_state, consensus_state, signer } = this;
        return {
            '@type': '/ibc.core.client.v1.MsgCreateClient',
            client_state,
            consensus_state,
            signer,
        };
    }
    static fromProto(proto) {
        return new MsgCreateClient(proto.clientState, proto.consensusState, proto.signer);
    }
    toProto() {
        const { client_state, consensus_state, signer } = this;
        return tx_1.MsgCreateClient.fromPartial({
            clientState: client_state,
            consensusState: consensus_state,
            signer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.core.client.v1.MsgCreateClient',
            value: tx_1.MsgCreateClient.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgCreateClient.fromProto(tx_1.MsgCreateClient.decode(msgAny.value));
    }
}
exports.MsgCreateClient = MsgCreateClient;
