// @initia/initia.js/dist/core/ibc/core/client/msgs/MsgSubmitMisbehaviour.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgSubmitMisbehaviour = void 0;
const json_1 = require("../../../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/core/client/v1/tx");
/**
 *  MsgSubmitMisbehaviour defines an sdk.Msg type that submits Evidence for light client misbehaviour.
 */
class MsgSubmitMisbehaviour extends json_1.JSONSerializable {
    /**
     * @param client_id client unique identifier
     * @param misbehaviour misbehaviour used for freezing the light client
     * @param signer signer address
     */
    constructor(client_id, misbehaviour, signer) {
        super();
        this.client_id = client_id;
        this.misbehaviour = misbehaviour;
        this.signer = signer;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { client_id, misbehaviour, signer } = data;
        return new MsgSubmitMisbehaviour(client_id, misbehaviour, signer);
    }
    toData() {
        const { client_id, misbehaviour, signer } = this;
        return {
            '@type': '/ibc.core.client.v1.MsgSubmitMisbehaviour',
            client_id,
            misbehaviour,
            signer,
        };
    }
    static fromProto(proto) {
        return new MsgSubmitMisbehaviour(proto.clientId, proto.misbehaviour, proto.signer);
    }
    toProto() {
        const { client_id, misbehaviour, signer } = this;
        return tx_1.MsgSubmitMisbehaviour.fromPartial({
            clientId: client_id,
            misbehaviour,
            signer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.core.client.v1.MsgSubmitMisbehaviour',
            value: tx_1.MsgSubmitMisbehaviour.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgSubmitMisbehaviour.fromProto(tx_1.MsgSubmitMisbehaviour.decode(msgAny.value));
    }
}
exports.MsgSubmitMisbehaviour = MsgSubmitMisbehaviour;
