// @initia/initia.js/dist/core/ibc/core/client/msgs/MsgUpdateClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateClient = void 0;
const json_1 = require("../../../../../util/json");
const any_1 = require("@initia/initia.proto/google/protobuf/any");
const tx_1 = require("@initia/initia.proto/ibc/core/client/v1/tx");
/**
 * MsgUpdateClient defines an sdk.Msg to update a IBC client state using the given client message
 */
class MsgUpdateClient extends json_1.JSONSerializable {
    /**
     * @param client_id client unique identifier
     * @param client_message client message to update the light client
     * @param signer signer address
     */
    constructor(client_id, client_message, signer) {
        super();
        this.client_id = client_id;
        this.client_message = client_message;
        this.signer = signer;
    }
    static fromAmino(_) {
        _;
        throw new Error('Amino not supported');
    }
    toAmino() {
        throw new Error('Amino not supported');
    }
    static fromData(data) {
        const { client_id, client_message, signer } = data;
        return new MsgUpdateClient(client_id, client_message, signer);
    }
    toData() {
        const { client_id, client_message, signer } = this;
        return {
            '@type': '/ibc.core.client.v1.MsgUpdateClient',
            client_id,
            client_message,
            signer,
        };
    }
    static fromProto(proto) {
        return new MsgUpdateClient(proto.clientId, proto.clientMessage, proto.signer);
    }
    toProto() {
        const { client_id, client_message, signer } = this;
        return tx_1.MsgUpdateClient.fromPartial({
            clientId: client_id,
            clientMessage: client_message,
            signer,
        });
    }
    packAny() {
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.core.client.v1.MsgUpdateClient',
            value: tx_1.MsgUpdateClient.encode(this.toProto()).finish(),
        });
    }
    static unpackAny(msgAny) {
        return MsgUpdateClient.fromProto(tx_1.MsgUpdateClient.decode(msgAny.value));
    }
}
exports.MsgUpdateClient = MsgUpdateClient;
