// @initia/initia.proto/slinky/oracle/v1/query.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcWebError = exports.GrpcWebImpl = exports.QueryGetPricesDesc = exports.QueryGetPriceDesc = exports.QueryGetAllCurrencyPairsDesc = exports.QueryDesc = exports.QueryClientImpl = exports.GetPricesResponse = exports.GetPricesRequest = exports.GetPriceResponse = exports.GetPriceRequest = exports.GetAllCurrencyPairsResponse = exports.GetAllCurrencyPairsRequest = exports.protobufPackage = void 0;
/* eslint-disable */
const grpc_web_1 = require("@improbable-eng/grpc-web");
const browser_headers_1 = require("browser-headers");
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const genesis_1 = require("./genesis");
exports.protobufPackage = "slinky.oracle.v1";
function createBaseGetAllCurrencyPairsRequest() {
    return {};
}
exports.GetAllCurrencyPairsRequest = {
    encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetAllCurrencyPairsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.GetAllCurrencyPairsRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseGetAllCurrencyPairsRequest();
        return message;
    },
};
function createBaseGetAllCurrencyPairsResponse() {
    return { currencyPairs: [] };
}
exports.GetAllCurrencyPairsResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.currencyPairs) {
            genesis_1.CurrencyPair.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetAllCurrencyPairsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.currencyPairs.push(genesis_1.CurrencyPair.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            currencyPairs: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.currencyPairs)
                ? object.currencyPairs.map((e) => genesis_1.CurrencyPair.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.currencyPairs) === null || _a === void 0 ? void 0 : _a.length) {
            obj.currencyPairs = message.currencyPairs.map((e) => genesis_1.CurrencyPair.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.GetAllCurrencyPairsResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGetAllCurrencyPairsResponse();
        message.currencyPairs = ((_a = object.currencyPairs) === null || _a === void 0 ? void 0 : _a.map((e) => genesis_1.CurrencyPair.fromPartial(e))) || [];
        return message;
    },
};
function createBaseGetPriceRequest() {
    return { currencyPair: undefined, currencyPairId: undefined };
}
exports.GetPriceRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.currencyPair !== undefined) {
            genesis_1.CurrencyPair.encode(message.currencyPair, writer.uint32(10).fork()).ldelim();
        }
        if (message.currencyPairId !== undefined) {
            writer.uint32(18).string(message.currencyPairId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetPriceRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.currencyPair = genesis_1.CurrencyPair.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.currencyPairId = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            currencyPair: isSet(object.currencyPair) ? genesis_1.CurrencyPair.fromJSON(object.currencyPair) : undefined,
            currencyPairId: isSet(object.currencyPairId) ? globalThis.String(object.currencyPairId) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.currencyPair !== undefined) {
            obj.currencyPair = genesis_1.CurrencyPair.toJSON(message.currencyPair);
        }
        if (message.currencyPairId !== undefined) {
            obj.currencyPairId = message.currencyPairId;
        }
        return obj;
    },
    create(base) {
        return exports.GetPriceRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGetPriceRequest();
        message.currencyPair =
            object.currencyPair !== undefined && object.currencyPair !== null
                ? genesis_1.CurrencyPair.fromPartial(object.currencyPair)
                : undefined;
        message.currencyPairId = (_a = object.currencyPairId) !== null && _a !== void 0 ? _a : undefined;
        return message;
    },
};
function createBaseGetPriceResponse() {
    return { price: undefined, nonce: long_1.default.UZERO, decimals: long_1.default.UZERO, id: long_1.default.UZERO };
}
exports.GetPriceResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.price !== undefined) {
            genesis_1.QuotePrice.encode(message.price, writer.uint32(10).fork()).ldelim();
        }
        if (!message.nonce.isZero()) {
            writer.uint32(16).uint64(message.nonce);
        }
        if (!message.decimals.isZero()) {
            writer.uint32(24).uint64(message.decimals);
        }
        if (!message.id.isZero()) {
            writer.uint32(32).uint64(message.id);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetPriceResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.price = genesis_1.QuotePrice.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.nonce = reader.uint64();
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.decimals = reader.uint64();
                    continue;
                case 4:
                    if (tag !== 32) {
                        break;
                    }
                    message.id = reader.uint64();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            price: isSet(object.price) ? genesis_1.QuotePrice.fromJSON(object.price) : undefined,
            nonce: isSet(object.nonce) ? long_1.default.fromValue(object.nonce) : long_1.default.UZERO,
            decimals: isSet(object.decimals) ? long_1.default.fromValue(object.decimals) : long_1.default.UZERO,
            id: isSet(object.id) ? long_1.default.fromValue(object.id) : long_1.default.UZERO,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.price !== undefined) {
            obj.price = genesis_1.QuotePrice.toJSON(message.price);
        }
        if (!message.nonce.isZero()) {
            obj.nonce = (message.nonce || long_1.default.UZERO).toString();
        }
        if (!message.decimals.isZero()) {
            obj.decimals = (message.decimals || long_1.default.UZERO).toString();
        }
        if (!message.id.isZero()) {
            obj.id = (message.id || long_1.default.UZERO).toString();
        }
        return obj;
    },
    create(base) {
        return exports.GetPriceResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseGetPriceResponse();
        message.price =
            object.price !== undefined && object.price !== null ? genesis_1.QuotePrice.fromPartial(object.price) : undefined;
        message.nonce =
            object.nonce !== undefined && object.nonce !== null ? long_1.default.fromValue(object.nonce) : long_1.default.UZERO;
        message.decimals =
            object.decimals !== undefined && object.decimals !== null
                ? long_1.default.fromValue(object.decimals)
                : long_1.default.UZERO;
        message.id = object.id !== undefined && object.id !== null ? long_1.default.fromValue(object.id) : long_1.default.UZERO;
        return message;
    },
};
function createBaseGetPricesRequest() {
    return { currencyPairIds: [] };
}
exports.GetPricesRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.currencyPairIds) {
            writer.uint32(10).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetPricesRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.currencyPairIds.push(reader.string());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            currencyPairIds: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.currencyPairIds)
                ? object.currencyPairIds.map((e) => globalThis.String(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.currencyPairIds) === null || _a === void 0 ? void 0 : _a.length) {
            obj.currencyPairIds = message.currencyPairIds;
        }
        return obj;
    },
    create(base) {
        return exports.GetPricesRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGetPricesRequest();
        message.currencyPairIds = ((_a = object.currencyPairIds) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        return message;
    },
};
function createBaseGetPricesResponse() {
    return { prices: [] };
}
exports.GetPricesResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.prices) {
            exports.GetPriceResponse.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetPricesResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.prices.push(exports.GetPriceResponse.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            prices: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.prices)
                ? object.prices.map((e) => exports.GetPriceResponse.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.prices) === null || _a === void 0 ? void 0 : _a.length) {
            obj.prices = message.prices.map((e) => exports.GetPriceResponse.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.GetPricesResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGetPricesResponse();
        message.prices = ((_a = object.prices) === null || _a === void 0 ? void 0 : _a.map((e) => exports.GetPriceResponse.fromPartial(e))) || [];
        return message;
    },
};
class QueryClientImpl {
    constructor(rpc) {
        this.rpc = rpc;
        this.GetAllCurrencyPairs = this.GetAllCurrencyPairs.bind(this);
        this.GetPrice = this.GetPrice.bind(this);
        this.GetPrices = this.GetPrices.bind(this);
    }
    GetAllCurrencyPairs(request, metadata) {
        return this.rpc.unary(exports.QueryGetAllCurrencyPairsDesc, exports.GetAllCurrencyPairsRequest.fromPartial(request), metadata);
    }
    GetPrice(request, metadata) {
        return this.rpc.unary(exports.QueryGetPriceDesc, exports.GetPriceRequest.fromPartial(request), metadata);
    }
    GetPrices(request, metadata) {
        return this.rpc.unary(exports.QueryGetPricesDesc, exports.GetPricesRequest.fromPartial(request), metadata);
    }
}
exports.QueryClientImpl = QueryClientImpl;
exports.QueryDesc = { serviceName: "slinky.oracle.v1.Query" };
exports.QueryGetAllCurrencyPairsDesc = {
    methodName: "GetAllCurrencyPairs",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.GetAllCurrencyPairsRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.GetAllCurrencyPairsResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.QueryGetPriceDesc = {
    methodName: "GetPrice",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.GetPriceRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.GetPriceResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.QueryGetPricesDesc = {
    methodName: "GetPrices",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.GetPricesRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.GetPricesResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
class GrpcWebImpl {
    constructor(host, options) {
        this.host = host;
        this.options = options;
    }
    unary(methodDesc, _request, metadata) {
        var _a;
        const request = Object.assign(Object.assign({}, _request), methodDesc.requestType);
        const maybeCombinedMetadata = metadata && this.options.metadata
            ? new browser_headers_1.BrowserHeaders(Object.assign(Object.assign({}, (_a = this.options) === null || _a === void 0 ? void 0 : _a.metadata.headersMap), metadata === null || metadata === void 0 ? void 0 : metadata.headersMap))
            : metadata !== null && metadata !== void 0 ? metadata : this.options.metadata;
        return new Promise((resolve, reject) => {
            var _a;
            grpc_web_1.grpc.unary(methodDesc, Object.assign(Object.assign({ request, host: this.host, metadata: maybeCombinedMetadata !== null && maybeCombinedMetadata !== void 0 ? maybeCombinedMetadata : {} }, (this.options.transport !== undefined ? { transport: this.options.transport } : {})), { debug: (_a = this.options.debug) !== null && _a !== void 0 ? _a : false, onEnd: function (response) {
                    if (response.status === grpc_web_1.grpc.Code.OK) {
                        resolve(response.message.toObject());
                    }
                    else {
                        const err = new GrpcWebError(response.statusMessage, response.status, response.trailers);
                        reject(err);
                    }
                } }));
        });
    }
}
exports.GrpcWebImpl = GrpcWebImpl;
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
class GrpcWebError extends globalThis.Error {
    constructor(message, code, metadata) {
        super(message);
        this.code = code;
        this.metadata = metadata;
    }
}
exports.GrpcWebError = GrpcWebError;
